// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-table/ontario-table.js
import { h, getAssetPath } from '@stencil/core';
import { validateTableColumns, validateTableRowOptions } from './utils/ontario-table-validation';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { ConsoleType } from '../../utils/console-message/console-message.enum';
import { extractValuesByKey, organizeObjectKeys, removeObjectsBySpecificKey } from '../../utils/helper/utils';
export class OntarioTable {
	element;
	tableScrollWrapper;
	tableScrollDiv;
	table;
	/**
	 * Specifies the caption (or title) of the table.
	 *
	 * This is optional.
	 */
	caption;
	/**
	 * Used to define the columns of the table.
	 *
	 * @example;
	 * <ontario-table
	 *    table-columns='[
	 *      { "title": "Type of service", "key": "service" },
	 *      { "title": "Processing and delivery", "key": "processing" },
	 *      { "title": "Cost", "key": "cost", "type": "numeric" }
	 *    ]'
	 * >
	 * </ontario-table>
	 */
	tableColumns;
	/**
	 * Used to define the table body data. Note that the keys passed to the `data` object in the tableData should match the keys of the columns defined in the tableColumns prop.
	 *
	 * @example
	 * <ontario-table
	 *  table-data='[
	 *    {
	 *      "data": {
	 *        "service": "Regular service (online)",
	 *        "processing": "15 business days plus delivery by Canada Post",
	 *        "cost": "$15"
	 *      }
	 *    },
	 *    {
	 *      "data": {
	 *        "service": "Premium service (online)",
	 *        "cost": "$45",
	 *        "processing": "5 business days including delivery by courier"
	 *      }
	 *    }
	 *  ]'
	 * >
	 * </ontario-table>
	 */
	tableData;
	/**
	 * Indicates whether or not the table data should have alternate row zebra striping.
	 *
	 * This is optional. By default, zebra striping will be added when the table rows extend 5 rows. If zebra striping is needed to table rows less than 5 rows, the prop should be set to “enabled”. If no zebra stripes are needed, it should be set to “disabled”.
	 *
	 * The default will be set to “auto”.
	 */
	zebraStripes = 'auto';
	/**
	 * Used to specify whether or not table data in cells should have reduced top and bottom padding. This is useful for pages with multiple data-heavy tables such as a budget or financial data.
	 *
	 * This is optional. By default it will be set to “false”.
	 */
	condensed = false;
	/**
	 * Used to specify whether or not the table should extend the full width of its container.
	 *
	 * This is optional. By default, it will be set to “false”
	 */
	fullWidth = false;
	tableColumnsState;
	tableDataState;
	tableFooterState;
	processTableColumns() {
		this.tableColumns && this.parseOptions(this.tableColumns);
		validateTableColumns(this.tableColumnsState);
	}
	processTableData() {
		this.tableData && this.parseOptions(this.tableData);
		validateTableRowOptions(this.tableDataState);
	}
	// Parse the tableColumn & tableData options if they are strings
	// Transforms the data and stores it in respective state objects
	parseOptions(options) {
		const isString = typeof options === 'string';
		if (!options) {
			return;
		}
		try {
			if (options === this.tableColumns) {
				this.tableColumnsState = isString ? JSON.parse(options) : options;
			} else {
				this.tableDataState = isString ? JSON.parse(options) : options;
				this.transformTableData(this.tableDataState);
			}
		} catch (error) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addRegularText(' failed to parse props for ')
				.addMonospaceText('<ontario-table>')
				.addRegularText(' in ')
				.addMonospaceText('parseOptions()')
				.addRegularText(' method \n ')
				.addMonospaceText(error.stack)
				.printMessage(ConsoleType.Error);
		}
	}
	transformTableData = (tableData) => {
		// get column keys
		const columns = extractValuesByKey(this.tableColumnsState, 'key').filter((column) => column !== undefined);
		// organize table data data object according to the order of the column keys
		const organizedTableData = organizeObjectKeys(tableData, columns);
		// remove the footer table data into it's own state so that it can be rendered in a `tfoot` element
		const [updatedTableData, tableFooterData] = removeObjectsBySpecificKey(organizedTableData, 'footer', true);
		this.tableDataState = updatedTableData;
		this.tableFooterState = tableFooterData;
	};
	getZebraStripeClass = () => {
		return this.zebraStripes === 'disabled'
			? `ontario-table--no-zebra-stripes`
			: this.zebraStripes === 'enabled'
				? `ontario-table--zebra-stripes-enabled`
				: ``;
	};
	getTableClasses = () => {
		let tableClass = '';
		if (this.condensed && this.fullWidth) {
			tableClass = `ontario-table--condensed ontario-table--full-container-width`;
		} else if (this.condensed) {
			tableClass = `ontario-table--condensed`;
		} else if (this.fullWidth) {
			tableClass = `ontario-table--full-container-width`;
		}
		const zebraStripesClass = this.getZebraStripeClass();
		if (tableClass) {
			tableClass = `${tableClass} ${zebraStripesClass}`;
		} else {
			tableClass = zebraStripesClass;
		}
		return tableClass;
	};
	getColumnClasses = (columnData) => {
		if (columnData.type === 'numeric' && columnData.colSpan) {
			return `ontario-table-cell--numeric ontario-table-header--column-span-${columnData.colSpan}`;
		} else if (columnData.type === 'numeric') {
			return `ontario-table-cell--numeric`;
		} else if (columnData.colSpan) {
			return `ontario-table-header--column-span-${columnData.colSpan}`;
		}
		return;
	};
	// A function used to generate the table header and table definitions for the table body and table footer sections.
	generateTableDataHTML = (dataType, rowData, columns, rowClass) => {
		return h(
			'tr',
			{ class: dataType === 'tableData' ? rowClass : '' },
			this.tableColumnsState.map((column, index) => {
				const type = column.type;
				const tdClass = type === 'numeric' ? `ontario-table-cell--numeric` : '';
				return index === 0
					? h(
							'th',
							{ scope: 'row', innerHTML: rowData.data[`${columns[0]}`] },
							dataType === 'tableData' &&
								rowData.highlight &&
								h('img', {
									class: 'ontario-table--highlight-indicator',
									src: getAssetPath('./assets/highlight-indicator.svg'),
								}),
						)
					: h('td', { class: tdClass, innerHTML: rowData.data[`${columns[index]}`] });
			}),
		);
	};
	// Helper function to apply the scrollbar styles to the tops of tables
	applyScrollbar(tableElement, scrollerDiv) {
		scrollerDiv.style.visibility = 'visible';
		scrollerDiv.style.height = '20px';
		scrollerDiv.style.width = `${tableElement.scrollWidth}px`;
	}
	// The following logic adds scrollbar functionality to the tops of tables depending on their size.
	componentDidLoad() {
		const tables = this.table;
		const scrollerDivs = this.tableScrollDiv;
		const scrollerWrappers = this.tableScrollWrapper;
		let resizeObserver = new ResizeObserver(() => {
			this.applyScrollbar(tables, scrollerDivs);
		});
		this.applyScrollbar(tables, scrollerDivs);
		resizeObserver.observe(tables);
		tables.addEventListener('scroll', () => {
			this.applyScrollbar(tables, scrollerDivs);
			scrollerWrappers.scrollLeft = tables.scrollLeft;
		});
		scrollerWrappers.addEventListener('scroll', () => {
			this.applyScrollbar(tables, scrollerDivs);
			tables.scrollLeft = scrollerWrappers.scrollLeft;
		});
	}
	componentWillLoad() {
		this.processTableColumns();
		this.processTableData();
	}
	render() {
		const columns = extractValuesByKey(this.tableColumnsState, 'key');
		return h(
			'div',
			{ key: '4f8842108ac7007576fe5ea75ec83de5cfbb2853', class: 'ontario-table-container' },
			h(
				'div',
				{
					key: '69130c9552b854e80d267728f947032222eacbe7',
					class: 'ontario-table-scroll--wrapper',
					ref: (el) => (this.tableScrollWrapper = el),
				},
				h('div', {
					key: 'cbd058d782c1d5a2aa70787af688b1fbd3607ba6',
					class: 'ontario-table-scroll--div',
					ref: (el) => (this.tableScrollDiv = el),
				}),
			),
			h(
				'div',
				{ key: '743c382ed2826bc49014af4d26283c812faf0199', class: 'ontario-table-div', ref: (el) => (this.table = el) },
				h(
					'table',
					{ key: '6a1f6baed2928d4508aeb3260e750d1b52a189fa', class: this.getTableClasses() },
					this.caption && h('caption', { key: 'aa662bd73b1b2315642e624e31e288ea096f583c' }, this.caption),
					this.tableColumnsState &&
						h(
							'thead',
							{ key: 'c6844d209ce5eb7e5220cec715c1b0d858bd03a4' },
							h(
								'tr',
								{ key: '160fc34659ac78321cd3c9309282c42c4f83819e' },
								this.tableColumnsState.map((columnData) => {
									const columnClass = this.getColumnClasses(columnData);
									return h('th', {
										'scope': 'col',
										'key': columnData.key,
										'data-type': columnData.type,
										'class': columnClass,
										'innerHTML': columnData.title,
									});
								}),
							),
						),
					this.tableDataState &&
						h(
							'tbody',
							{ key: '7774577acb957958aa317917636ee27c52ae10be' },
							this.tableDataState.map((rowData) => {
								const rowClass = rowData.highlight
									? `ontario-table-row--highlight`
									: rowData.subtotal
										? `ontario-table-row--subtotal`
										: undefined;
								return this.generateTableDataHTML('tableData', rowData, columns, rowClass);
							}),
						),
					!!this.tableFooterState.length &&
						h(
							'tfoot',
							{ key: 'a293134380a5f33e008b05c419f371bd0c93308f' },
							this.tableFooterState.map((footerData) => {
								return this.generateTableDataHTML('tableData', footerData, columns, undefined);
							}),
						),
				),
			),
		);
	}
	static get is() {
		return 'ontario-table';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-table.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-table.css'],
		};
	}
	static get assetsDirs() {
		return ['./assets'];
	}
	static get properties() {
		return {
			caption: {
				type: 'string',
				attribute: 'caption',
				mutable: false,
				complexType: {
					original: 'string | undefined',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Specifies the caption (or title) of the table.\n\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			tableColumns: {
				type: 'string',
				attribute: 'table-columns',
				mutable: false,
				complexType: {
					original: 'string | TableColumnOptions[]',
					resolved: 'TableColumnOptions[] | string',
					references: {
						TableColumnOptions: {
							location: 'import',
							path: './table.interface',
							id: 'src/components/ontario-table/table.interface.ts::TableColumnOptions',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: ';\n<ontario-table\ntable-columns=\'[\n{ "title": "Type of service", "key": "service" },\n{ "title": "Processing and delivery", "key": "processing" },\n{ "title": "Cost", "key": "cost", "type": "numeric" }\n]\'\n>\n</ontario-table>',
						},
					],
					text: 'Used to define the columns of the table.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			tableData: {
				type: 'string',
				attribute: 'table-data',
				mutable: false,
				complexType: {
					original: 'string | TableRowOptions[]',
					resolved: 'TableRowOptions[] | string',
					references: {
						TableRowOptions: {
							location: 'import',
							path: './table.interface',
							id: 'src/components/ontario-table/table.interface.ts::TableRowOptions',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-table\n table-data=\'[\n   {\n     "data": {\n       "service": "Regular service (online)",\n       "processing": "15 business days plus delivery by Canada Post",\n       "cost": "$15"\n     }\n   },\n   {\n     "data": {\n       "service": "Premium service (online)",\n       "cost": "$45",\n       "processing": "5 business days including delivery by courier"\n     }\n   }\n ]\'\n>\n</ontario-table>',
						},
					],
					text: 'Used to define the table body data. Note that the keys passed to the `data` object in the tableData should match the keys of the columns defined in the tableColumns prop.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			zebraStripes: {
				type: 'string',
				attribute: 'zebra-stripes',
				mutable: false,
				complexType: {
					original: "'auto' | 'disabled' | 'enabled' | undefined",
					resolved: '"auto" | "disabled" | "enabled" | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Indicates whether or not the table data should have alternate row zebra striping.\n\nThis is optional. By default, zebra striping will be added when the table rows extend 5 rows. If zebra striping is needed to table rows less than 5 rows, the prop should be set to \u201Cenabled\u201D. If no zebra stripes are needed, it should be set to \u201Cdisabled\u201D.\n\nThe default will be set to \u201Cauto\u201D.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'auto'",
			},
			condensed: {
				type: 'boolean',
				attribute: 'condensed',
				mutable: false,
				complexType: {
					original: 'boolean | undefined',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to specify whether or not table data in cells should have reduced top and bottom padding. This is useful for pages with multiple data-heavy tables such as a budget or financial data.\n\nThis is optional. By default it will be set to \u201Cfalse\u201D.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			fullWidth: {
				type: 'boolean',
				attribute: 'full-width',
				mutable: false,
				complexType: {
					original: 'boolean | undefined',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to specify whether or not the table should extend the full width of its container.\n\nThis is optional. By default, it will be set to \u201Cfalse\u201D',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
		};
	}
	static get states() {
		return {
			tableColumnsState: {},
			tableDataState: {},
			tableFooterState: {},
		};
	}
	static get elementRef() {
		return 'element';
	}
	static get watchers() {
		return [
			{
				propName: 'tableColumns',
				methodName: 'processTableColumns',
			},
			{
				propName: 'tableData',
				methodName: 'processTableData',
			},
		];
	}
}
