// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-radio-buttons/ontario-radio-buttons.js
import { h } from '@stencil/core';
import { Input } from '../../utils/common/input/input';
import { InputCaption } from '../../utils/common/input-caption/input-caption';
import {
	validateObjectExists,
	validatePropExists,
	validateLanguage,
} from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { constructHintTextObject } from '../../utils/components/hints/hints';
import { EventType } from '../../utils/events/event-handler.interface';
import { handleInputEvent } from '../../utils/events/event-handler';
import { default as translations } from '../../translations/global.i18n.json';
import { ErrorMessage } from '../../utils/components/error-message/error-message';
export class OntarioRadioButtons {
	element;
	internals;
	hintTextRef;
	/**
	 * The text to display for the radio button legend.
	 *
	 * @example
	 * <ontario-radio-buttons
	 *   caption='{
	 *     "captionText": "Radio legend",
	 *     "captionType": "heading",
	 *    }'
	 *   required="true"
	 *   ...>
	 * </ontario-radio-buttons>
	 */
	caption;
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
	 */
	language;
	/**
	 * The name assigned to the radio button. The name value is used to reference form data after a form is submitted.
	 */
	name;
	/**
	 * Used to include the ontario-hint-text component for radio button group.
	 * This is optional.
	 */
	hintText;
	/**
	 * Used to include the ontario-hint-expander component for the radio button group.
	 * This is passed in as an object with key-value pairs.
	 *
	 * This is optional.
	 *
	 * @example
	 * <ontario-radio-buttons
	 *   caption='{
	 *     "captionText": "Radio legend",
	 *     "captionType": "heading",
	 *   }'
	 * 	 name="radios"
	 *   options='[
	 * 	   {
	 *        "value": "radio-option-1",
	 * 		  "elementId": "radio-1",
	 *        "label": "Radio option 1 label",
	 *        "hintExpander": {
	 *			  "hint": "Hint expander for radio option 1",
	 * 		      "content": "Example hint expander content for radio option 1."
	 *		  }
	 *     }
	 *   ]'
	 *   hint-expander='{
	 *     "hint": "Hint expander for the radio button group",
	 *     "content": "Example hint expander content for the radio button group."
	 *   }'
	 *   required="true"
	 * >
	 * </ontario-radio-buttons>
	 */
	hintExpander;
	/**
	 * This is used to determine whether the radio button is required or not.
	 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
	 * If no prop is set, it will default to false (optional).
	 */
	required = false;
	/**
	 * The options for the radio button group.
	 *
	 * Each property will be passed in through an object in the options array.
	 * This can either be passed in as an object directly (if using react), or as a string in HTML.
	 * If there are multiple radio buttons in a group, each radio button will be displayed as an option.
	 *
	 * In the example below, the options are being passed in as a string and there are two radio buttons to be displayed in the group.
	 *
	 * @example
	 * <ontario-radio-buttons
	 *   caption='{
	 *     "captionText": "Radio legend",
	 *     "captionType": "heading",
	 *   }'
	 *   name="radios"
	 *   hint-text="Hint text for the radio button group."
	 *   options='[
	 *     {
	 *        "value": "radio-option-1",
	 * 		  "elementId": "radio-1",
	 *        "label": "Radio option 1 label"
	 *     },
	 *     {
	 *        "value": "radio-option-2",
	 * 		  "elementId": "radio-2",
	 *        "label": "Radio option 2 label",
	 *        "hintExpander": {
	 *			  "hint": "Hint expander for radio option 2",
	 * 		      "content": "Example hint expander content for radio option 2."
	 *		  }
	 *      }
	 *   ]'
	 *   required="true"
	 * >
	 * </ontario-radio-buttons>
	 */
	options;
	/**
	 * Set this to display an error message
	 */
	errorMessage;
	/**
	 * Used to add a custom function to the radio input onChange event.
	 */
	customOnChange;
	/**
	 * Used to add a custom function to the radio input onBlur event.
	 */
	customOnBlur;
	/**
	 * Used to add a custom function to the radio input onFocus event.
	 */
	customOnFocus;
	/**
	 * Used for the `aria-describedby` value of the radio button group. This will match with the id of the hint text.
	 */
	hintTextId;
	/**
	 * The hint text options are re-assigned to the internalHintText array.
	 */
	internalHintText;
	/**
	 * The hint expander options are re-assigned to the internalHintExpander array.
	 */
	internalHintExpander;
	/**
	 * The options are re-assigned to the internalOptions array.
	 */
	internalOptions;
	/**
	 * Instantiate an InputCaption object for internal logic use
	 */
	captionState;
	/**
	 * Emitted when a keyboard input or mouse event occurs when a radio option has been changed.
	 */
	radioOnChange;
	/**
	 * Emitted when a keyboard input event occurs when a radio option has lost focus.
	 */
	radioOnBlur;
	/**
	 * Emitted when a keyboard input event occurs when a radio option has gained focus.
	 */
	radioOnFocus;
	/**
	 * Emitted when an error message is reported to the component.
	 */
	inputErrorOccurred;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 */
	handleSetAppLanguage(event) {
		if (!this.language) {
			this.language = validateLanguage(event);
		}
	}
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event);
	}
	/**
	 * Watch for changes to the `hintText` prop.
	 *
	 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
	 */
	parseHintText() {
		if (this.hintText) {
			const hintTextObject = constructHintTextObject(this.hintText);
			this.internalHintText = hintTextObject;
		}
	}
	/**
	 * Watch for changes to the `hintExpander` prop.
	 *
	 * If a `hintExpander` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalHintExpander` state.
	 */
	parseHintExpander() {
		const hintExpander = this.hintExpander;
		if (hintExpander) {
			if (typeof hintExpander === 'string') this.internalHintExpander = JSON.parse(hintExpander);
			else this.internalHintExpander = hintExpander;
		}
	}
	/**
	 * Watch for changes to the `options` prop.
	 *
	 * If an `options` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalOptions` state. The result will be run through a validation function.
	 */
	parseOptions() {
		if (typeof this.options !== 'undefined') {
			if (!Array.isArray(this.options)) {
				this.internalOptions = JSON.parse(this.options);
			} else {
				this.internalOptions = this.options;
			}
		}
	}
	/*
	 * Watch for changes in the `name` prop for validation purposes.
	 *
	 * Validate the `name` and make sure the `name` prop has a value.
	 * Log a warning if user doesn't input a value for the `name`.
	 */
	validateName(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' name ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-radio-buttons> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/*
     * Watch for changes in the `options` prop for validation purposes.

     * Validate the `options` and make sure the `options` prop has a value.
     * Log a warning if user doesn't input a value for the `options`.
     */
	validateOptions(newValue) {
		if (validateObjectExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' options ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-radio-buttons> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/**
	 * Watch for changes to the `caption` prop.
	 *
	 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
	 * @param newValue: Caption | string
	 */
	updateCaptionState(newValue) {
		this.captionState = new InputCaption(
			this.element.tagName,
			newValue,
			translations,
			this.language,
			true,
			this.required,
		);
	}
	/**
	 * Watch for changes to the `language` prop to render either the English or French translations
	 */
	updateLanguage() {
		this.updateCaptionState(this.caption);
	}
	broadcastInputErrorOccurredEvent() {
		// Emit event to notify anyone who wants to listen for errors occurring
		this.inputErrorOccurred.emit({ errorMessage: this.errorMessage ?? '' });
	}
	/**
	 * Function to handle radio buttons events and the information pertaining to the radio buttons to emit.
	 */
	handleEvent(event, eventType) {
		const input = event.target;
		// Reset all internalOptions checked states
		const changedOption = this.internalOptions
			.map((x) => {
				x.checked = false;
				return x;
			})
			.find((x) => x.value === input?.value);
		// Set the new checked state for the selected value
		if (changedOption) changedOption.checked = !changedOption?.checked;
		// Set the value within the form
		this.internals?.setFormValue?.(
			this.internalOptions
				.filter((x) => !!x.checked)
				.reduce((formData, currentValue) => {
					formData.append(this.name, currentValue.value);
					return formData;
				}, new FormData()),
		);
		handleInputEvent(
			event,
			eventType,
			input,
			this.radioOnChange,
			this.radioOnFocus,
			this.radioOnBlur,
			undefined,
			'radio',
			this.customOnChange,
			this.customOnFocus,
			this.customOnBlur,
			undefined,
			this.element,
		);
	}
	/**
	 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the fieldset `aria-describedBy` attribute.
	 */
	async componentDidLoad() {
		this.hintTextId = await this.hintTextRef?.getHintTextId();
	}
	componentWillLoad() {
		this.updateCaptionState(this.caption);
		this.parseOptions();
		this.parseHintText();
		this.parseHintExpander();
		this.validateName(this.name);
		this.validateOptions(this.internalOptions);
		this.language = validateLanguage(this.language);
	}
	render() {
		const error = !!this.errorMessage;
		return h(
			'div',
			{
				key: 'ba478aa73f71d17fba7f1a403f4758f6853b220b',
				class: `ontario-form-group ${error ? 'ontario-input--error' : ''}`,
			},
			h(
				'fieldset',
				{
					'key': 'cae7a78b9c05a3b69b9af229e5c3ba1e4328b340',
					'class': 'ontario-fieldset',
					'aria-describedby': this.hintTextId,
				},
				this.captionState.getCaption(undefined, !!this.internalHintExpander),
				this.internalHintText &&
					h('ontario-hint-text', {
						key: '122c784ca3b0219b6478b443cabc258d13ab3e7c',
						hint: this.internalHintText.hint,
						hintContentType: this.internalHintText.hintContentType,
						ref: (el) => (this.hintTextRef = el),
					}),
				h(
					'div',
					{ key: '1580f017c32064bc1ec65cb21a28aa3e253c2fdd', class: 'ontario-radios' },
					h(ErrorMessage, {
						key: 'd3c602b449e84c5d9b21836378835552fdeba767',
						message: this.errorMessage,
						error: error,
					}),
					this.internalOptions?.map((radioOption) =>
						h(
							'div',
							{ class: 'ontario-radios__item' },
							h(Input, {
								className: 'ontario-radios__input',
								id: radioOption.elementId,
								name: this.name,
								type: 'radio',
								value: radioOption.value,
								required: !!this.required,
								checked: !!radioOption.checked,
								onChange: (e) => this.handleEvent(e, EventType.Change),
								onBlur: (e) => this.handleEvent(e, EventType.Blur),
								onFocus: (e) => this.handleEvent(e, EventType.Focus),
							}),
							h(
								'label',
								{ class: 'ontario-radios__label', htmlFor: radioOption.elementId },
								radioOption.label,
								radioOption.hintExpander && this.captionState.getHintExpanderAccessibilityText(radioOption.label, true),
							),
							radioOption.hintExpander &&
								h(
									'div',
									{ class: 'ontario-radios__hint-expander' },
									h('ontario-hint-expander', {
										'hint': radioOption.hintExpander.hint,
										'content': radioOption.hintExpander.content,
										'hintContentType': radioOption.hintExpander.hintContentType,
										'input-exists': true,
									}),
								),
						),
					),
					this.internalHintExpander &&
						h(
							'div',
							{ key: '6cd63cae4be79a78c328c13288d348537c8551ad', class: 'ontario-radios__hint-expander' },
							h('ontario-hint-expander', {
								'key': 'e9abb0ecb680b50283438469d1425293e3a80440',
								'hint': this.internalHintExpander.hint,
								'content': this.internalHintExpander.content,
								'hintContentType': this.internalHintExpander.hintContentType,
								'input-exists': true,
							}),
						),
				),
			),
		);
	}
	static get is() {
		return 'ontario-radio-buttons';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get formAssociated() {
		return true;
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-radio-buttons.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-radio-buttons.css'],
		};
	}
	static get properties() {
		return {
			caption: {
				type: 'string',
				attribute: 'caption',
				mutable: false,
				complexType: {
					original: 'Caption | string',
					resolved: 'Caption | string',
					references: {
						Caption: {
							location: 'import',
							path: '../../utils/common/input-caption/caption.interface',
							id: 'src/utils/common/input-caption/caption.interface.ts::Caption',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-radio-buttons\n  caption=\'{\n    "captionText": "Radio legend",\n    "captionType": "heading",\n   }\'\n  required="true"\n  ...>\n</ontario-radio-buttons>',
						},
					],
					text: 'The text to display for the radio button legend.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr" | undefined',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			name: {
				type: 'string',
				attribute: 'name',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The name assigned to the radio button. The name value is used to reference form data after a form is submitted.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			hintText: {
				type: 'string',
				attribute: 'hint-text',
				mutable: false,
				complexType: {
					original: 'string | Hint',
					resolved: 'Hint | string | undefined',
					references: {
						Hint: {
							location: 'import',
							path: '../../utils/common/common.interface',
							id: 'src/utils/common/common.interface.ts::Hint',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to include the ontario-hint-text component for radio button group.\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			hintExpander: {
				type: 'string',
				attribute: 'hint-expander',
				mutable: false,
				complexType: {
					original: 'HintExpander | string',
					resolved: 'HintExpander | string | undefined',
					references: {
						HintExpander: {
							location: 'import',
							path: '../ontario-hint-expander/hint-expander.interface',
							id: 'src/components/ontario-hint-expander/hint-expander.interface.ts::HintExpander',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-radio-buttons\n  caption=\'{\n    "captionText": "Radio legend",\n    "captionType": "heading",\n  }\'\n\t name="radios"\n  options=\'[\n\t   {\n       "value": "radio-option-1",\n\t\t  "elementId": "radio-1",\n       "label": "Radio option 1 label",\n       "hintExpander": {\n\t\t  "hint": "Hint expander for radio option 1",\n\t\t      "content": "Example hint expander content for radio option 1."\n\t  }\n    }\n  ]\'\n  hint-expander=\'{\n    "hint": "Hint expander for the radio button group",\n    "content": "Example hint expander content for the radio button group."\n  }\'\n  required="true"\n>\n</ontario-radio-buttons>',
						},
					],
					text: 'Used to include the ontario-hint-expander component for the radio button group.\nThis is passed in as an object with key-value pairs.\n\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			required: {
				type: 'boolean',
				attribute: 'required',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'This is used to determine whether the radio button is required or not.\nThis prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.\nIf no prop is set, it will default to false (optional).',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			options: {
				type: 'string',
				attribute: 'options',
				mutable: false,
				complexType: {
					original: 'string | RadioOption[]',
					resolved: 'RadioOption[] | string',
					references: {
						RadioOption: {
							location: 'import',
							path: './radio-option.interface',
							id: 'src/components/ontario-radio-buttons/radio-option.interface.ts::RadioOption',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-radio-buttons\n  caption=\'{\n    "captionText": "Radio legend",\n    "captionType": "heading",\n  }\'\n  name="radios"\n  hint-text="Hint text for the radio button group."\n  options=\'[\n    {\n       "value": "radio-option-1",\n\t\t  "elementId": "radio-1",\n       "label": "Radio option 1 label"\n    },\n    {\n       "value": "radio-option-2",\n\t\t  "elementId": "radio-2",\n       "label": "Radio option 2 label",\n       "hintExpander": {\n\t\t  "hint": "Hint expander for radio option 2",\n\t\t      "content": "Example hint expander content for radio option 2."\n\t  }\n     }\n  ]\'\n  required="true"\n>\n</ontario-radio-buttons>',
						},
					],
					text: 'The options for the radio button group.\n\nEach property will be passed in through an object in the options array.\nThis can either be passed in as an object directly (if using react), or as a string in HTML.\nIf there are multiple radio buttons in a group, each radio button will be displayed as an option.\n\nIn the example below, the options are being passed in as a string and there are two radio buttons to be displayed in the group.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			errorMessage: {
				type: 'string',
				attribute: 'error-message',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Set this to display an error message',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			customOnChange: {
				type: 'unknown',
				attribute: 'custom-on-change',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the radio input onChange event.',
				},
				getter: false,
				setter: false,
			},
			customOnBlur: {
				type: 'unknown',
				attribute: 'custom-on-blur',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the radio input onBlur event.',
				},
				getter: false,
				setter: false,
			},
			customOnFocus: {
				type: 'unknown',
				attribute: 'custom-on-focus',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the radio input onFocus event.',
				},
				getter: false,
				setter: false,
			},
		};
	}
	static get states() {
		return {
			hintTextId: {},
			internalHintText: {},
			internalHintExpander: {},
			internalOptions: {},
			captionState: {},
		};
	}
	static get events() {
		return [
			{
				method: 'radioOnChange',
				name: 'radioOnChange',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input or mouse event occurs when a radio option has been changed.',
				},
				complexType: {
					original: 'RadioAndCheckboxChangeEvent',
					resolved: 'InputInteractionEvent & { checked: boolean; }',
					references: {
						RadioAndCheckboxChangeEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::RadioAndCheckboxChangeEvent',
						},
					},
				},
			},
			{
				method: 'radioOnBlur',
				name: 'radioOnBlur',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when a radio option has lost focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'radioOnFocus',
				name: 'radioOnFocus',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when a radio option has gained focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'inputErrorOccurred',
				name: 'inputErrorOccurred',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when an error message is reported to the component.',
				},
				complexType: {
					original: '{ errorMessage: string }',
					resolved: '{ errorMessage: string; }',
					references: {},
				},
			},
		];
	}
	static get elementRef() {
		return 'element';
	}
	static get watchers() {
		return [
			{
				propName: 'hintText',
				methodName: 'parseHintText',
			},
			{
				propName: 'hintExpander',
				methodName: 'parseHintExpander',
			},
			{
				propName: 'options',
				methodName: 'parseOptions',
			},
			{
				propName: 'name',
				methodName: 'validateName',
			},
			{
				propName: 'options',
				methodName: 'validateOptions',
			},
			{
				propName: 'caption',
				methodName: 'updateCaptionState',
			},
			{
				propName: 'language',
				methodName: 'updateLanguage',
			},
			{
				propName: 'errorMessage',
				methodName: 'broadcastInputErrorOccurredEvent',
			},
		];
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
		];
	}
	static get attachInternalsMemberName() {
		return 'internals';
	}
}
