// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-accordion/ontario-accordion.js
import { h } from '@stencil/core';
import { validateLanguage } from '../../utils/validation/validation-functions';
import translations from '../../translations/global.i18n.json';
export class OntarioAccordion {
	host;
	/**
	 * The name of the accordion component.
	 *
	 * This is not optional.
	 */
	name;
	/**
	 * Custom Expand/Collapse button text.
	 *
	 * @example
	 *  <ontario-accordion
	 *		name="My Accordion"
	 *		expand-collapse-button='{
	 *			"expandAllSectionsLabel": "Expand All",
	 *			"collapseAllSectionsLabel": "Collapse All"
	 *		}'
	 *		accordion-data='[
	 *			{"label": "Accordion 1", "content": ["Item 1", "Item 2", "Item 3"]},
	 *			{"label": "Accordion 2", "content": ["Item A", "Item B", "Item C"]}
	 *		]'
	 *	></ontario-accordion>
	 */
	expandCollapseButton;
	/**
	 * Used to include individual accordion data for the accordion component.
	 * This is passed in as an array of objects with key-value pairs.
	 *
	 * The `content` is expecting a string, that can either be written as HTML or a just a plain string, depending on the accordionContentType.
	 *
	 * @example
	 * 	<ontario-accordion
	 *		name="My Accordion"
	 *		accordion-data='[
	 *			{"label": "Accordion 1", "content": "This is a string"},
	 *			{"label": "Accordion 2", "accordionContentType": "html", "content": "<ul><li>List A</li><li>List B</li><li>List C</li></ul>"}
	 *		]'
	 *	></ontario-accordion>
	 */
	accordionData;
	/**
	 * Used to show whether the accordion is opened or closed.
	 */
	isOpen = false;
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If none are passed, it will default to English.
	 */
	language;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 */
	handleSetAppLanguage(event) {
		if (!this.language) {
			this.language = validateLanguage(event);
		}
	}
	/**
	 * Handle the language being toggled from the `<ontario-header>`.
	 * @param event Event object passed when the event is fired.
	 */
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event);
	}
	/**
	 * Store the translation dictionary for use within the component.
	 */
	translations = translations;
	/**
	 * The label for the expand/collapse button.
	 * This is internal and udpdated dynamically.
	 */
	expandCollapseLabel;
	/**
	 * Internal state of the expand/collapse label information.
	 */
	internalExpandCollapseLabelDetails;
	/**
	 * Internal state containing the parsed Accordion Data
	 */
	internalAccordionData = [];
	/**
	 * This state tracks which accordions are open.
	 */
	openAccordionIndexes = [];
	/**
	 * Parse Accordion data, this is used to handle JSON strings from HTML.
	 */
	parseAccordionData() {
		if (typeof this.accordionData !== 'undefined') {
			this.internalAccordionData = Array.isArray(this.accordionData)
				? this.accordionData
				: JSON.parse(this.accordionData);
		}
		// Initialize the label based on the initial accordion state
		this.updateLabel();
	}
	/**
	 * Parse Expand/Collapse Button Details, this is used to handle JSON strings from HTML.
	 */
	parseExpandCollapseButtonDetails() {
		if (typeof this.expandCollapseButton !== 'undefined') {
			this.internalExpandCollapseLabelDetails =
				typeof this.expandCollapseButton === 'string'
					? JSON.parse(this.expandCollapseButton)
					: this.expandCollapseButton;
		}
	}
	// Toggle the accordion state when it's clicked
	toggleAccordion(index) {
		if (this.openAccordionIndexes.includes(index)) {
			// If the accordion is already open, close it
			this.openAccordionIndexes = this.openAccordionIndexes.filter((i) => i !== index);
		} else {
			// If the accordion is closed, open it
			this.openAccordionIndexes = [...this.openAccordionIndexes, index];
		}
		this.updateLabel();
	}
	/**
	 * Toggle all accordions open/close
	 */
	toggleAll() {
		if (this.openAccordionIndexes.length === this.internalAccordionData.length) {
			// All accordions are open, close all
			this.openAccordionIndexes = [];
		} else {
			// At least one accordion is closed, open all
			this.openAccordionIndexes = this.internalAccordionData.map((_, index) => index);
		}
		this.updateLabel();
	}
	/**
	 * Update the label based on the current accordion state
	 */
	updateLabel() {
		const allOpen = this.internalAccordionData.every((_, index) => this.openAccordionIndexes.includes(index));
		if (allOpen) {
			// All accordions are open, set label to "Collapse all"
			this.expandCollapseLabel = 'collapse';
		} else {
			// At least one accordion is closed, set label to "Expand all"
			this.expandCollapseLabel = 'expand';
		}
	}
	componentWillLoad() {
		this.parseAccordionData();
		this.parseExpandCollapseButtonDetails();
		this.language = validateLanguage(this.language);
	}
	render() {
		return h(
			'div',
			{ key: '1ec9adce66ba9651eb72287a6268e041b8dc4fb3' },
			h('h2', { key: 'b13104816ae4a4ae5292afbc94ffb526c020babe' }, this.name),
			h(
				'div',
				{ key: '53e28363acaec45d55d0a021a68ebea3845b034d', class: 'ontario-accordions__container' },
				h(
					'div',
					{ key: 'eb4ee3f92067cce8ff9b169026eb1df0801fa230', class: 'ontario-accordion__controls' },
					h(
						'button',
						{
							'key': 'c018efcef59dcb152b3646e217df9b4ae09ed5c0',
							'class': 'ontario-accordion__button--expand-all',
							'onClick': () => this.toggleAll(),
							'aria-expanded':
								this.openAccordionIndexes.length === this.internalAccordionData.length ? 'true' : 'false',
							'aria-label': this.internalExpandCollapseLabelDetails?.ariaLabelText,
						},
						h(
							'span',
							{ key: '9e0f9959fb6f83e14b02e9843652d7a66a3fae54', class: 'ontario-accordion--expand-open-all' },
							this.expandCollapseLabel === 'expand'
								? h(
										'div',
										null,
										this.internalExpandCollapseLabelDetails?.expandAllSectionsLabel ??
											this.translations.accordion.expand[`${this.language}`],
									)
								: h(
										'div',
										null,
										this.internalExpandCollapseLabelDetails?.collapseAllSectionsLabel ??
											this.translations.accordion.collapse[`${this.language}`],
									),
						),
					),
				),
				this.internalAccordionData?.map((accordion, index) =>
					h(
						'div',
						{
							class: `ontario-accordion ${this.openAccordionIndexes.includes(index) ? 'open' : ''}`,
							key: `accordion-${index}`,
						},
						h(
							'h3',
							{
								class: `ontario-accordion__heading ${this.openAccordionIndexes.includes(index) ? 'ontario-expander--active' : ''}`,
							},
							h(
								'button',
								{
									'class': 'ontario-accordion__button',
									'aria-expanded': this.openAccordionIndexes.includes(index) ? 'true' : 'false',
									'data-toggle': 'ontario-collapse',
									'onClick': () => this.toggleAccordion(index),
									'aria-label': accordion.ariaLabelText,
								},
								h(
									'span',
									{ class: 'ontario-accordion__button-icon--close' },
									h('ontario-icon-chevron-up', { colour: 'blue' }),
								),
								h(
									'span',
									{ class: 'ontario-accordion__button-icon--open' },
									h('ontario-icon-chevron-down', { colour: 'blue' }),
								),
								accordion.label,
							),
						),
						h(
							'section',
							{
								'class': `ontario-accordion__content ${
									this.openAccordionIndexes.includes(index)
										? 'ontario-accordion__content--opened'
										: 'ontario-accordion__content--closed'
								}`,
								'aria-hidden': !this.openAccordionIndexes.includes(index),
								'data-toggle': 'ontario-expander-content',
							},
							accordion.accordionContentType === 'html'
								? h('div', { innerHTML: accordion.content }) // Render HTML content
								: h('div', null, accordion.content),
						),
					),
				),
			),
		);
	}
	static get is() {
		return 'ontario-accordion';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-accordion.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-accordion.css'],
		};
	}
	static get properties() {
		return {
			name: {
				type: 'string',
				attribute: 'name',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The name of the accordion component.\n\nThis is not optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			expandCollapseButton: {
				type: 'string',
				attribute: 'expand-collapse-button',
				mutable: false,
				complexType: {
					original: 'string | ExpandCollapseButtonDetails',
					resolved: 'ExpandCollapseButtonDetails | string | undefined',
					references: {
						ExpandCollapseButtonDetails: {
							location: 'import',
							path: './expandCollapseButtonDetails.interface',
							id: 'src/components/ontario-accordion/expandCollapseButtonDetails.interface.ts::ExpandCollapseButtonDetails',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: ' <ontario-accordion\n\tname="My Accordion"\n\texpand-collapse-button=\'{\n\t\t"expandAllSectionsLabel": "Expand All",\n\t\t"collapseAllSectionsLabel": "Collapse All"\n\t}\'\n\taccordion-data=\'[\n\t\t{"label": "Accordion 1", "content": ["Item 1", "Item 2", "Item 3"]},\n\t\t{"label": "Accordion 2", "content": ["Item A", "Item B", "Item C"]}\n\t]\'\n></ontario-accordion>',
						},
					],
					text: 'Custom Expand/Collapse button text.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			accordionData: {
				type: 'string',
				attribute: 'accordion-data',
				mutable: false,
				complexType: {
					original: 'string | Accordion[]',
					resolved: 'Accordion[] | string',
					references: {
						Accordion: {
							location: 'import',
							path: './accordion.interface',
							id: 'src/components/ontario-accordion/accordion.interface.ts::Accordion',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '\t<ontario-accordion\n\tname="My Accordion"\n\taccordion-data=\'[\n\t\t{"label": "Accordion 1", "content": "This is a string"},\n\t\t{"label": "Accordion 2", "accordionContentType": "html", "content": "<ul><li>List A</li><li>List B</li><li>List C</li></ul>"}\n\t]\'\n></ontario-accordion>',
						},
					],
					text: 'Used to include individual accordion data for the accordion component.\nThis is passed in as an array of objects with key-value pairs.\n\nThe `content` is expecting a string, that can either be written as HTML or a just a plain string, depending on the accordionContentType.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			isOpen: {
				type: 'boolean',
				attribute: 'is-open',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'Used to show whether the accordion is opened or closed.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr" | undefined',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If none are passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			translations: {},
			expandCollapseLabel: {},
			internalExpandCollapseLabelDetails: {},
			internalAccordionData: {},
			openAccordionIndexes: {},
		};
	}
	static get elementRef() {
		return 'host';
	}
	static get watchers() {
		return [
			{
				propName: 'accordionData',
				methodName: 'parseAccordionData',
			},
			{
				propName: 'expandCollapseButton',
				methodName: 'parseExpandCollapseButtonDetails',
			},
		];
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
		];
	}
}
