// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-step-indicator/ontario-step-indicator.js
import { h } from '@stencil/core';
import { validateLanguage } from '../../utils/validation/validation-functions';
import translations from '../../translations/global.i18n.json';
export class OntarioStepIndicator {
	host;
	/**
	 * A boolean value to determine whether or not the back button is displayed for the step indicator.
	 *
	 * This is optional. If no prop is passed, it will default to `false`.
	 */
	showBackButton = false;
	/**
	 * URL for the back element to set a path for where the link will lead.
	 *
	 * If a URL is passed in, the back element will display as an anchor tag.
	 * The back element will require either the backButtonURL prop or the customOnClick prop to be passed in order for the back element to display.
	 */
	backButtonUrl;
	/**
	 * A number value to indicate which step the user is currently on.
	 */
	currentStep;
	/**
	 * A number value to indicate to the user the total number of steps the form has.
	 */
	numberOfSteps;
	/**
	 * A number value to indicate to the user the percentage of the form that has been completed.
	 */
	percentageComplete;
	/**
	 * Used to add a custom function to the back button onClick event.
	 *
	 * If this function is passed in, the back element will display as a button.
	 * The back element will require either the backButtonURL prop or the customOnClick prop to be passed in order for the back element to display.
	 */
	customOnClick;
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If none are passed, it will default to English.
	 */
	language;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 */
	handleSetAppLanguage(event) {
		if (!this.language) {
			this.language = validateLanguage(event);
		}
	}
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event);
	}
	translations = translations;
	handleCustomOnClick = (e) => {
		this.customOnClick && this.customOnClick(e);
	};
	componentWillLoad() {
		this.language = validateLanguage(this.language);
	}
	render() {
		return h(
			'div',
			{ key: '99b28d54588042901db733dc07515778730104e0', class: 'ontario-step-indicator' },
			h(
				'div',
				{ key: '4d7d0de7f7573d237eb1efdb189d47aa8edb0c97', class: 'ontario-row' },
				h(
					'div',
					{ key: '59beefde92218fac06ea12335937a7eaa05f4a05', class: 'ontario-columns ontario-small-12' },
					h(
						'div',
						{
							key: '6f6b3b31ad381ec965e24d2d0cb71fcb87a55e8e',
							class: `ontario-step-indicator--with-back-button--${this.showBackButton}`,
						},
						this.showBackButton === true &&
							!this.backButtonUrl &&
							h(
								'button',
								{
									key: 'f81596ff5ef6e6c53921be03dd9d2156b6bebbfd',
									class: 'ontario-button ontario-button--tertiary',
									onClick: (e) => this.handleCustomOnClick(e),
								},
								h('ontario-icon-chevron-left', { key: 'a7f93e91eb046d3927abd6c10db07302e1aa09a8', colour: 'blue' }),
								this.translations.stepIndicator.back[`${this.language}`],
							),
						this.showBackButton === true &&
							this.backButtonUrl &&
							h(
								'a',
								{
									key: 'dcb41f75d9074ac2a2e423f75f5eafd1a628a221',
									class: 'ontario-button ontario-button--tertiary',
									href: this.backButtonUrl,
								},
								h('ontario-icon-chevron-left', { key: '6d396d47d555552128c442d96e0f90ed8a1d2290', colour: 'blue' }),
								this.translations.stepIndicator.back[`${this.language}`],
							),
						this.percentageComplete
							? h(
									'span',
									{ class: 'ontario-h4' },
									this.percentageComplete,
									this.language === 'en' ? '%' : h('span', null, '\u00A0%'),
									' ',
									this.translations.stepIndicator.complete[`${this.language}`],
								)
							: h(
									'span',
									{ class: 'ontario-h4' },
									this.translations.stepIndicator.step[`${this.language}`],
									'\u00A0',
									this.currentStep,
									' ',
									this.translations.stepIndicator.of[`${this.language}`],
									'\u00A0',
									this.numberOfSteps,
								),
					),
					h('hr', { key: '50951805695a30af40d20f7500f8dc8ede620f71' }),
				),
			),
		);
	}
	static get is() {
		return 'ontario-step-indicator';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-step-indicator.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-step-indicator.css'],
		};
	}
	static get properties() {
		return {
			showBackButton: {
				type: 'boolean',
				attribute: 'show-back-button',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'A boolean value to determine whether or not the back button is displayed for the step indicator.\n\nThis is optional. If no prop is passed, it will default to `false`.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			backButtonUrl: {
				type: 'string',
				attribute: 'back-button-url',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'URL for the back element to set a path for where the link will lead.\n\nIf a URL is passed in, the back element will display as an anchor tag.\nThe back element will require either the backButtonURL prop or the customOnClick prop to be passed in order for the back element to display.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			currentStep: {
				type: 'number',
				attribute: 'current-step',
				mutable: false,
				complexType: {
					original: 'number',
					resolved: 'number | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'A number value to indicate which step the user is currently on.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			numberOfSteps: {
				type: 'number',
				attribute: 'number-of-steps',
				mutable: false,
				complexType: {
					original: 'number',
					resolved: 'number | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'A number value to indicate to the user the total number of steps the form has.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			percentageComplete: {
				type: 'number',
				attribute: 'percentage-complete',
				mutable: false,
				complexType: {
					original: 'number',
					resolved: 'number | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'A number value to indicate to the user the percentage of the form that has been completed.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			customOnClick: {
				type: 'unknown',
				attribute: 'custom-on-click',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the back button onClick event.\n\nIf this function is passed in, the back element will display as a button.\nThe back element will require either the backButtonURL prop or the customOnClick prop to be passed in order for the back element to display.',
				},
				getter: false,
				setter: false,
			},
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr" | undefined',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If none are passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			translations: {},
		};
	}
	static get elementRef() {
		return 'host';
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
		];
	}
}
