// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-button/ontario-button.js
import { h } from '@stencil/core';
import { ButtonTypes, HtmlTypes } from './ontario-button.types';
import { validatePropExists, validateValueAgainstArray } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
export class OntarioButton {
	host;
	internals;
	/**
	 * The type of button to render.
	 *
	 * If no type is passed, it will default to 'secondary'.
	 */
	type = 'secondary';
	/**
	 * The native HTML button type the button should use.
	 *
	 * If no htmlType is passed, it will default to 'button'.
	 */
	htmlType = 'button';
	/**
	 * Text to be displayed within the button. This will override the text provided through the host element textContent.
	 *
	 * @example
	 * <ontario-button label="Label Text">Text</ontario-button>
	 *
	 * The resulting button will have the label `"Label Text"`.
	 */
	label;
	/**
	 * Provides more context as to what the button interaction is doing. This should only be used for accessibility purposes, if the button interaction requires more description than what the text provides.
	 *
	 *  This is optional.
	 *
	 * @example
	 * <ontario-button aria-label-text="Click button to open map">Open</ontario button>
	 */
	ariaLabelText;
	/**
	 * The unique identifier of the button. This is optional - if no ID is passed, one will be generated.
	 */
	elementId;
	/**
	 * Mutable variable, for internal use only.
	 * Set the button's type depending on validation result.
	 */
	typeState;
	/**
	 * Mutable variable, for internal use only.
	 * Set the native HTML button type depending on validation result.
	 */
	htmlTypeState;
	labelState;
	/**
	 * A reference to the internal button element.
	 */
	buttonRef;
	/*
	 * Watch for changes to the `label` property for validation purposes.
	 *
	 * If  no `label` prop is provided, the `label` prop will be set to the host element textContent (if it exists).
	 */
	updateLabelContent() {
		this.labelState = this.label ?? this.host.textContent ?? '';
		this.validateLabelContent(this.labelState);
	}
	/**
	 * Watch for changes to the `type` property for validation purposes.
	 *
	 * If the user input doesn't match one of the array values then `type` will be set to its default (`secondary`).
	 * If a match is found in one of the array values then `type` will be set to the matching array key value.
	 */
	validateType() {
		const isValid = validateValueAgainstArray(this.type, ButtonTypes);
		this.typeState = isValid ? this.type : this.warnDefaultType();
	}
	/**
	 *  Watch for changes to the `htmlType` property for validation purposes.
	 *
	 * If the user input doesn't match one of the array values then `htmlType` will be set to its default (`submit`).
	 * If a match is found in one of the array values then `htmlType` will be set to the matching array key value.
	 */
	validateHtmlType() {
		const isValid = validateValueAgainstArray(this.htmlType, HtmlTypes);
		this.htmlTypeState = isValid ? this.htmlType : this.warnDefaultHtmlType();
	}
	/**
	 * Print the missing `label` prop warning message
	 */
	validateLabelContent(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' label ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-button> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/**
	 * Print the invalid `type` prop warning message
	 * @returns default type (secondary)
	 */
	warnDefaultType() {
		const message = new ConsoleMessageClass();
		message
			.addDesignSystemTag()
			.addMonospaceText(' type ')
			.addRegularText('on')
			.addMonospaceText(' <ontario-button> ')
			.addRegularText('was set to an invalid type; only')
			.addMonospaceText(' primary, secondary, or tertiary ')
			.addRegularText('are supported. The default type')
			.addMonospaceText(' secondary ')
			.addRegularText('is assumed.')
			.printMessage();
		return 'secondary';
	}
	/**
	 * Print the invalid `htmlType` warning message
	 * @returns default htmlType (button)
	 */
	warnDefaultHtmlType() {
		const message = new ConsoleMessageClass();
		message
			.addDesignSystemTag()
			.addMonospaceText(' htmlType ')
			.addRegularText('on')
			.addMonospaceText(' <ontario-button> ')
			.addRegularText('was set to an invalid htmlType; only')
			.addMonospaceText(' button, reset, or submit ')
			.addRegularText('are supported. The default type')
			.addMonospaceText(' button ')
			.addRegularText('is assumed.')
			.printMessage();
		return 'button';
	}
	/**
	 * @returns the classes of the button based of the button's `type`.
	 */
	getClass() {
		return `ontario-button ontario-button--${this.typeState}`;
	}
	getId() {
		return this.elementId ?? '';
	}
	/**
	 * Set `buttonId`, `label`, and `ariaLabel` using internal component logic.
	 */
	componentWillLoad() {
		this.updateLabelContent();
		this.validateHtmlType();
		this.validateType();
		this.ariaLabelText = this.ariaLabelText ?? this.labelState;
	}
	componentDidLoad() {
		// Used to help load translations for any slots + text content passed in by the user.
		const observer = new MutationObserver((mutations) => {
			mutations.forEach((mutation) => {
				if (mutation.type === 'attributes') {
					this.updateLabelContent();
				}
			});
		});
		const options = { attributes: true };
		observer.observe(this.host, options);
		// Add a click event listener to handle submitting a form
		if (this.htmlTypeState === 'submit') {
			this.buttonRef.addEventListener('click', () => {
				const { form } = this.internals;
				form?.requestSubmit();
			});
		}
	}
	render() {
		return h(
			'button',
			{
				'key': '38158ea47d07a6da873acc3fba49ea3fecc534ce',
				'ref': (el) => (this.buttonRef = el),
				'type': this.htmlTypeState,
				'class': this.getClass(),
				'aria-label': this.ariaLabelText,
				'id': this.getId(),
			},
			this.labelState,
		);
	}
	static get is() {
		return 'ontario-button';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get formAssociated() {
		return true;
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-button.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-button.css'],
		};
	}
	static get properties() {
		return {
			type: {
				type: 'string',
				attribute: 'type',
				mutable: false,
				complexType: {
					original: 'ButtonType',
					resolved: '"internalThemeDark" | "primary" | "secondary" | "tertiary"',
					references: {
						ButtonType: {
							location: 'import',
							path: './ontario-button.types',
							id: 'src/components/ontario-button/ontario-button.types.ts::ButtonType',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "The type of button to render.\n\nIf no type is passed, it will default to 'secondary'.",
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'secondary'",
			},
			htmlType: {
				type: 'string',
				attribute: 'html-type',
				mutable: false,
				complexType: {
					original: 'HtmlType',
					resolved: '"button" | "reset" | "submit"',
					references: {
						HtmlType: {
							location: 'import',
							path: './ontario-button.types',
							id: 'src/components/ontario-button/ontario-button.types.ts::HtmlType',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "The native HTML button type the button should use.\n\nIf no htmlType is passed, it will default to 'button'.",
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'button'",
			},
			label: {
				type: 'string',
				attribute: 'label',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-button label="Label Text">Text</ontario-button>\n\nThe resulting button will have the label `"Label Text"`.',
						},
					],
					text: 'Text to be displayed within the button. This will override the text provided through the host element textContent.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			ariaLabelText: {
				type: 'string',
				attribute: 'aria-label-text',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-button aria-label-text="Click button to open map">Open</ontario button>',
						},
					],
					text: 'Provides more context as to what the button interaction is doing. This should only be used for accessibility purposes, if the button interaction requires more description than what the text provides.\n\n This is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			elementId: {
				type: 'string',
				attribute: 'element-id',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The unique identifier of the button. This is optional - if no ID is passed, one will be generated.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			typeState: {},
			htmlTypeState: {},
			labelState: {},
		};
	}
	static get elementRef() {
		return 'host';
	}
	static get watchers() {
		return [
			{
				propName: 'label',
				methodName: 'updateLabelContent',
			},
			{
				propName: 'type',
				methodName: 'validateType',
			},
			{
				propName: 'htmlType',
				methodName: 'validateHtmlType',
			},
		];
	}
	static get attachInternalsMemberName() {
		return 'internals';
	}
}
