// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-fieldset/ontario-fieldset.js
import { h } from '@stencil/core';
import { CaptionTypes } from '../../utils/common/input-caption/input-caption.types';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { validatePropExists, validateValueAgainstArray } from '../../utils/validation/validation-functions';
export class OntarioFieldset {
	/**
	 * The text value used for the legend of the fieldset.
	 */
	legend;
	/**
	 * The size of the fieldset legend. If no prop is passed, it will be `default`.
	 */
	legendSize = 'default';
	/**
	 * Watch for changes to the legendSize prop.
	 * This is for validation purposes to make sure the legendSize matches one of the CaptionTypes.
	 */
	validateLegendSize() {
		const isValid = validateValueAgainstArray(this.legendSize, CaptionTypes);
		if (isValid) {
			return this.legendSize;
		} else {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(` legendSize ${this.legendSize} `)
				.addRegularText('for')
				.addMonospaceText(' <ontario-fieldset> ')
				.addRegularText('is not a valid type. A default size will be applied.')
				.printMessage();
			return 'default';
		}
	}
	/*
	 * Watch for changes in the `legend` prop for validation purposes.
	 */
	validateLegend() {
		this.validateLegendText(this.legend);
	}
	/**
	 * Print the legend warning message
	 */
	validateLegendText(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' legend ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-fieldset> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	componentWillLoad() {
		this.validateLegend();
		this.validateLegendSize();
	}
	getClass() {
		switch (this.legendSize) {
			case 'large':
				return `ontario-fieldset__legend ontario-fieldset__legend--large`;
			case 'heading':
				return `ontario-fieldset__legend ontario-fieldset__legend--heading`;
			case 'default':
			default:
				return `ontario-fieldset__legend`;
		}
	}
	render() {
		return h(
			'div',
			{ key: '43597f1eb7279ccb96e8ef27b0a2bdab57e45a10', class: 'ontario-form-group' },
			h(
				'fieldset',
				{ key: '68737bd9cae2ce8871e07458adc97e878d668d0b', class: 'ontario-fieldset' },
				h(
					'legend',
					{ key: '6a831b206e46ba7531ed9ba47f145bc3279bf166', class: this.getClass() },
					this.legendSize === 'heading' ? h('h1', null, this.legend) : this.legend,
				),
				h('slot', { key: 'e6835d747a0a988fe1d890875af812ed66f86434' }),
			),
		);
	}
	static get is() {
		return 'ontario-fieldset';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-fieldset.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-fieldset.css'],
		};
	}
	static get properties() {
		return {
			legend: {
				type: 'string',
				attribute: 'legend',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The text value used for the legend of the fieldset.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			legendSize: {
				type: 'string',
				attribute: 'legend-size',
				mutable: true,
				complexType: {
					original: 'CaptionType',
					resolved: '"default" | "heading" | "large"',
					references: {
						CaptionType: {
							location: 'import',
							path: '../../utils/common/input-caption/input-caption.types',
							id: 'src/utils/common/input-caption/input-caption.types.ts::CaptionType',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The size of the fieldset legend. If no prop is passed, it will be `default`.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'default'",
			},
		};
	}
	static get watchers() {
		return [
			{
				propName: 'legendSize',
				methodName: 'validateLegendSize',
			},
			{
				propName: 'legend',
				methodName: 'validateLegend',
			},
		];
	}
}
