// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-date-input/components/input.js
import { h } from '@stencil/core';
import { Input as CommonInput } from '../../../utils/common/input/input';
export const Input = ({
	id,
	type,
	label,
	accessibilityLabel,
	placeholder,
	onInput,
	onChange,
	onFocus,
	onBlur,
	ariaDescribedBy,
	error = false,
	required = false,
}) => {
	const handleInputInput = (event) => {
		const newValue = event.target?.value ?? '';
		onInput(newValue, type);
	};
	const handleInputChange = (event) => {
		const newValue = event.target?.value ?? '';
		onChange(newValue, type);
	};
	const handleInputFocus = () => {
		onFocus(type);
	};
	const handleInputBlur = () => {
		onBlur(type);
	};
	return h(
		'div',
		{ class: `ontario-date__group-input ${error ? 'ontario-date--error' : ''}` },
		h('label', { htmlFor: id }, label, h('span', { class: 'ontario-show-for-sr' }, '(', accessibilityLabel, ')')),
		h(CommonInput, {
			className: 'ontario-input ontario-input--4-char-width',
			type: 'text',
			inputMode: 'numeric',
			id: id,
			required: !!required,
			placeholder: placeholder,
			onInput: handleInputInput,
			onChange: handleInputChange,
			onFocus: handleInputFocus,
			onBlur: handleInputBlur,
			ariaInvalid: error,
			ariaDescribedBy: ariaDescribedBy,
		}),
	);
};
