// @ongov/ontario-design-system-component-library/dist/components/ontario-button.js
import { p as proxyCustomElement, H, h } from './index.js';
import { v as validateValueAgainstArray, a as validatePropExists } from './p-CNqFEifG.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';

const ButtonTypes = ['primary', 'secondary', 'tertiary', 'internalThemeDark'];
const HtmlTypes = ['button', 'reset', 'submit'];

const ontarioButtonCss =
	'.ontario-button:focus,.ontario-button:active{box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:box-shadow 0.1s ease-in-out}.ontario-button{border:none;border-radius:4px;box-sizing:border-box;box-shadow:none;display:inline-block;font-size:1.125rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-weight:600;line-height:1.5555555556;margin:0 1.75rem 1.5rem 0;min-width:10rem;padding:0.625rem 1.5rem;text-align:center;text-decoration:none;cursor:pointer}@media screen and (max-width: 40em){.ontario-button{margin-right:0;display:block;width:100%}}.ontario-button .ontario-icon{margin-right:4px}.ontario-button--primary{background-color:#0066CC;color:#FFFFFF}.ontario-button--primary:hover{background-color:#00478F;color:#FFFFFF}.ontario-button--primary:focus{background-color:#00478F;color:#FFFFFF;transition:background-color 0.2s ease-out, box-shadow 0.1s ease-in-out}.ontario-button--primary:active{background-color:#002142;color:#FFFFFF;transition:background-color 0s, box-shadow 0.1s ease-in-out}.ontario-button--primary:visited{color:#FFFFFF}.ontario-button--secondary{background-color:#FFFFFF;border:2px solid #0066CC;color:#0066CC;padding-top:0.5rem;padding-bottom:0.5rem}.ontario-button--secondary:hover{background-color:#e0f0ff;border-color:#00478F;color:#00478F}.ontario-button--secondary:focus{background-color:#e0f0ff;border-color:#00478F;color:#00478F;transition:background-color 0.2s ease-out, box-shadow 0.1s ease-in-out}.ontario-button--secondary:active{background-color:#c2e0ff;border-color:#002142;color:#002142;transition:background-color 0s, box-shadow 0.1s ease-in-out}.ontario-button--secondary:visited{color:#0066CC}.ontario-button--tertiary{background-color:transparent;color:#0066CC;text-decoration:underline}.ontario-button--tertiary:hover{background-color:#e8e8e8;color:#00478F;text-decoration:underline}.ontario-button--tertiary:focus{background-color:#e8e8e8;color:#00478F;text-decoration:underline;transition:background-color 0.2s ease-out, box-shadow 0.1s ease-in-out}.ontario-button--tertiary:active{background-color:#d1d1d1;color:#002142;text-decoration:underline;transition:background-color 0s, box-shadow 0.1s ease-in-out}.ontario-button--tertiary:visited{color:#0066CC}.ontario-button--internalThemeDark{background-color:#1a1a1a;border:2px solid #FFFFFF;color:#FFFFFF}.ontario-button--internalThemeDark:hover{background-color:#4d4d4d}.ontario-button--internalThemeDark:focus{background-color:#4d4d4d;transition:background-color 0.2s ease-out, box-shadow 0.1s ease-in-out}.ontario-button--internalThemeDark:active{background-color:#4d4d4d;transition:background-color 0s, box-shadow 0.1s ease-in-out}.ontario-button--internalThemeDark:visited{color:#0066CC}';

const OntarioButton$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioButton extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
			this.internals = this.attachInternals();
		}
		get host() {
			return this;
		}
		internals;
		/**
		 * The type of button to render.
		 *
		 * If no type is passed, it will default to 'secondary'.
		 */
		type = 'secondary';
		/**
		 * The native HTML button type the button should use.
		 *
		 * If no htmlType is passed, it will default to 'button'.
		 */
		htmlType = 'button';
		/**
		 * Text to be displayed within the button. This will override the text provided through the host element textContent.
		 *
		 * @example
		 * <ontario-button label="Label Text">Text</ontario-button>
		 *
		 * The resulting button will have the label `"Label Text"`.
		 */
		label;
		/**
		 * Provides more context as to what the button interaction is doing. This should only be used for accessibility purposes, if the button interaction requires more description than what the text provides.
		 *
		 *  This is optional.
		 *
		 * @example
		 * <ontario-button aria-label-text="Click button to open map">Open</ontario button>
		 */
		ariaLabelText;
		/**
		 * The unique identifier of the button. This is optional - if no ID is passed, one will be generated.
		 */
		elementId;
		/**
		 * Mutable variable, for internal use only.
		 * Set the button's type depending on validation result.
		 */
		typeState;
		/**
		 * Mutable variable, for internal use only.
		 * Set the native HTML button type depending on validation result.
		 */
		htmlTypeState;
		labelState;
		/**
		 * A reference to the internal button element.
		 */
		buttonRef;
		/*
		 * Watch for changes to the `label` property for validation purposes.
		 *
		 * If  no `label` prop is provided, the `label` prop will be set to the host element textContent (if it exists).
		 */
		updateLabelContent() {
			this.labelState = this.label ?? this.host.textContent ?? '';
			this.validateLabelContent(this.labelState);
		}
		/**
		 * Watch for changes to the `type` property for validation purposes.
		 *
		 * If the user input doesn't match one of the array values then `type` will be set to its default (`secondary`).
		 * If a match is found in one of the array values then `type` will be set to the matching array key value.
		 */
		validateType() {
			const isValid = validateValueAgainstArray(this.type, ButtonTypes);
			this.typeState = isValid ? this.type : this.warnDefaultType();
		}
		/**
		 *  Watch for changes to the `htmlType` property for validation purposes.
		 *
		 * If the user input doesn't match one of the array values then `htmlType` will be set to its default (`submit`).
		 * If a match is found in one of the array values then `htmlType` will be set to the matching array key value.
		 */
		validateHtmlType() {
			const isValid = validateValueAgainstArray(this.htmlType, HtmlTypes);
			this.htmlTypeState = isValid ? this.htmlType : this.warnDefaultHtmlType();
		}
		/**
		 * Print the missing `label` prop warning message
		 */
		validateLabelContent(newValue) {
			if (validatePropExists(newValue)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' label ')
					.addRegularText('for')
					.addMonospaceText(' <ontario-button> ')
					.addRegularText('was not provided')
					.printMessage();
			}
		}
		/**
		 * Print the invalid `type` prop warning message
		 * @returns default type (secondary)
		 */
		warnDefaultType() {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' type ')
				.addRegularText('on')
				.addMonospaceText(' <ontario-button> ')
				.addRegularText('was set to an invalid type; only')
				.addMonospaceText(' primary, secondary, or tertiary ')
				.addRegularText('are supported. The default type')
				.addMonospaceText(' secondary ')
				.addRegularText('is assumed.')
				.printMessage();
			return 'secondary';
		}
		/**
		 * Print the invalid `htmlType` warning message
		 * @returns default htmlType (button)
		 */
		warnDefaultHtmlType() {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' htmlType ')
				.addRegularText('on')
				.addMonospaceText(' <ontario-button> ')
				.addRegularText('was set to an invalid htmlType; only')
				.addMonospaceText(' button, reset, or submit ')
				.addRegularText('are supported. The default type')
				.addMonospaceText(' button ')
				.addRegularText('is assumed.')
				.printMessage();
			return 'button';
		}
		/**
		 * @returns the classes of the button based of the button's `type`.
		 */
		getClass() {
			return `ontario-button ontario-button--${this.typeState}`;
		}
		getId() {
			return this.elementId ?? '';
		}
		/**
		 * Set `buttonId`, `label`, and `ariaLabel` using internal component logic.
		 */
		componentWillLoad() {
			this.updateLabelContent();
			this.validateHtmlType();
			this.validateType();
			this.ariaLabelText = this.ariaLabelText ?? this.labelState;
		}
		componentDidLoad() {
			// Used to help load translations for any slots + text content passed in by the user.
			const observer = new MutationObserver((mutations) => {
				mutations.forEach((mutation) => {
					if (mutation.type === 'attributes') {
						this.updateLabelContent();
					}
				});
			});
			const options = { attributes: true };
			observer.observe(this.host, options);
			// Add a click event listener to handle submitting a form
			if (this.htmlTypeState === 'submit') {
				this.buttonRef.addEventListener('click', () => {
					const { form } = this.internals;
					form?.requestSubmit();
				});
			}
		}
		render() {
			return h(
				'button',
				{
					'key': '38158ea47d07a6da873acc3fba49ea3fecc534ce',
					'ref': (el) => (this.buttonRef = el),
					'type': this.htmlTypeState,
					'class': this.getClass(),
					'aria-label': this.ariaLabelText,
					'id': this.getId(),
				},
				this.labelState,
			);
		}
		static get formAssociated() {
			return true;
		}
		static get watchers() {
			return {
				label: ['updateLabelContent'],
				type: ['validateType'],
				htmlType: ['validateHtmlType'],
			};
		}
		static get style() {
			return ontarioButtonCss;
		}
	},
	[
		65,
		'ontario-button',
		{
			type: [1],
			htmlType: [1, 'html-type'],
			label: [1],
			ariaLabelText: [1025, 'aria-label-text'],
			elementId: [1025, 'element-id'],
			typeState: [32],
			htmlTypeState: [32],
			labelState: [32],
		},
		undefined,
		{
			label: ['updateLabelContent'],
			type: ['validateType'],
			htmlType: ['validateHtmlType'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-button'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-button':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioButton$1);
				}
				break;
		}
	});
}

const OntarioButton = OntarioButton$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioButton, defineCustomElement };
