// @ongov/ontario-design-system-component-library/dist/components/ontario-textarea.js
import { p as proxyCustomElement, H, c as createEvent, h } from './index.js';
import { I as InputCaption, h as handleInputEvent, E as EventType } from './p-CJ1sSyqT.js';
import { b as validateLanguage, a as validatePropExists } from './p-CNqFEifG.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';
import { c as constructHintTextObject } from './p-CkusvkRr.js';
import { t as translations } from './p-DgMhpSmi.js';
import { E as ErrorMessage } from './p-DcoC7PVV.js';
import { d as defineCustomElement$6 } from './p-orSa-v1r.js';
import { d as defineCustomElement$5 } from './p-Br8HOVJk.js';
import { d as defineCustomElement$4 } from './p-BeNeXjVo.js';
import { d as defineCustomElement$3 } from './p-DyZbYpqr.js';
import { d as defineCustomElement$2 } from './p-pYZtHvSO.js';
import { v as v4 } from './p-BF0_OXTe.js';

const ontarioTextareaCss =
	'@charset "UTF-8";.ontario-textarea:focus,.ontario-textarea:active,.ontario-input:focus,.ontario-input:active{box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:box-shadow 0.1s ease-in-out}h6,.ontario-h6,h5,.ontario-h5,h4,.ontario-h4,h3,.ontario-h3,h2,.ontario-h2,h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-style:normal;font-weight:700;text-rendering:optimizeLegibility;margin-bottom:1rem;font-feature-settings:normal;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif}h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2rem;letter-spacing:0.04rem;line-height:1.29;margin:2.5rem 0 1.5rem 0;max-width:70rem}@media screen and (min-width: 40em){h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2.5rem;letter-spacing:0.04rem;line-height:1.2}}h2,.ontario-h2{font-size:1.6875rem;letter-spacing:0.03rem;line-height:1.37;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h2,.ontario-h2{font-size:2.0625rem;letter-spacing:0.02rem;line-height:1.33}}h3,.ontario-h3{font-size:1.4375rem;letter-spacing:0.02rem;line-height:1.39;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h3,.ontario-h3{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.43}}h4,.ontario-h4{font-size:1.25rem;letter-spacing:0.03rem;line-height:1.5;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h4,.ontario-h4{font-size:1.5rem;letter-spacing:0.0313rem;line-height:1.5}}h5,.ontario-h5{font-size:1.125rem;letter-spacing:0.03rem;line-height:1.56;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h5,.ontario-h5{font-size:1.1875rem;letter-spacing:0.025rem;line-height:1.5}}h6,.ontario-h6{font-size:1rem;line-height:1.56;letter-spacing:0.03rem;max-width:48rem;margin:0 0 0.75rem 0}@media screen and (min-width: 40em){h6,.ontario-h6{font-size:1rem;letter-spacing:0.025rem;line-height:1.5}}.ontario-lead-statement{font-size:1.25rem;line-height:1.6;max-width:70rem}@media screen and (min-width: 40em){.ontario-lead-statement{font-size:1.375rem}}html{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem}body{font-size:1rem;line-height:1.6;background-color:#FFFFFF;color:#1a1a1a;margin:0}p{font-size:inherit;margin-bottom:1.25rem;text-rendering:optimizeLegibility;margin:0 0 1.5rem 0}cite{font-style:italic;line-height:inherit}strong{font-weight:700;line-height:inherit}*,*::before,*::after{box-sizing:inherit}main p,main p:not(.ontario-lead-statement),main dd,main li{max-width:48rem}hr{border:0;padding:0;margin:2.5rem 0;border-bottom:4px solid #cccccc}hr.ontario-hr--dark{border-bottom:4px solid #1a1a1a}a{color:#0066CC;text-decoration:underline}a:visited{color:#551A8B}a:hover{color:#00478F}a:active{color:#002142}a:focus{-moz-box-shadow:0 0 0 4px #009ADB;-webkit-box-shadow:0 0 0 4px #009ADB;box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:all 0.1s ease-in-out}​p{margin:0 0 1.5rem 0}​p+h1,p+h2,p+h3,p+h4,p+h5,p+h6,ul+h1,ul+h2,ul+h3,ul+h4,ul+h5,ul+h6,ol+h1,ol+h2,ol+h3,ol+h4,ol+h5,ol+h6,h1+h2,h2+h3,h3+h4,h4+h5,h5+h6{margin-top:2rem}p code,td code,li code{line-height:2.25rem}input[type=text],input[type=number],input[type=email],input[type=date],input[type=color],input[type=search],input[type=tel],textarea{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}input[type=checkbox]+label,input[type=radio]+label{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400;margin-left:0;margin-right:0}fieldset{border:0;margin:0;padding:0;page-break-inside:avoid}legend,label{display:block;margin:0}select{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}ul,ol{padding:0}ul{margin-left:2rem}ul li{padding:0 0 0.75rem 0}ol{margin-left:1.75rem}ol li{padding:0 0 0.75rem 0.25rem}ul li:last-child,ol li:last-child{padding-bottom:0}ul[class*=columns-] li:last-child,ol[class*=columns-] li:last-child{padding-bottom:0.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 2rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 1.75rem}p+ul,p+ol{margin-top:-0.75rem}@media screen and (min-width: 40em){ul{margin:0.5rem 0 1.5rem 3rem}ol{margin:0.5rem 0 1.5rem 2.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 3rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 2.75rem}}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-form-group:last-of-type{margin-bottom:3rem}.ontario-fieldset__legend{color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;margin:0 0 1rem 0;line-height:1.5;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-fieldset__legend{font-size:1.125rem;line-height:1.56}}.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{letter-spacing:0.02rem;line-height:1.2}.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1{display:inline-block;margin:0}@media screen and (max-width: 40em){.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{font-size:1.4375rem;line-height:1.39}}.ontario-fieldset__legend--large{font-size:1.75rem}.ontario-fieldset__legend-required-flag,.ontario-fieldset__legend-optional-flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-fieldset__legend-required-flag:before,.ontario-fieldset__legend-optional-flag:before{content:" "}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-label{cursor:pointer;color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;line-height:1.5;margin:0 0 1rem 0;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-label{font-size:1.125rem;line-height:1.56}}.ontario-label--large{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.2;max-width:26.25em}@media screen and (max-width: 40em){.ontario-label--large{font-size:1.4375rem;line-height:1.39}}h1>.ontario-label--heading{display:inline-block;margin:0}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-input{border:2px solid #1a1a1a;border-radius:4px;box-sizing:border-box;color:#1a1a1a;display:block;font-size:1rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;line-height:1.5;margin:0 0 2.5rem;max-width:48rem;width:100%;padding:0.625rem 1rem;transition:box-shadow 0.1s ease-in-out}.ontario-input--2-char-width{max-width:8ex}.ontario-input--3-char-width{max-width:10ex}.ontario-input--4-char-width{max-width:12ex}.ontario-input--5-char-width{max-width:14ex}.ontario-input--7-char-width{max-width:17ex}.ontario-input--10-char-width{max-width:23ex}.ontario-input--20-char-width{max-width:41ex}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-hide{display:none !important}.ontario-invisible{visibility:hidden !important}@media screen and (max-width: 39.99875em){.ontario-hide-for-small-only{display:none !important}}@media screen and (max-width: 0em), screen and (min-width: 40em){.ontario-show-for-small-only{display:none !important}}@media print, screen and (min-width: 40em){.ontario-hide-for-medium{display:none !important}}@media screen and (max-width: 39.99875em){.ontario-show-for-medium{display:none !important}}@media screen and (min-width: 40em) and (max-width: 72.99875em){.ontario-hide-for-medium-only{display:none !important}}@media screen and (max-width: 39.99875em), screen and (min-width: 73em){.ontario-show-for-medium-only{display:none !important}}@media print, screen and (min-width: 73em){.ontario-hide-for-large{display:none !important}}@media screen and (max-width: 72.99875em){.ontario-show-for-large{display:none !important}}@media screen and (min-width: 73em) and (max-width: 95.99875em){.ontario-hide-for-large-only{display:none !important}}@media screen and (max-width: 72.99875em), screen and (min-width: 96em){.ontario-show-for-large-only{display:none !important}}.ontario-show-for-sr,.ontario-show-on-focus{position:absolute;width:1px;height:1px;overflow:hidden;clip:rect(0, 0, 0, 0)}.ontario-show-on-focus:active,.ontario-show-on-focus:focus{position:static;width:auto;height:auto;overflow:visible;clip:auto}.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}@media screen and (orientation: landscape){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}}@media screen and (orientation: portrait){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:none !important}}.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}@media screen and (orientation: landscape){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}}@media screen and (orientation: portrait){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:block !important}}.ontario-error-messaging{display:flex;align-items:flex-start;margin:-0.25rem 0 0.5rem 0;color:#D81A21;max-width:48rem}.ontario-error-messaging__content{margin-left:0.5rem}.ontario-input__error{border-color:#D81A21 !important}.ontario-error__hidden{display:none}.ontario-textarea{border:2px solid #1a1a1a;border-radius:4px;box-sizing:border-box;color:#1a1a1a;font-size:1rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;line-height:1.5;display:block;max-width:48rem;min-height:144px;margin:0 0 2.5rem;padding:0.625rem 1rem;transition:box-shadow 0.1s ease-in-out;white-space:pre-wrap;width:100%}.ontario-textarea.ontario-textarea-hint-expander--true{margin-bottom:1rem}.ontario-textarea--error textarea{border:0.125rem solid #CD0000}';

const OntarioTextarea$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioTextarea extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
			this.inputOnInput = createEvent(this, 'inputOnInput');
			this.inputOnChange = createEvent(this, 'inputOnChange');
			this.inputOnBlur = createEvent(this, 'inputOnBlur');
			this.inputOnFocus = createEvent(this, 'inputOnFocus');
			this.inputErrorOccurred = createEvent(this, 'inputErrorOccurred');
			this.internals = this.attachInternals();
		}
		get element() {
			return this;
		}
		internals;
		hintTextRef;
		/**
		 * The text to display as the textarea label.
		 *
		 * @example
		 * <ontario-input
		 *   caption='{
		 *     "captionText": "Address",
		 *     "captionType": "heading",
		 *   }'
		 *   required="true"
		 *   ...>
		 * </ontario-input>
		 */
		caption;
		/**
		 * The name assigned to the textarea. The name value is used to reference form data after a form is submitted.
		 */
		name;
		/**
		 * The unique identifier of the textarea. This is optional - if no ID is passed, one will be generated.
		 */
		elementId;
		/**
		 * This is used to determine whether the textarea is required or not.
		 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
		 * If no prop is set, it will default to false (optional).
		 */
		required = false;
		/**
		 * The textarea content value.
		 */
		value;
		/**
     * Used to include the ontario-hint-text component for the textarea.
     * This is optional.

     */
		hintText;
		/**
		 * Used to include the ontario-hint-expander component for the textarea component.
		 * This is passed in as an object with key-value pairs.
		 *
		 * This is optional.
		 *
		 * @example
		 * <ontario-textarea
		 *   caption='{
		 *     "captionText": "What are your thoughts",
		 *     "captionType": "heading",
		 *   }'
		 * 	 name: "textarea"
		 *   hint-expander='{
		 *     "hint": "This is the hint expander",
		 *     "content": "This is the content for the hint expander"
		 *   }'
		 *   required="true"
		 * >
		 * </ontario-textarea>
		 */
		hintExpander;
		/**
		 * The language of the component.
		 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
		 */
		language;
		/**
		 * Set this to display an error message
		 */
		errorMessage;
		/**
		 * Used to add a custom function to the textarea onInput event.
		 */
		customOnInput;
		/**
		 * Used to add a custom function to the textarea onChange event.
		 */
		customOnChange;
		/**
		 * Used to add a custom function to the textarea onBlur event.
		 */
		customOnBlur;
		/**
		 * Used to add a custom function to the textarea onFocus event.
		 */
		customOnFocus;
		/**
		 * Used for the `aria-describedby` value of the textarea. This will match with the id of the hint text.
		 */
		hintTextId;
		/**
		 * The hint text options are re-assigned to the internalHintText array.
		 */
		internalHintText;
		/**
		 * The hint expander options are re-assigned to the internalHintExpander array.
		 */
		internalHintExpander;
		/**
		 * Instantiate an InputCaption object for internal logic use
		 */
		captionState;
		/**
		 * Emitted when a input event occurs when an input has been changed.
		 */
		inputOnInput;
		/**
		 * Emitted when a keyboard input or mouse event occurs when an input has been changed.
		 */
		inputOnChange;
		/**
		 * Emitted when a keyboard input event occurs when an input has lost focus.
		 */
		inputOnBlur;
		/**
		 * Emitted when a keyboard input event occurs when an input has gained focus.
		 */
		inputOnFocus;
		/**
		 * Emitted when an error message is reported to the component.
		 */
		inputErrorOccurred;
		/**
		 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the textarea component loads.
		 */
		handleSetAppLanguage(event) {
			if (!this.language) {
				this.language = validateLanguage(event);
			}
		}
		handleHeaderLanguageToggled(event) {
			this.language = validateLanguage(event);
		}
		/**
		 * Watch for changes to the `hintText` prop.
		 *
		 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
		 */
		parseHintText() {
			if (this.hintText) {
				const hintTextObject = constructHintTextObject(this.hintText);
				this.internalHintText = hintTextObject;
			}
		}
		/**
		 * Watch for changes to the `hintExpander` prop.
		 *
		 * If a `hintExpander` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalHintExpander` state.
		 */
		parseHintExpander() {
			const hintExpander = this.hintExpander;
			if (hintExpander) {
				if (typeof hintExpander === 'string') this.internalHintExpander = JSON.parse(hintExpander);
				else this.internalHintExpander = hintExpander;
			}
		}
		/*
		 * Watch for changes in the `name` prop for validation purposes.
		 *
		 * Validate the `name` and make sure the `name` prop has a value.
		 * Log a warning if user doesn't input a value for the `name`.
		 */
		validateName(newValue) {
			if (validatePropExists(newValue)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' name ')
					.addRegularText('for')
					.addMonospaceText(' <ontario-textarea> ')
					.addRegularText('was not provided')
					.printMessage();
			}
		}
		/**
		 * Watch for changes to the `caption` prop.
		 *
		 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
		 * @param newValue: Caption | string
		 */
		updateCaptionState(newValue) {
			this.captionState = new InputCaption(
				this.element.tagName,
				newValue,
				translations,
				this.language,
				false,
				this.required,
			);
		}
		/**
		 * Watch for changes to the `language` prop to render either the English or French translations
		 */
		updateLanguage() {
			this.updateCaptionState(this.caption);
		}
		broadcastInputErrorOccurredEvent() {
			// Emit event to notify anyone who wants to listen for errors occurring
			this.inputErrorOccurred.emit({ inputId: this.getId(), errorMessage: this.errorMessage ?? '' });
		}
		/**
		 * Function to handle textarea events and the information pertaining to the textarea to emit.
		 */
		handleEvent(event, eventType) {
			const input = event.target;
			this.internals?.setFormValue?.(input?.value ?? '');
			handleInputEvent(
				event,
				eventType,
				input,
				this.inputOnChange,
				this.inputOnFocus,
				this.inputOnBlur,
				this.inputOnInput,
				'input',
				this.customOnChange,
				this.customOnFocus,
				this.customOnBlur,
				this.customOnInput,
				this.element,
			);
		}
		getId() {
			return this.elementId ?? '';
		}
		getValue() {
			return this.value ?? '';
		}
		getClass() {
			return this.hintExpander ? `ontario-textarea ontario-textarea-hint-expander--true` : `ontario-textarea`;
		}
		/**
		 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the textarea `aria-describedBy` attribute.
		 */
		async componentDidLoad() {
			this.hintTextId = await this.hintTextRef?.getHintTextId();
		}
		componentWillLoad() {
			this.updateCaptionState(this.caption);
			this.elementId = this.elementId ?? v4();
			this.parseHintText();
			this.parseHintExpander();
			this.validateName(this.name);
			this.language = validateLanguage(this.language);
		}
		render() {
			const error = !!this.errorMessage;
			return h(
				'div',
				{
					key: '0e8456b0840ae71541d3b201df095680d63650db',
					class: `ontario-form-group ${error ? 'ontario-textarea--error' : ''}`,
				},
				this.captionState.getCaption(this.getId(), !!this.internalHintExpander),
				this.internalHintText &&
					h('ontario-hint-text', {
						key: '21c33eb1f306b8280dcfa1d0def3b3d6aa0baf4b',
						hint: this.internalHintText.hint,
						hintContentType: this.internalHintText.hintContentType,
						ref: (el) => (this.hintTextRef = el),
					}),
				h(ErrorMessage, { key: 'd908a60fc99c09936564cb8d1f6652d2476f6e70', message: this.errorMessage, error: error }),
				h('textarea', {
					'key': '51d90ea63daf55f8c20b563551a7e20a5ece16f2',
					'aria-describedby': this.hintTextId,
					'class': this.getClass(),
					'id': this.getId(),
					'name': this.name,
					'value': this.getValue(),
					'onInput': (e) => this.handleEvent(e, EventType.Input),
					'onChange': (e) => this.handleEvent(e, EventType.Change),
					'onBlur': (e) => this.handleEvent(e, EventType.Blur),
					'onFocus': (e) => this.handleEvent(e, EventType.Focus),
					'required': !!this.required,
				}),
				this.internalHintExpander &&
					h('ontario-hint-expander', {
						'key': 'db0a618f1c0296dd1a6dcd014153b00925371d97',
						'hint': this.internalHintExpander.hint,
						'content': this.internalHintExpander.content,
						'hintContentType': this.internalHintExpander.hintContentType,
						'input-exists': true,
					}),
			);
		}
		static get formAssociated() {
			return true;
		}
		static get watchers() {
			return {
				hintText: ['parseHintText'],
				hintExpander: ['parseHintExpander'],
				name: ['validateName'],
				caption: ['updateCaptionState'],
				language: ['updateLanguage'],
				errorMessage: ['broadcastInputErrorOccurredEvent'],
			};
		}
		static get style() {
			return ontarioTextareaCss;
		}
	},
	[
		65,
		'ontario-textarea',
		{
			caption: [1],
			name: [1],
			elementId: [1025, 'element-id'],
			required: [4],
			value: [1025],
			hintText: [1, 'hint-text'],
			hintExpander: [1, 'hint-expander'],
			language: [1025],
			errorMessage: [1025, 'error-message'],
			customOnInput: [16, 'custom-on-input'],
			customOnChange: [16, 'custom-on-change'],
			customOnBlur: [16, 'custom-on-blur'],
			customOnFocus: [16, 'custom-on-focus'],
			hintTextId: [32],
			internalHintText: [32],
			internalHintExpander: [32],
			captionState: [32],
		},
		[
			[8, 'setAppLanguage', 'handleSetAppLanguage'],
			[8, 'headerLanguageToggled', 'handleHeaderLanguageToggled'],
		],
		{
			hintText: ['parseHintText'],
			hintExpander: ['parseHintExpander'],
			name: ['validateName'],
			caption: ['updateCaptionState'],
			language: ['updateLanguage'],
			errorMessage: ['broadcastInputErrorOccurredEvent'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = [
		'ontario-textarea',
		'ontario-hint-expander',
		'ontario-hint-text',
		'ontario-icon-alert-error',
		'ontario-icon-chevron-down',
		'ontario-icon-chevron-up',
	];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-textarea':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioTextarea$1);
				}
				break;
			case 'ontario-hint-expander':
				if (!customElements.get(tagName)) {
					defineCustomElement$6();
				}
				break;
			case 'ontario-hint-text':
				if (!customElements.get(tagName)) {
					defineCustomElement$5();
				}
				break;
			case 'ontario-icon-alert-error':
				if (!customElements.get(tagName)) {
					defineCustomElement$4();
				}
				break;
			case 'ontario-icon-chevron-down':
				if (!customElements.get(tagName)) {
					defineCustomElement$3();
				}
				break;
			case 'ontario-icon-chevron-up':
				if (!customElements.get(tagName)) {
					defineCustomElement$2();
				}
				break;
		}
	});
}

const OntarioTextarea = OntarioTextarea$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioTextarea, defineCustomElement };
