// @ongov/ontario-design-system-component-library/dist/components/ontario-input.js
import { p as proxyCustomElement, H, c as createEvent, h } from './index.js';
import { I as Input } from './p-BgeHsfjA.js';
import { I as InputCaption, h as handleInputEvent, E as EventType } from './p-CJ1sSyqT.js';
import { b as validateLanguage, a as validatePropExists } from './p-CNqFEifG.js';
import { t as translations } from './p-DgMhpSmi.js';
import { c as constructHintTextObject } from './p-CkusvkRr.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';
import { E as ErrorMessage } from './p-DcoC7PVV.js';
import { d as defineCustomElement$6 } from './p-orSa-v1r.js';
import { d as defineCustomElement$5 } from './p-Br8HOVJk.js';
import { d as defineCustomElement$4 } from './p-BeNeXjVo.js';
import { d as defineCustomElement$3 } from './p-DyZbYpqr.js';
import { d as defineCustomElement$2 } from './p-pYZtHvSO.js';
import { v as v4 } from './p-BF0_OXTe.js';

const ontarioInputCss =
	'@charset "UTF-8";.ontario-input:focus,.ontario-input:active{box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:box-shadow 0.1s ease-in-out}h6,.ontario-h6,h5,.ontario-h5,h4,.ontario-h4,h3,.ontario-h3,h2,.ontario-h2,h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-style:normal;font-weight:700;text-rendering:optimizeLegibility;margin-bottom:1rem;font-feature-settings:normal;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif}h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2rem;letter-spacing:0.04rem;line-height:1.29;margin:2.5rem 0 1.5rem 0;max-width:70rem}@media screen and (min-width: 40em){h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2.5rem;letter-spacing:0.04rem;line-height:1.2}}h2,.ontario-h2{font-size:1.6875rem;letter-spacing:0.03rem;line-height:1.37;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h2,.ontario-h2{font-size:2.0625rem;letter-spacing:0.02rem;line-height:1.33}}h3,.ontario-h3{font-size:1.4375rem;letter-spacing:0.02rem;line-height:1.39;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h3,.ontario-h3{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.43}}h4,.ontario-h4{font-size:1.25rem;letter-spacing:0.03rem;line-height:1.5;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h4,.ontario-h4{font-size:1.5rem;letter-spacing:0.0313rem;line-height:1.5}}h5,.ontario-h5{font-size:1.125rem;letter-spacing:0.03rem;line-height:1.56;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h5,.ontario-h5{font-size:1.1875rem;letter-spacing:0.025rem;line-height:1.5}}h6,.ontario-h6{font-size:1rem;line-height:1.56;letter-spacing:0.03rem;max-width:48rem;margin:0 0 0.75rem 0}@media screen and (min-width: 40em){h6,.ontario-h6{font-size:1rem;letter-spacing:0.025rem;line-height:1.5}}.ontario-lead-statement{font-size:1.25rem;line-height:1.6;max-width:70rem}@media screen and (min-width: 40em){.ontario-lead-statement{font-size:1.375rem}}html{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem}body{font-size:1rem;line-height:1.6;background-color:#FFFFFF;color:#1a1a1a;margin:0}p{font-size:inherit;margin-bottom:1.25rem;text-rendering:optimizeLegibility;margin:0 0 1.5rem 0}cite{font-style:italic;line-height:inherit}strong{font-weight:700;line-height:inherit}*,*::before,*::after{box-sizing:inherit}main p,main p:not(.ontario-lead-statement),main dd,main li{max-width:48rem}hr{border:0;padding:0;margin:2.5rem 0;border-bottom:4px solid #cccccc}hr.ontario-hr--dark{border-bottom:4px solid #1a1a1a}a{color:#0066CC;text-decoration:underline}a:visited{color:#551A8B}a:hover{color:#00478F}a:active{color:#002142}a:focus{-moz-box-shadow:0 0 0 4px #009ADB;-webkit-box-shadow:0 0 0 4px #009ADB;box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:all 0.1s ease-in-out}​p{margin:0 0 1.5rem 0}​p+h1,p+h2,p+h3,p+h4,p+h5,p+h6,ul+h1,ul+h2,ul+h3,ul+h4,ul+h5,ul+h6,ol+h1,ol+h2,ol+h3,ol+h4,ol+h5,ol+h6,h1+h2,h2+h3,h3+h4,h4+h5,h5+h6{margin-top:2rem}p code,td code,li code{line-height:2.25rem}input[type=text],input[type=number],input[type=email],input[type=date],input[type=color],input[type=search],input[type=tel],textarea{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}input[type=checkbox]+label,input[type=radio]+label{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400;margin-left:0;margin-right:0}fieldset{border:0;margin:0;padding:0;page-break-inside:avoid}legend,label{display:block;margin:0}select{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}ul,ol{padding:0}ul{margin-left:2rem}ul li{padding:0 0 0.75rem 0}ol{margin-left:1.75rem}ol li{padding:0 0 0.75rem 0.25rem}ul li:last-child,ol li:last-child{padding-bottom:0}ul[class*=columns-] li:last-child,ol[class*=columns-] li:last-child{padding-bottom:0.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 2rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 1.75rem}p+ul,p+ol{margin-top:-0.75rem}@media screen and (min-width: 40em){ul{margin:0.5rem 0 1.5rem 3rem}ol{margin:0.5rem 0 1.5rem 2.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 3rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 2.75rem}}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-form-group:last-of-type{margin-bottom:3rem}.ontario-fieldset__legend{color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;margin:0 0 1rem 0;line-height:1.5;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-fieldset__legend{font-size:1.125rem;line-height:1.56}}.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{letter-spacing:0.02rem;line-height:1.2}.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1{display:inline-block;margin:0}@media screen and (max-width: 40em){.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{font-size:1.4375rem;line-height:1.39}}.ontario-fieldset__legend--large{font-size:1.75rem}.ontario-fieldset__legend-required-flag,.ontario-fieldset__legend-optional-flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-fieldset__legend-required-flag:before,.ontario-fieldset__legend-optional-flag:before{content:" "}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-label{cursor:pointer;color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;line-height:1.5;margin:0 0 1rem 0;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-label{font-size:1.125rem;line-height:1.56}}.ontario-label--large{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.2;max-width:26.25em}@media screen and (max-width: 40em){.ontario-label--large{font-size:1.4375rem;line-height:1.39}}h1>.ontario-label--heading{display:inline-block;margin:0}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-input{border:2px solid #1a1a1a;border-radius:4px;box-sizing:border-box;color:#1a1a1a;display:block;font-size:1rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;line-height:1.5;margin:0 0 2.5rem;max-width:48rem;width:100%;padding:0.625rem 1rem;transition:box-shadow 0.1s ease-in-out}.ontario-input--2-char-width{max-width:8ex}.ontario-input--3-char-width{max-width:10ex}.ontario-input--4-char-width{max-width:12ex}.ontario-input--5-char-width{max-width:14ex}.ontario-input--7-char-width{max-width:17ex}.ontario-input--10-char-width{max-width:23ex}.ontario-input--20-char-width{max-width:41ex}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-hide{display:none !important}.ontario-invisible{visibility:hidden !important}@media screen and (max-width: 39.99875em){.ontario-hide-for-small-only{display:none !important}}@media screen and (max-width: 0em), screen and (min-width: 40em){.ontario-show-for-small-only{display:none !important}}@media print, screen and (min-width: 40em){.ontario-hide-for-medium{display:none !important}}@media screen and (max-width: 39.99875em){.ontario-show-for-medium{display:none !important}}@media screen and (min-width: 40em) and (max-width: 72.99875em){.ontario-hide-for-medium-only{display:none !important}}@media screen and (max-width: 39.99875em), screen and (min-width: 73em){.ontario-show-for-medium-only{display:none !important}}@media print, screen and (min-width: 73em){.ontario-hide-for-large{display:none !important}}@media screen and (max-width: 72.99875em){.ontario-show-for-large{display:none !important}}@media screen and (min-width: 73em) and (max-width: 95.99875em){.ontario-hide-for-large-only{display:none !important}}@media screen and (max-width: 72.99875em), screen and (min-width: 96em){.ontario-show-for-large-only{display:none !important}}.ontario-show-for-sr,.ontario-show-on-focus{position:absolute;width:1px;height:1px;overflow:hidden;clip:rect(0, 0, 0, 0)}.ontario-show-on-focus:active,.ontario-show-on-focus:focus{position:static;width:auto;height:auto;overflow:visible;clip:auto}.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}@media screen and (orientation: landscape){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}}@media screen and (orientation: portrait){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:none !important}}.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}@media screen and (orientation: landscape){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}}@media screen and (orientation: portrait){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:block !important}}.ontario-error-messaging{display:flex;align-items:flex-start;margin:-0.25rem 0 0.5rem 0;color:#D81A21;max-width:48rem}.ontario-error-messaging__content{margin-left:0.5rem}.ontario-input__error{border-color:#D81A21 !important}.ontario-error__hidden{display:none}.ontario-input{border:2px solid #1a1a1a;border-radius:4px;box-sizing:border-box;color:#1a1a1a;display:block;font-size:1rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;line-height:1.5;margin:0 0 2.5rem;max-width:48rem;padding:0.625rem 1rem;transition:box-shadow 0.1s ease-in-out;width:100%}.ontario-input-hint-expander--true{margin-bottom:1rem}.ontario-input--2-char-width{max-width:8ex}.ontario-input--3-char-width{max-width:10ex}.ontario-input--4-char-width{max-width:12ex}.ontario-input--5-char-width{max-width:14ex}.ontario-input--7-char-width{max-width:17ex}.ontario-input--10-char-width{max-width:23ex}.ontario-input--20-char-width{max-width:41ex}.ontario-input--error input{border:0.125rem solid #CD0000}';

const OntarioInput$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioInput extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
			this.inputOnInput = createEvent(this, 'inputOnInput');
			this.inputOnChange = createEvent(this, 'inputOnChange');
			this.inputOnBlur = createEvent(this, 'inputOnBlur');
			this.inputOnFocus = createEvent(this, 'inputOnFocus');
			this.inputErrorOccurred = createEvent(this, 'inputErrorOccurred');
			this.internals = this.attachInternals();
		}
		get element() {
			return this;
		}
		internals;
		hintTextRef;
		/**
		 * The text to display as the input label
		 *
		 * @example
		 * <ontario-input
		 *   caption='{
		 *     "captionText": "Address",
		 *     "captionType": "heading",
		 *   }
		 *   required="true"
		 *   ...>
		 * </ontario-input>
		 */
		caption;
		/**
		 * The unique identifier of the input. This is optional - if no ID is passed, one will be generated.
		 */
		elementId;
		/**
		 * The width of the input field. If no value is assigned, it will present as the `default` input width.
		 */
		inputWidth = 'default';
		/**
		 * The name assigned to the input. The name value is used to reference form data after a form is submitted.
		 */
		name;
		/**
		 * Used to include the ontario-hint-text component for the input.
		 * This is optional.
		 */
		hintText;
		/**
		 * This is used to determine whether the input is required or not.
		 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
		 * If no prop is set, it will default to false (optional).
		 *
		 * _Please add a validation messaging using `requiredValidationMessage` if setting this property._
		 *
		 * @example
		 * <ontario-input
		 *		id="address-line-1"
		 *		caption="Address line 1"
		 *		required
		 *		required-validation-message="Please enter an address, including street number and street name"
		 *		name="address-line-1"
		 *		hint-text="Street and number or P.O. box."
		 *	></ontario-input>
		 */
		required = false;
		/**
		 * The input type value.
		 *
		 * If no `type` is provided, it will default to 'text'.
		 */
		type = 'text';
		/**
		 * The input content value.
		 *
		 * This is optional.
		 */
		value;
		/**
		 * Set this to display an
		 */
		errorMessage;
		/**
		 * The language of the component.
		 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
		 */
		language = 'en';
		/**
		 * Used to include the ontario-hint-expander component for the input component.
		 * This is passed in as an object with key-value pairs.
		 *
		 * This is optional.
		 *
		 * @example
		 * <ontario-input
		 *   caption='{
		 *     "caption": "Address",
		 *     "captionType": "heading",
		 *   }
		 *   hint-expander='{
		 *    "hint": "Hint expander",
		 *    "content": "This is the content"
		 *   }'
		 *   required="true"
		 * >
		 * </ontario-input>
		 */
		hintExpander;
		/**
		 * Used for the `aria-describedby` value of the input. This will match with the id of the hint text.
		 */
		hintTextId;
		/**
		 * Enable live validation on the input.  Custom live validation can be performed using an `inputValidator`
		 * validation function.  It will also validate the `required` state if no errors are returned from
		 * the `inputValidator`.  Please set a `requiredValidationMessage` to report concisely to the end user what
		 * they are required to set.
		 */
		enableLiveValidation = false;
		/**
		 * Validate the validity of the input value `onBlur`.  This `async` function should return a result
		 * to trigger an error message.  Returning `undefined` or `null` will clear it.
		 */
		inputValidator;
		/**
		 * Used to add a custom function to the input onInput event.
		 */
		customOnInput;
		/**
		 * Used to add a custom function to the input onChange event.
		 */
		customOnChange;
		/**
		 * Used to add a custom function to the input onBlur event.
		 */
		customOnBlur;
		/**
		 * Used to add a custom function to the input onFocus event.
		 */
		customOnFocus;
		/**
		 * Custom error message to display if a required field is not filled out.  _Please add a
		 * custom message when setting an input as required_.
		 */
		requiredValidationMessage;
		/**
		 * The hint text options are re-assigned to the internalHintText array.
		 */
		internalHintText;
		/**
		 * The hint expander options are re-assigned to the internalHintExpander array.
		 */
		internalHintExpander;
		/**
		 * Instantiate an InputCaption object for internal logic use
		 */
		captionState;
		/**
		 * Track if the input has been interacted with, used to validate if
		 * a `required` field is in error.
		 */
		hasBeenInteractedWith = false;
		/**
		 * Global translations for accessing built-in translations
		 */
		translations = translations;
		/**
		 * Emitted when a input  occurs when an input has been changed.
		 */
		inputOnInput;
		/**
		 * Emitted when a keyboard input or mouse event occurs when an input has been changed.
		 */
		inputOnChange;
		/**
		 * Emitted when a keyboard input event occurs when an input has lost focus.
		 */
		inputOnBlur;
		/**
		 * Emitted when a keyboard input event occurs when an input has gained focus.
		 */
		inputOnFocus;
		/**
		 * Emitted when an error message is reported to the component.
		 */
		inputErrorOccurred;
		/**
		 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
		 * @param event The language that has been selected.
		 */
		handleSetAppLanguage(event) {
			this.language = validateLanguage(event);
		}
		/**
		 * Handles an update to the language should the user request a language update from the language toggle.
		 * @param event The language that has been selected.
		 */
		handleHeaderLanguageToggled(event) {
			this.language = validateLanguage(event.detail.newLanguage);
		}
		/**
		 * Handle the change in the `value` property and validate if the input has been interacted with by
		 * the user to aid in determining if the required state should produce an error.
		 */
		handleValueChange() {
			this.hasBeenInteractedWith = this.hasBeenInteractedWith || !!this.value;
		}
		/*
		 * Watch for changes in the `name` prop for validation purposes.
		 *
		 * Validate the `name` and make sure the `name` prop has a value.
		 * Log a warning if user doesn't input a value for the `name`.
		 */
		validateName(newValue) {
			if (validatePropExists(newValue)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' name ')
					.addRegularText('for')
					.addMonospaceText(' <ontario-input> ')
					.addRegularText('was not provided')
					.printMessage();
			}
		}
		/**
		 * Watch for changes to the `hintText` prop.
		 *
		 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
		 */
		parseHintText() {
			if (this.hintText) {
				const hintTextObject = constructHintTextObject(this.hintText);
				this.internalHintText = hintTextObject;
			}
		}
		/**
		 * Watch for changes to the `hintExpander` prop.
		 *
		 * If a `hintExpander` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalHintExpander` state.
		 */
		parseHintExpander() {
			const hintExpander = this.hintExpander;
			if (hintExpander) {
				if (typeof hintExpander === 'string') this.internalHintExpander = JSON.parse(hintExpander);
				else this.internalHintExpander = hintExpander;
			}
		}
		/**
		 * Watch for changes to the `caption` prop.
		 *
		 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
		 * @param newValue: Caption | string
		 */
		updateCaptionState(newValue) {
			this.captionState = new InputCaption(
				this.element.tagName,
				newValue,
				this.translations,
				this.language,
				false,
				this.required,
			);
		}
		/**
		 * Watch for changes in the `language` prop to render either the English or French translations
		 */
		updateLanguage() {
			this.updateCaptionState(this.caption);
		}
		/**
		 * Handle the component being blurred and perform validation logic on the input.  Custom validation
		 * takes persistance, followed by validating the required state.
		 *
		 * Finally, an event is emitted to notify anything listening for the `inputErrorOccurred` that
		 * an error occurred.
		 */
		async handleComponentBlur() {
			if (this.enableLiveValidation) {
				// Call inputValidator function to perform custom validation
				const validationResult = this.inputValidator && this.inputValidator(this.value);
				await validationResult?.then((x) => (this.errorMessage = x?.errorMessage));
				// Validate the `required` field
				// Only report a required error if no other error message is reported via validation
				if (this.required && this.hasBeenInteractedWith && !validationResult)
					if (!this.value)
						this.errorMessage =
							this.requiredValidationMessage || this.translations.input.requiredFieldError[this.getComponentLanguage()];
					else this.errorMessage = undefined;
			}
		}
		broadcastInputErrorOccurredEvent() {
			// Emit event to notify anyone who wants to listen for errors occurring
			this.inputErrorOccurred.emit({ inputId: this.getId(), errorMessage: this.errorMessage ?? '' });
		}
		/**
		 * Function to handle input events and the information pertaining to the input to emit.
		 */
		handleEvent(event, eventType) {
			const input = event.target;
			// Update the component value to match the value of the input element.
			this.value = input?.value;
			this.internals?.setFormValue?.(this.value ?? '');
			handleInputEvent(
				event,
				eventType,
				input,
				this.inputOnChange,
				this.inputOnFocus,
				this.inputOnBlur,
				this.inputOnInput,
				'input',
				this.customOnChange,
				this.customOnFocus,
				this.customOnBlur,
				this.customOnInput,
				this.element,
			);
		}
		getId() {
			// A UUID is assigned in `componentWillLoad` if there is no given `elementId`.
			return this.elementId ?? '';
		}
		getValue() {
			return this.value ?? '';
		}
		getClass() {
			if (this.hintExpander) {
				return this.inputWidth === 'default'
					? `ontario-input ontario-input-hint-expander--true`
					: `ontario-input ontario-input--${this.inputWidth} ontario-input-hint-expander--true`;
			} else {
				return this.inputWidth === 'default' ? `ontario-input` : `ontario-input ontario-input--${this.inputWidth}`;
			}
		}
		getComponentLanguage() {
			return this.language ?? 'en';
		}
		/**
		 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the input `aria-describedBy` attribute.
		 */
		async componentDidLoad() {
			this.hintTextId = await this.hintTextRef?.getHintTextId();
		}
		componentWillLoad() {
			this.updateCaptionState(this.caption);
			this.elementId = this.elementId ?? v4();
			this.parseHintText();
			this.parseHintExpander();
			this.validateName(this.name);
			this.language = validateLanguage(this.language);
		}
		render() {
			const error = !!this.errorMessage;
			return h(
				'div',
				{
					key: 'a7222e6476454f5d3d7063306b93db212abb3b01',
					class: `ontario-form-group ${error ? 'ontario-input--error' : ''}`,
				},
				this.captionState.getCaption(this.getId(), !!this.internalHintExpander),
				this.internalHintText &&
					h('ontario-hint-text', {
						key: '89d1b50018069434bf7ff24d784b6dc065ea7205',
						hint: this.internalHintText.hint,
						hintContentType: this.internalHintText.hintContentType,
						ref: (el) => (this.hintTextRef = el),
					}),
				h(ErrorMessage, { key: 'dd096e7cfe445817532bc7c0cf681048f417c390', message: this.errorMessage, error: error }),
				h(Input, {
					'key': '6f8e1f78a9c9d3f21335ff5b3dca8f4d0d53e3a7',
					'aria-describedBy': this.hintTextId,
					'className': this.getClass(),
					'id': this.getId(),
					'name': this.name,
					'onInput': (e) => this.handleEvent(e, EventType.Input),
					'onChange': (e) => this.handleEvent(e, EventType.Change),
					'onBlur': (e) => this.handleEvent(e, EventType.Blur),
					'onFocus': (e) => this.handleEvent(e, EventType.Focus),
					'type': this.type,
					'value': this.getValue(),
					'required': !!this.required,
				}),
				this.internalHintExpander &&
					h('ontario-hint-expander', {
						key: 'f25a94ba009fa257660a0dbe6c77d00d28668351',
						hint: this.internalHintExpander.hint,
						content: this.internalHintExpander.content,
						hintContentType: this.internalHintExpander.hintContentType,
					}),
			);
		}
		static get formAssociated() {
			return true;
		}
		static get watchers() {
			return {
				value: ['handleValueChange'],
				name: ['validateName'],
				hintText: ['parseHintText'],
				hintExpander: ['parseHintExpander'],
				caption: ['updateCaptionState'],
				language: ['updateLanguage'],
				errorMessage: ['broadcastInputErrorOccurredEvent'],
			};
		}
		static get style() {
			return ontarioInputCss;
		}
	},
	[
		65,
		'ontario-input',
		{
			caption: [1],
			elementId: [1025, 'element-id'],
			inputWidth: [1025, 'input-width'],
			name: [1],
			hintText: [1, 'hint-text'],
			required: [4],
			type: [1025],
			value: [1025],
			errorMessage: [1025, 'error-message'],
			language: [1025],
			hintExpander: [1, 'hint-expander'],
			enableLiveValidation: [4, 'enable-live-validation'],
			inputValidator: [16, 'input-validator'],
			customOnInput: [16, 'custom-on-input'],
			customOnChange: [16, 'custom-on-change'],
			customOnBlur: [16, 'custom-on-blur'],
			customOnFocus: [16, 'custom-on-focus'],
			requiredValidationMessage: [1, 'required-validation-message'],
			hintTextId: [32],
			internalHintText: [32],
			internalHintExpander: [32],
			captionState: [32],
			hasBeenInteractedWith: [32],
			translations: [32],
		},
		[
			[8, 'setAppLanguage', 'handleSetAppLanguage'],
			[8, 'headerLanguageToggled', 'handleHeaderLanguageToggled'],
			[2, 'blur', 'handleComponentBlur'],
		],
		{
			value: ['handleValueChange'],
			name: ['validateName'],
			hintText: ['parseHintText'],
			hintExpander: ['parseHintExpander'],
			caption: ['updateCaptionState'],
			language: ['updateLanguage'],
			errorMessage: ['broadcastInputErrorOccurredEvent'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = [
		'ontario-input',
		'ontario-hint-expander',
		'ontario-hint-text',
		'ontario-icon-alert-error',
		'ontario-icon-chevron-down',
		'ontario-icon-chevron-up',
	];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-input':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioInput$1);
				}
				break;
			case 'ontario-hint-expander':
				if (!customElements.get(tagName)) {
					defineCustomElement$6();
				}
				break;
			case 'ontario-hint-text':
				if (!customElements.get(tagName)) {
					defineCustomElement$5();
				}
				break;
			case 'ontario-icon-alert-error':
				if (!customElements.get(tagName)) {
					defineCustomElement$4();
				}
				break;
			case 'ontario-icon-chevron-down':
				if (!customElements.get(tagName)) {
					defineCustomElement$3();
				}
				break;
			case 'ontario-icon-chevron-up':
				if (!customElements.get(tagName)) {
					defineCustomElement$2();
				}
				break;
		}
	});
}

const OntarioInput = OntarioInput$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioInput, defineCustomElement };
