// @ongov/ontario-design-system-component-library/dist/components/ontario-blockquote.js
import { p as proxyCustomElement, H, h } from './index.js';
import { a as validatePropExists } from './p-CNqFEifG.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';

const ontarioBlockquoteCss =
	'@charset "UTF-8";.ontario-blockquote{border-left:4px solid #4d4d4d;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:400;letter-spacing:0.025rem;line-height:1.56;max-width:48rem;margin:2rem;padding:1.5rem 0 1.5rem 1.5rem;quotes:"“" "”"}@media screen and (max-width: 40em){.ontario-blockquote{font-size:1.125rem;letter-spacing:0.03rem;line-height:1.5;margin:1.5rem 0 2rem 0}}.ontario-blockquote p{margin:0 0 1.5rem 0}.ontario-blockquote p::before{content:open-quote}.ontario-blockquote p::after{content:close-quote}.ontario-blockquote--short{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.5}@media screen and (max-width: 40em){.ontario-blockquote--short{font-size:1.4375rem;line-height:1.43}}.ontario-blockquote__attribution,.ontario-blockquote__byline{display:block;font-size:1rem;letter-spacing:0.025rem;line-height:1.56;text-align:right}@media screen and (max-width: 40em){.ontario-blockquote__attribution,.ontario-blockquote__byline{letter-spacing:0.03rem;line-height:1.5}}.ontario-blockquote__attribution{font-style:normal;font-weight:700}.ontario-blockquote__attribution::before{content:"— "}html[lang=fr] blockquote{quotes:"« " " »"}';

const OntarioBlockquote$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioBlockquote extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
		}
		get host() {
			return this;
		}
		/**
		 * Text to be displayed as the quote.
		 *
		 * Note that wrapping the quotes in quotations is not needed - this is handled through the component styles
		 */
		quote;
		/**
		 * Optional text to be displayed as the attribution (the author) of the quote.
		 */
		attribution;
		/**
		 * Optional text to be displayed for additional information about the attribution/author.
		 */
		byline;
		shortQuoteLength = 140;
		shortQuote = false;
		quoteState;
		/*
		 * Watch for changes in the `quote` prop for validation purposes.
		 *
		 * If the `quote` prop is not provided, set the `quote` prop to the host textContent (if it exists).
		 * If the `quote` prop length is 140 characters or less, set the `shortQuote` state to true - this will add additonal classes for the blockquote styles.
		 */
		validateQuote() {
			this.quoteState = this.quote ?? this.host.textContent ?? '';
			this.validateQuoteContent(this.quoteState);
			this.shortQuote = this.quoteState?.length <= this.shortQuoteLength ?? true;
		}
		/**
		 * Print the missing `quote` prop warning message
		 */
		validateQuoteContent(newValue) {
			if (validatePropExists(newValue)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' quote ')
					.addRegularText('for')
					.addMonospaceText(' <ontario-blockquote> ')
					.addRegularText('was not provided')
					.printMessage();
			}
		}
		componentWillLoad() {
			this.validateQuote();
		}
		/**
		 * This helper is used to help load translations for any slots + text content passed in by the user.
		 */
		componentDidLoad() {
			const observer = new MutationObserver((mutations) => {
				mutations.forEach((mutation) => {
					if (mutation.type === 'attributes') {
						this.validateQuote();
					}
				});
			});
			const options = { attributes: true };
			observer.observe(this.host, options);
		}
		render() {
			return h(
				'blockquote',
				{
					key: '2ee572aa7d71ef844a7bb1fa544962339b19b5d0',
					class: this.shortQuote ? `ontario-blockquote ontario-blockquote--short` : `ontario-blockquote`,
				},
				h('p', { key: 'edda8b3587f2eab9c8e06ee3fbc4162e980cc4a2' }, this.quoteState),
				this.attribution &&
					h(
						'cite',
						{ key: '968f3739eae9fa5d5ae4a88462b8b23bc911e5c1', class: 'ontario-blockquote__attribution' },
						this.attribution,
					),
				this.byline &&
					h(
						'cite',
						{ key: 'd6a8cd59764e78c64d0163f6a30f18ec215e32e9', class: 'ontario-blockquote__byline' },
						this.byline,
					),
			);
		}
		static get watchers() {
			return {
				quote: ['validateQuote'],
			};
		}
		static get style() {
			return ontarioBlockquoteCss;
		}
	},
	[
		1,
		'ontario-blockquote',
		{
			quote: [1025],
			attribution: [1],
			byline: [1],
			shortQuoteLength: [32],
			shortQuote: [32],
			quoteState: [32],
		},
		undefined,
		{
			quote: ['validateQuote'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-blockquote'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-blockquote':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioBlockquote$1);
				}
				break;
		}
	});
}

const OntarioBlockquote = OntarioBlockquote$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioBlockquote, defineCustomElement };
