// @ongov/ontario-design-system-component-library/dist/components/ontario-search-box.js
import { p as proxyCustomElement, H, c as createEvent, h } from './index.js';
import { I as Input } from './p-BgeHsfjA.js';
import { O as OntarioIconSearch } from './p-CXfntdg5.js';
import { I as InputCaption, h as handleInputEvent, E as EventType } from './p-CJ1sSyqT.js';
import { t as translations } from './p-DgMhpSmi.js';
import { c as constructHintTextObject } from './p-CkusvkRr.js';
import { d as defineCustomElement$2 } from './p-Br8HOVJk.js';

const ontarioSearchBoxCss =
	'@charset "UTF-8";.ontario-search__input.ontario-input:focus,.ontario-input:focus,.ontario-input:active{box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:box-shadow 0.1s ease-in-out}h6,.ontario-h6,h5,.ontario-h5,h4,.ontario-h4,h3,.ontario-h3,h2,.ontario-h2,h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-style:normal;font-weight:700;text-rendering:optimizeLegibility;margin-bottom:1rem;font-feature-settings:normal;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif}h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2rem;letter-spacing:0.04rem;line-height:1.29;margin:2.5rem 0 1.5rem 0;max-width:70rem}@media screen and (min-width: 40em){h1>.ontario-label--heading,.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1,h1,.ontario-h1{font-size:2.5rem;letter-spacing:0.04rem;line-height:1.2}}h2,.ontario-h2{font-size:1.6875rem;letter-spacing:0.03rem;line-height:1.37;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h2,.ontario-h2{font-size:2.0625rem;letter-spacing:0.02rem;line-height:1.33}}h3,.ontario-h3{font-size:1.4375rem;letter-spacing:0.02rem;line-height:1.39;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h3,.ontario-h3{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.43}}h4,.ontario-h4{font-size:1.25rem;letter-spacing:0.03rem;line-height:1.5;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h4,.ontario-h4{font-size:1.5rem;letter-spacing:0.0313rem;line-height:1.5}}h5,.ontario-h5{font-size:1.125rem;letter-spacing:0.03rem;line-height:1.56;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h5,.ontario-h5{font-size:1.1875rem;letter-spacing:0.025rem;line-height:1.5}}h6,.ontario-h6{font-size:1rem;line-height:1.56;letter-spacing:0.03rem;max-width:48rem;margin:0 0 0.75rem 0}@media screen and (min-width: 40em){h6,.ontario-h6{font-size:1rem;letter-spacing:0.025rem;line-height:1.5}}.ontario-lead-statement{font-size:1.25rem;line-height:1.6;max-width:70rem}@media screen and (min-width: 40em){.ontario-lead-statement{font-size:1.375rem}}html{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem}body{font-size:1rem;line-height:1.6;background-color:#FFFFFF;color:#1a1a1a;margin:0}p{font-size:inherit;margin-bottom:1.25rem;text-rendering:optimizeLegibility;margin:0 0 1.5rem 0}cite{font-style:italic;line-height:inherit}strong{font-weight:700;line-height:inherit}*,*::before,*::after{box-sizing:inherit}main p,main p:not(.ontario-lead-statement),main dd,main li{max-width:48rem}hr{border:0;padding:0;margin:2.5rem 0;border-bottom:4px solid #cccccc}hr.ontario-hr--dark{border-bottom:4px solid #1a1a1a}a{color:#0066CC;text-decoration:underline}a:visited{color:#551A8B}a:hover{color:#00478F}a:active{color:#002142}a:focus{-moz-box-shadow:0 0 0 4px #009ADB;-webkit-box-shadow:0 0 0 4px #009ADB;box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:all 0.1s ease-in-out}​p{margin:0 0 1.5rem 0}​p+h1,p+h2,p+h3,p+h4,p+h5,p+h6,ul+h1,ul+h2,ul+h3,ul+h4,ul+h5,ul+h6,ol+h1,ol+h2,ol+h3,ol+h4,ol+h5,ol+h6,h1+h2,h2+h3,h3+h4,h4+h5,h5+h6{margin-top:2rem}p code,td code,li code{line-height:2.25rem}input[type=text],input[type=number],input[type=email],input[type=date],input[type=color],input[type=search],input[type=tel],textarea{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}input[type=checkbox]+label,input[type=radio]+label{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400;margin-left:0;margin-right:0}fieldset{border:0;margin:0;padding:0;page-break-inside:avoid}legend,label{display:block;margin:0}select{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}ul,ol{padding:0}ul{margin-left:2rem}ul li{padding:0 0 0.75rem 0}ol{margin-left:1.75rem}ol li{padding:0 0 0.75rem 0.25rem}ul li:last-child,ol li:last-child{padding-bottom:0}ul[class*=columns-] li:last-child,ol[class*=columns-] li:last-child{padding-bottom:0.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 2rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 1.75rem}p+ul,p+ol{margin-top:-0.75rem}@media screen and (min-width: 40em){ul{margin:0.5rem 0 1.5rem 3rem}ol{margin:0.5rem 0 1.5rem 2.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 3rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 2.75rem}}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-form-group:last-of-type{margin-bottom:3rem}.ontario-fieldset__legend{color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;margin:0 0 1rem 0;line-height:1.5;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-fieldset__legend{font-size:1.125rem;line-height:1.56}}.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{letter-spacing:0.02rem;line-height:1.2}.ontario-fieldset__legend--large h1,.ontario-fieldset__legend--heading h1{display:inline-block;margin:0}@media screen and (max-width: 40em){.ontario-fieldset__legend--large,.ontario-fieldset__legend--heading{font-size:1.4375rem;line-height:1.39}}.ontario-fieldset__legend--large{font-size:1.75rem}.ontario-fieldset__legend-required-flag,.ontario-fieldset__legend-optional-flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-fieldset__legend-required-flag:before,.ontario-fieldset__legend-optional-flag:before{content:" "}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-label{cursor:pointer;color:#1a1a1a;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1.1875rem;font-weight:700;line-height:1.5;margin:0 0 1rem 0;max-width:38.75em;white-space:normal;width:100%;hyphens:auto}@media screen and (max-width: 40em){.ontario-label{font-size:1.125rem;line-height:1.56}}.ontario-label--large{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.2;max-width:26.25em}@media screen and (max-width: 40em){.ontario-label--large{font-size:1.4375rem;line-height:1.39}}h1>.ontario-label--heading{display:inline-block;margin:0}.ontario-label__flag{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400}.ontario-label__flag:before{content:" "}.ontario-input{border:2px solid #1a1a1a;border-radius:4px;box-sizing:border-box;color:#1a1a1a;display:block;font-size:1rem;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;line-height:1.5;margin:0 0 2.5rem;max-width:48rem;width:100%;padding:0.625rem 1rem;transition:box-shadow 0.1s ease-in-out}.ontario-input--2-char-width{max-width:8ex}.ontario-input--3-char-width{max-width:10ex}.ontario-input--4-char-width{max-width:12ex}.ontario-input--5-char-width{max-width:14ex}.ontario-input--7-char-width{max-width:17ex}.ontario-input--10-char-width{max-width:23ex}.ontario-input--20-char-width{max-width:41ex}.ontario-form-group:last-of-type{margin-bottom:2.5rem}.ontario-hide{display:none !important}.ontario-invisible{visibility:hidden !important}@media screen and (max-width: 39.99875em){.ontario-hide-for-small-only{display:none !important}}@media screen and (max-width: 0em), screen and (min-width: 40em){.ontario-show-for-small-only{display:none !important}}@media print, screen and (min-width: 40em){.ontario-hide-for-medium{display:none !important}}@media screen and (max-width: 39.99875em){.ontario-show-for-medium{display:none !important}}@media screen and (min-width: 40em) and (max-width: 72.99875em){.ontario-hide-for-medium-only{display:none !important}}@media screen and (max-width: 39.99875em), screen and (min-width: 73em){.ontario-show-for-medium-only{display:none !important}}@media print, screen and (min-width: 73em){.ontario-hide-for-large{display:none !important}}@media screen and (max-width: 72.99875em){.ontario-show-for-large{display:none !important}}@media screen and (min-width: 73em) and (max-width: 95.99875em){.ontario-hide-for-large-only{display:none !important}}@media screen and (max-width: 72.99875em), screen and (min-width: 96em){.ontario-show-for-large-only{display:none !important}}.ontario-show-for-sr,.ontario-show-on-focus{position:absolute;width:1px;height:1px;overflow:hidden;clip:rect(0, 0, 0, 0)}.ontario-show-on-focus:active,.ontario-show-on-focus:focus{position:static;width:auto;height:auto;overflow:visible;clip:auto}.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}@media screen and (orientation: landscape){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:block !important}}@media screen and (orientation: portrait){.ontario-show-for-landscape,.ontario-hide-for-portrait{display:none !important}}.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}@media screen and (orientation: landscape){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:none !important}}@media screen and (orientation: portrait){.ontario-hide-for-landscape,.ontario-show-for-portrait{display:block !important}}.ontario-search__container{max-width:34rem;margin-bottom:2.5rem}.ontario-search__input-container{position:relative}.ontario-search__input.ontario-input{border:2px solid #1a1a1a;margin-bottom:0;height:3rem;padding-left:0.75rem;padding-right:8.5rem;width:100%}.ontario-search__input.ontario-input:invalid+input[type=reset]{display:none}input[type=reset].ontario-search__reset{position:absolute;display:flex;align-items:center;top:0.75rem;right:6rem;height:1.5rem;width:1.5rem;color:#1a1a1a;margin:0;padding:0.25rem;background-size:1.5rem;background-image:url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgZmlsbD0ibm9uZSI+PHBhdGggZD0iTTE5IDYuNEwxNy42IDUgMTIgMTAuNiA2LjQgNSA1IDYuNGw1LjYgNS42TDUgMTcuNiA2LjQgMTlsNS42LTUuNiA1LjYgNS42IDEuNC0xLjQtNS42LTUuNkwxOSA2LjR6IiBmaWxsPSIjMDAwIi8+PC9zdmc+);background-position:center center;background-repeat:no-repeat;background-color:transparent;outline:none;border:none;cursor:pointer}input[type=reset].ontario-search__reset:focus{box-shadow:inset 0 0 0 4px #009ADB}@media screen and (max-width: 40em){input[type=reset].ontario-search__reset{right:5.5rem}}.ontario-search__submit{border:0;border-radius:0 3px 3px 0;background-color:#0066CC;color:#1a1a1a;display:flex;justify-content:center;align-items:center;font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-weight:600;font-size:1.125rem;width:5rem;line-height:1.56;position:absolute;right:0;bottom:0;top:0;cursor:pointer}.ontario-search__submit:hover{background-color:#00478F;color:#FFFFFF}.ontario-search__submit:focus{outline:none;box-shadow:0 0 0 4px #009ADB;background-color:#00478F;color:#FFFFFF}.ontario-search__submit:active{background-color:#002142;color:#FFFFFF}.ontario-search__submit svg{fill:#FFFFFF;margin-right:0;margin-bottom:0;width:2rem;height:2rem;display:inline-block;vertical-align:middle;overflow:hidden}@media screen and (max-width: 40em){.ontario-search__submit{width:3rem}}input[type=search].ontario-search__input::-webkit-search-decoration,input[type=search].ontario-search__input::-webkit-search-cancel-button,input[type=search].ontario-search__input::-webkit-search-results-button,input[type=search].ontario-search__input::-webkit-search-results-decoration{-webkit-appearance:none}';

const OntarioSearchBox$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioSearchBox extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
			this.searchOnSubmit = createEvent(this, 'searchOnSubmit');
			this.inputOnInput = createEvent(this, 'inputOnInput');
			this.inputOnChange = createEvent(this, 'inputOnChange');
			this.inputOnBlur = createEvent(this, 'inputOnBlur');
			this.inputOnFocus = createEvent(this, 'inputOnFocus');
		}
		get element() {
			return this;
		}
		internals;
		/**
		 * This Ref is used get a direct reference to the hint text element
		 */
		hintTextRef;
		/**
		 * This Ref is used get a direct reference to the search input element
		 */
		inputFieldRef;
		/**
		 * The language of the component.
		 * This is used for translations. If none is passed, it will default to English.
		 */
		language = 'en';
		/**
		 * The unique identifier of the search-box component. This is optional - if no ID is passed, one will be generated.
		 */
		elementId;
		/**
		 * The value of the search term.
		 * This is optional.
		 */
		value;
		/**
		 * The text to display as the input label
		 *
		 * @example
		 * <ontario-search-box
		 *   caption='{
		 *			"captionText": "Search directory",
		 *			"captionType": "default"
		 *		}'
		 *		required = "true"
		 * >
		 * </ontario-search-box>
		 */
		caption;
		/**
		 * This is used to determine whether the dropdown list is required or not.
		 * This prop gets passed to the InputCaption utility to display either an optional or required flag in the label.
		 * If no prop is set, it will default to false (optional).
		 */
		required = false;
		/**
		 * This Function to perform a search operation.
		 * This function will be called when the search submit button is triggered.
		 * The value argument is used for as search term to use for the search operation. This parameter is optional.
		 * The performSearch prop can be set dynamically using JavaScript, allowing you to define custom search functionality when the search form is submitted.
		 *
		 * @example
		 * <ontario-search-box
		 *	  id="ontario-search-box"
		 *	  caption='Search directory'
		 * ></ontario-search-box>
		 *
		 * <script>
		 *	window.addEventListener('load', () => {
		 *		const searchBox = document.getElementById('ontario-search-box');
		 *		searchBox.performSearch = async (value) => {
		 * 			console.log('Performing search with value:', value);
		 *		};
		 *	});
		 * </script>
		 */
		performSearch;
		/**
		 * Used to include the ontario-hint-text component for the search-box.
		 * This is optional.
		 */
		hintText;
		/**
		 * Used to add a custom function to the input onInput event.
		 */
		customOnInput;
		/**
		 * Used to add a custom function to the input onChange event.
		 */
		customOnChange;
		/**
		 * Used to add a custom function to the input onBlur event.
		 */
		customOnBlur;
		/**
		 * Used to add a custom function to the input onFocus event.
		 */
		customOnFocus;
		/**
		 * Emitted when the search is submitted.
		 * Below is an example on how to hook into the event to get the event details.
		 *
		 * @example
		 *	<script>
		 *		document.getElementById('ontario-search-box').addEventListener('searchOnSubmit', (event) => {
		 *	 		const searchValue = event.detail;
		 *			console.log('Search submitted with value:', searchValue);
		 *	  };
		 * 	</script>
		 */
		searchOnSubmit;
		/**
		 * Emitted when a input  occurs when an input has been changed.
		 */
		inputOnInput;
		/**
		 * Emitted when a keyboard input or mouse event occurs when an input has been changed.
		 */
		inputOnChange;
		/**
		 * Emitted when a keyboard input event occurs when an input has lost focus.
		 */
		inputOnBlur;
		/**
		 * Emitted when a keyboard input event occurs when an input has gained focus.
		 */
		inputOnFocus;
		/**
		 * The hint text options are re-assigned to the internalHintText array.
		 */
		internalHintText;
		/**
		 * Instantiate an InputCaption object for internal logic use
		 */
		captionState;
		/**
		 * Used for the `aria-describedby` value of the dropdown list. This will match with the id of the hint text.
		 */
		hintTextId;
		/**
		 * Watch for changes to the `hintText` prop.
		 *
		 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
		 */
		parseHintText() {
			if (this.hintText) {
				const hintTextObject = constructHintTextObject(this.hintText);
				this.internalHintText = hintTextObject;
			}
		}
		/**
		 * Watch for changes to the `caption` prop.
		 *
		 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
		 * @param newValue: Caption | string
		 */
		updateCaptionState(newValue) {
			this.captionState = new InputCaption(
				this.element.tagName,
				newValue,
				translations,
				this.language,
				false,
				this.required,
			);
		}
		/**
		 * Watch for changes to the `language` prop to render either the English or French translations
		 */
		updateLanguage() {
			this.updateCaptionState(this.caption);
		}
		/**
		 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the select `aria-describedBy` attribute.
		 */
		async componentDidLoad() {
			this.hintTextId = await this.hintTextRef?.getHintTextId();
		}
		componentWillLoad() {
			this.elementId = this.elementId;
			this.parseHintText();
			this.updateCaptionState(this.caption);
		}
		/**
		 * Function to handle input events and the information pertaining to the input to emit.
		 */
		handleEvent(event, eventType) {
			const input = event.target;
			// Update the component value to match the value of the input element.
			this.value = input?.value;
			this.internals?.setFormValue?.(this.value ?? '');
			handleInputEvent(
				event,
				eventType,
				input,
				this.inputOnChange,
				this.inputOnFocus,
				this.inputOnBlur,
				this.inputOnInput,
				'input',
				this.customOnChange,
				this.customOnFocus,
				this.customOnBlur,
				this.customOnInput,
				this.element,
			);
		}
		/**
		 * handleSearch function is called when the search submit button is clicked
		 */
		async handleSearch(event) {
			event.preventDefault();
			this.searchOnSubmit.emit(this.value);
			this.performSearch && (await this.performSearch(this.value));
		}
		getId() {
			return this.elementId ?? '';
		}
		getValue() {
			return this.value ?? '';
		}
		/**
		 *This function ensures that the focus returns to the search input field when the reset button is clicked.
		 */
		setFocus(inputRef) {
			if (inputRef) {
				inputRef.focus();
			}
		}
		/**
		 * when the reset button is clicked this function gets called
		 */
		handleFocus = () => {
			this.setFocus(this.inputFieldRef);
		};
		render() {
			const searchInputFieldId = 'ontario-search-input-field';
			return h(
				'form',
				{
					key: 'e3c63d000fc89ce897173eb391ae1f384433039e',
					name: 'searchForm',
					id: 'ontario-search-form-container',
					class:
						'ontario-search__container ontario-columns ontario-small-10 ontario-medium-offset-3 ontario-medium-6 ontario-large-offset-0 ontario-large-6',
					novalidate: true,
				},
				this.captionState.getCaption(searchInputFieldId, false, true) /* Note the _required_ text is disabled */,
				this.internalHintText &&
					h('ontario-hint-text', {
						key: '7db0a418e18506a0d30519127faa32bc6008c4e0',
						hint: this.internalHintText.hint,
						hintContentType: this.internalHintText.hintContentType,
						ref: (el) => (this.hintTextRef = el),
					}),
				h(
					'div',
					{ key: '47611717f5f3ede04d4506230c5bd9ba06072edd', class: 'ontario-search__input-container' },
					h(Input, {
						'key': '0a86eb7189908a9e3822ded6fcba0fac20596f11',
						'aria-describedBy': this.hintTextId,
						'type': 'search',
						'name': 'search',
						'id': searchInputFieldId,
						'autoComplete': 'off',
						'aria-autocomplete': 'none',
						'className': 'ontario-search__input ontario-input',
						'required': true,
						'ref': (el) => (this.inputFieldRef = el),
						'onInput': (e) => this.handleEvent(e, EventType.Input),
						'onChange': (e) => this.handleEvent(e, EventType.Change),
						'onBlur': (e) => this.handleEvent(e, EventType.Blur),
						'onFocus': (e) => this.handleEvent(e, EventType.Focus),
						'value': this.getValue(),
					}),
					h(Input, {
						key: '1523d69dcd1f37861eb5911dbc0a52779502a795',
						className: 'ontario-search__reset',
						id: 'ontario-search-reset',
						type: 'reset',
						value: '',
						onClick: () => this.handleFocus(),
					}),
					h(
						'button',
						{
							key: '3a7ad4366efaef64fd32f77a2ef1b871af15cb48',
							class: 'ontario-search__submit',
							type: 'submit',
							id: 'ontario-search-box__submit',
							onClick: (e) => this.handleSearch(e),
						},
						h('span', { key: '9e8068e22e63dae3f6cafbfcaaccbbc466976121', innerHTML: OntarioIconSearch }),
					),
				),
			);
		}
		static get watchers() {
			return {
				hintText: ['parseHintText'],
				caption: ['updateCaptionState'],
				language: ['updateLanguage'],
			};
		}
		static get style() {
			return ontarioSearchBoxCss;
		}
	},
	[
		1,
		'ontario-search-box',
		{
			language: [1025],
			elementId: [1025, 'element-id'],
			value: [1025],
			caption: [1],
			required: [4],
			performSearch: [16, 'perform-search'],
			hintText: [1, 'hint-text'],
			customOnInput: [16, 'custom-on-input'],
			customOnChange: [16, 'custom-on-change'],
			customOnBlur: [16, 'custom-on-blur'],
			customOnFocus: [16, 'custom-on-focus'],
			internalHintText: [32],
			captionState: [32],
			hintTextId: [32],
		},
		undefined,
		{
			hintText: ['parseHintText'],
			caption: ['updateCaptionState'],
			language: ['updateLanguage'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-search-box', 'ontario-hint-text'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-search-box':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioSearchBox$1);
				}
				break;
			case 'ontario-hint-text':
				if (!customElements.get(tagName)) {
					defineCustomElement$2();
				}
				break;
		}
	});
}

const OntarioSearchBox = OntarioSearchBox$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioSearchBox, defineCustomElement };
