// @ongov/ontario-design-system-component-library/dist/components/ontario-page-alert.js
import { p as proxyCustomElement, H, h } from './index.js';
import { d as defineCustomElement$5 } from './p-BeNeXjVo.js';
import { d as defineCustomElement$4 } from './p-DtoxyT30.js';
import { d as defineCustomElement$3 } from './p-C6R1RJva.js';
import { d as defineCustomElement$2 } from './p-oOEM9Utl.js';

const ontarioPageAlertCss =
	'@charset "UTF-8";h6,.ontario-h6,h5,.ontario-h5,h4,.ontario-h4,h3,.ontario-h3,h2,.ontario-h2,h1,.ontario-h1{font-style:normal;font-weight:700;text-rendering:optimizeLegibility;margin-bottom:1rem;font-feature-settings:normal;font-family:"Raleway", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif}h1,.ontario-h1{font-size:2rem;letter-spacing:0.04rem;line-height:1.29;margin:2.5rem 0 1.5rem 0;max-width:70rem}@media screen and (min-width: 40em){h1,.ontario-h1{font-size:2.5rem;letter-spacing:0.04rem;line-height:1.2}}h2,.ontario-h2{font-size:1.6875rem;letter-spacing:0.03rem;line-height:1.37;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h2,.ontario-h2{font-size:2.0625rem;letter-spacing:0.02rem;line-height:1.33}}h3,.ontario-h3{font-size:1.4375rem;letter-spacing:0.02rem;line-height:1.39;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h3,.ontario-h3{font-size:1.75rem;letter-spacing:0.02rem;line-height:1.43}}h4,.ontario-h4{font-size:1.25rem;letter-spacing:0.03rem;line-height:1.5;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h4,.ontario-h4{font-size:1.5rem;letter-spacing:0.0313rem;line-height:1.5}}h5,.ontario-h5{font-size:1.125rem;letter-spacing:0.03rem;line-height:1.56;margin:0 0 0.75rem 0;max-width:48rem}@media screen and (min-width: 40em){h5,.ontario-h5{font-size:1.1875rem;letter-spacing:0.025rem;line-height:1.5}}h6,.ontario-h6{font-size:1rem;line-height:1.56;letter-spacing:0.03rem;max-width:48rem;margin:0 0 0.75rem 0}@media screen and (min-width: 40em){h6,.ontario-h6{font-size:1rem;letter-spacing:0.025rem;line-height:1.5}}.ontario-lead-statement{font-size:1.25rem;line-height:1.6;max-width:70rem}@media screen and (min-width: 40em){.ontario-lead-statement{font-size:1.375rem}}html{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem}body{font-size:1rem;line-height:1.6;background-color:#FFFFFF;color:#1a1a1a;margin:0}p{font-size:inherit;margin-bottom:1.25rem;text-rendering:optimizeLegibility;margin:0 0 1.5rem 0}cite{font-style:italic;line-height:inherit}strong{font-weight:700;line-height:inherit}*,*::before,*::after{box-sizing:inherit}main p,main p:not(.ontario-lead-statement),main dd,main li{max-width:48rem}hr{border:0;padding:0;margin:2.5rem 0;border-bottom:4px solid #cccccc}hr.ontario-hr--dark{border-bottom:4px solid #1a1a1a}a{color:#0066CC;text-decoration:underline}a:visited{color:#551A8B}a:hover{color:#00478F}a:active{color:#002142}a:focus{-moz-box-shadow:0 0 0 4px #009ADB;-webkit-box-shadow:0 0 0 4px #009ADB;box-shadow:0 0 0 4px #009ADB;outline:4px solid transparent;transition:all 0.1s ease-in-out}​p{margin:0 0 1.5rem 0}​p+h1,p+h2,p+h3,p+h4,p+h5,p+h6,ul+h1,ul+h2,ul+h3,ul+h4,ul+h5,ul+h6,ol+h1,ol+h2,ol+h3,ol+h4,ol+h5,ol+h6,h1+h2,h2+h3,h3+h4,h4+h5,h5+h6{margin-top:2rem}p code,td code,li code{line-height:2.25rem}input[type=text],input[type=number],input[type=email],input[type=date],input[type=color],input[type=search],input[type=tel],textarea{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}input[type=checkbox]+label,input[type=radio]+label{font-family:"Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;font-size:1rem;font-weight:400;margin-left:0;margin-right:0}fieldset{border:0;margin:0;padding:0;page-break-inside:avoid}legend,label{display:block;margin:0}select{background-origin:initial;height:auto;-webkit-appearance:none;-moz-appearance:none}ul,ol{padding:0}ul{margin-left:2rem}ul li{padding:0 0 0.75rem 0}ol{margin-left:1.75rem}ol li{padding:0 0 0.75rem 0.25rem}ul li:last-child,ol li:last-child{padding-bottom:0}ul[class*=columns-] li:last-child,ol[class*=columns-] li:last-child{padding-bottom:0.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 2rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 1.75rem}p+ul,p+ol{margin-top:-0.75rem}@media screen and (min-width: 40em){ul{margin:0.5rem 0 1.5rem 3rem}ol{margin:0.5rem 0 1.5rem 2.75rem}ul ul,ol ul{margin:0.5rem 0 0.5rem 3rem}ul ol,ol ol{margin:0.5rem 0 0.5rem 2.75rem}}.ontario-alert{border-width:0 0 0 0.25rem;border-style:solid;padding:1.5rem;margin:2rem 0 2.5rem 0}.ontario-alert__header{display:-webkit-box;display:-ms-flexbox;display:flex}.ontario-alert__header-icon{display:inline-block;vertical-align:top;margin-right:0.95rem}.ontario-alert__header-title{margin-bottom:0.75rem;padding-top:0.125rem}.ontario-alert__body{margin-left:3.25rem;max-width:48rem}@media screen and (max-width: 40em){.ontario-alert__body{margin-left:0}}.ontario-alert__body p{margin-top:0}.ontario-alert__body p:last-of-type{margin-bottom:0.5rem}.ontario-alert__body p:last-of-type+ul{margin-top:1rem}.ontario-alert__body ul:last-of-type{margin-bottom:0.5rem}.ontario-alert--error ul{margin-left:1.25rem}.ontario-alert--error ul:last-of-type{margin-bottom:0.5rem}.ontario-alert--informational{border-color:#1080A6;background-color:#E2F0F4}.ontario-alert--warning{border-color:#FFD440;background-color:#FEF6DC}.ontario-alert--success{border-color:#118847;background-color:#E5f0E9}.ontario-alert--error{border-color:#CD0000;background-color:#FCEFF0}.ontario-alert--error a{color:#CD0000}.ontario-alert--error a:hover,.ontario-alert--error a:focus{color:#6b0000}.ontario-alert--error a:active{color:#280000}';

const OntarioPageAlert$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioPageAlert extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
		}
		/**
		 * The type of page alert to render. If no value is provided, the `informational` type alert would be rendered.
		 *
		 * There are four possible values for page alert: `informational`, `warning`, `success` or `error`.
		 *
		 * @example
		 * <ontario-page-alert type="error">
		 * </ontario-page-alert>
		 */
		type = 'informational';
		/**
		 * The heading for the page alert.
		 *
		 * @example
		 * <ontario-page-alert heading="Licence plates">
		 * </ontario-page-alert>
		 */
		heading;
		/**
		 * The main content for the page alert. This can be rendered as either string or HTML content.
		 *
		 * @example
		 * <ontario-page-alert content="Please look out for an email confirmation with your receipt and order number.">
		 * </ontario-page-alert>
		 *
		 * or
		 *
		 * <ontario-page-alert>
		 *  <p>This is a sample page alert component using slots. <a href="#">Learn more</a>.</p>
		 * </ontario-page-alert>
		 */
		content;
		/**
		 * @returns the classes of the page alert container based on the alert `type`.
		 */
		getClass() {
			return `ontario-alert ontario-alert--${this.type}`;
		}
		renderIcon() {
			const iconProps = { 'icon-width': 36 };
			switch (this.type) {
				case 'informational':
					return h('ontario-icon-alert-information', { ...iconProps });
				case 'success':
					return h('ontario-icon-alert-success', { ...iconProps });
				case 'warning':
					return h('ontario-icon-alert-warning', { ...iconProps });
				case 'error':
					return h('ontario-icon-alert-error', { ...iconProps });
			}
		}
		renderContent() {
			const body = this.content;
			if (typeof body === 'string') {
				return h('p', null, body);
			}
			return h('slot', null);
		}
		render() {
			return h(
				'div',
				{ key: 'bc89ba1ae733df97405d1431c21e94b6bde31659', class: this.getClass() },
				h(
					'div',
					{ key: '823d43473763c86b7d62e50c6a9f38377eb85931', class: 'ontario-alert__header' },
					h(
						'div',
						{ key: '186731136cf022db116adcb017c1d53be3589312', class: 'ontario-alert__header-icon' },
						this.renderIcon(),
					),
					h(
						'h2',
						{ key: '01f2f4f91f6567cff09130359ac2cd6fb949574e', class: 'ontario-alert__header-title ontario-h4' },
						this.heading,
					),
				),
				h(
					'div',
					{ key: 'c72fdccf42f2669259856c58be84085152dde302', class: 'ontario-alert__body' },
					this.renderContent(),
				),
			);
		}
		static get style() {
			return ontarioPageAlertCss;
		}
	},
	[
		1,
		'ontario-page-alert',
		{
			type: [1],
			heading: [1],
			content: [1],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = [
		'ontario-page-alert',
		'ontario-icon-alert-error',
		'ontario-icon-alert-information',
		'ontario-icon-alert-success',
		'ontario-icon-alert-warning',
	];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-page-alert':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioPageAlert$1);
				}
				break;
			case 'ontario-icon-alert-error':
				if (!customElements.get(tagName)) {
					defineCustomElement$5();
				}
				break;
			case 'ontario-icon-alert-information':
				if (!customElements.get(tagName)) {
					defineCustomElement$4();
				}
				break;
			case 'ontario-icon-alert-success':
				if (!customElements.get(tagName)) {
					defineCustomElement$3();
				}
				break;
			case 'ontario-icon-alert-warning':
				if (!customElements.get(tagName)) {
					defineCustomElement$2();
				}
				break;
		}
	});
}

const OntarioPageAlert = OntarioPageAlert$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioPageAlert, defineCustomElement };
