// @ongov/ontario-design-system-component-library/dist/components/ontario-card-collection.js
import { p as proxyCustomElement, H, h } from './index.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';

const ontarioCardCollectionCss =
	'.ontario-card-collection__container{display:flex;flex-wrap:wrap;column-gap:2rem;row-gap:0;margin:0;padding:0}.ontario-card-collecton--cards-per-row-4 ::slotted(ontario-card){width:calc(25% - 1.5rem)}@media screen and (max-width: 73em){.ontario-card-collecton--cards-per-row-4 ::slotted(ontario-card){width:calc(50% - 1rem)}}@media screen and (max-width: 40em){.ontario-card-collecton--cards-per-row-4 ::slotted(ontario-card){width:100%}}.ontario-card-collecton--cards-per-row-3 ::slotted(ontario-card){width:calc(33.3% - 1.33rem)}@media screen and (max-width: 40em){.ontario-card-collecton--cards-per-row-3 ::slotted(ontario-card){width:100%}}.ontario-card-collecton--cards-per-row-2 ::slotted(ontario-card){width:calc(50% - 1rem)}@media screen and (max-width: 40em){.ontario-card-collecton--cards-per-row-2 ::slotted(ontario-card){width:100%}}';

const OntarioCardCollection$1 = /*@__PURE__*/ proxyCustomElement(
	class OntarioCardCollection extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
		}
		get host() {
			return this;
		}
		/**
		 * The number of cards to display per row.
		 *
		 * If no number is passed, it will default to 3.
		 */
		cardsPerRow = 3;
		/**
		 * Mutable variable, for internal use only.
		 *
		 * Set number of cards per row depending on validation result.
		 */
		cardsPerRowState;
		/**
		 * Watch for changes to the `cardsPerRow` property for validation purposes.
		 *
		 * If the user input is not a number or is a negative number then `cardsPerRow` will be set to its default (3).
		 */
		validateCardsPerRow() {
			if (isNaN(this.cardsPerRow) || (!isNaN(this.cardsPerRow) && this.cardsPerRow <= 0)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' cards-per-row ')
					.addRegularText('on')
					.addMonospaceText(' <ontario-card> ')
					.addRegularText(
						`${isNaN(this.cardsPerRow) ? 'was set to a non-numeric value' : 'was set to a negative number'}; only a positive number is allowed. The default number of cards per row: `,
					)
					.addMonospaceText(' 3 ')
					.addRegularText('was assumed.')
					.printMessage();
				this.cardsPerRowState = 3;
			} else {
				this.cardsPerRowState = this.cardsPerRow;
			}
		}
		/**
		 * @returns the classes of the ontario cards based off the `cardType` and number of cards per row.
		 */
		getClass() {
			return `
      ontario-card-collection__container
      ontario-card-collecton--cards-per-row-${this.cardsPerRowState}
    `.trim();
		}
		componentWillLoad() {
			this.validateCardsPerRow();
		}
		render() {
			return h(
				'ul',
				{ key: '8991908dd16580e556964b37009bb25ad547232a', class: this.getClass() },
				h('slot', { key: 'f75f012338edd1a94ed1c5b3d197a9b909b36250' }),
			);
		}
		static get watchers() {
			return {
				cardsPerRow: ['validateCardsPerRow'],
			};
		}
		static get style() {
			return ontarioCardCollectionCss;
		}
	},
	[
		1,
		'ontario-card-collection',
		{
			cardsPerRow: [2, 'cards-per-row'],
			cardsPerRowState: [32],
		},
		undefined,
		{
			cardsPerRow: ['validateCardsPerRow'],
		},
	],
);
function defineCustomElement$1() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-card-collection'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-card-collection':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioCardCollection$1);
				}
				break;
		}
	});
}

const OntarioCardCollection = OntarioCardCollection$1;
const defineCustomElement = defineCustomElement$1;

export { OntarioCardCollection, defineCustomElement };
