// @ongov/ontario-design-system-component-library/dist/components/p-BeNeXjVo.js
import { p as proxyCustomElement, H, h } from './index.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';

const ontarioIconCss =
	'.ontario-icon svg,.ontario-icon{display:inline-block;stroke-width:0;stroke:currentColor;fill:currentColor;vertical-align:middle;overflow:hidden}.ontario-icon--blue svg{stroke:#0066CC;fill:#0066CC}.ontario-icon--grey svg{stroke:#666666;fill:#666666}.ontario-icon--black svg{stroke:#1A1A1A;fill:#1A1A1A}.ontario-icon--white svg{stroke:#FFFFFF;fill:#FFFFFF}.ontario-icon--white svg{stroke:#FFFFFF;fill:#FFFFFF}';

const OntarioIconAlertError = /*@__PURE__*/ proxyCustomElement(
	class OntarioIconAlertError extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
		}
		/**
		 * The icon width will autogenerate the height since the icons are in square format, thus preserving
		 * the aspect ratio.
		 */
		iconWidth = 24;
		/**
		 * Mutable variable, for internal use only.
		 * Set the icon's width depending on validation result.
		 */
		iconWidthState;
		/**
		 * Watch for changes in the `iconWidth` variable for validation purpose.
		 * If the user input is not a number or is a negative number then `iconWidth` will be set to its default (24).
		 */
		validateWidth() {
			if (isNaN(this.iconWidth) || (!isNaN(this.iconWidth) && this.iconWidth <= 0)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' icon-width ')
					.addRegularText('on')
					.addMonospaceText(' <ontario-icon-alert-error> ')
					.addRegularText(
						`${isNaN(this.iconWidth) ? 'was set to a non-numeric value' : 'was set to a negative number'}; only a positive number is allowed. The default size of`,
					)
					.addMonospaceText(' 24px ')
					.addRegularText('was assumed.')
					.printMessage();
				this.iconWidthState = 24;
			} else {
				this.iconWidthState = this.iconWidth;
			}
		}
		get host() {
			return this;
		}
		/**
		 * Validate that the `colour` attribute is not set by users
		 * Prints a warning message if the `colour` attribute is set
		 */
		validateColour() {
			if (this.host.hasAttribute('colour')) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' colour ')
					.addRegularText('on')
					.addMonospaceText(' <ontario-icon-alert-error> ')
					.addRegularText('cannot be set. The provided colour is ignored.')
					.printMessage();
			}
		}
		/**
		 * Stencil component lifecycle method that is called once after the component is first connected to the DOM.
		 */
		componentWillLoad() {
			this.validateColour();
			this.validateWidth();
		}
		/**
		 * Returns the HTML code to be rendered into a custom element.
		 */
		render() {
			return h(
				'div',
				{
					key: '080d9965b36782f0044f4c6b0728fd820df73830',
					class: `ontario-icon ontario-icon--width-${this.iconWidthState}`,
					style: { width: `${this.iconWidthState}px` },
				},
				h(
					'svg',
					{
						key: '0c066c0777fdf1adea3fa6b49a42e5b29ec72a40',
						class: 'svg-icon',
						role: 'img',
						xmlns: 'http://www.w3.org/2000/svg',
						fill: 'none',
						viewBox: '0 0 24 24',
						id: 'alert-error',
					},
					h('path', {
						key: 'b5c90bab331d620b89732891e89b36cca164b107',
						d: 'M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z',
						fill: '#cd0000',
					}),
					h('path', {
						key: '8e73e0ca55b20a1637e73ff3bb2ad6869c77ba8b',
						d: 'M11 17h2v-2h-2v2zm0-4h2V7h-2v6z',
						fill: '#fff',
					}),
				),
			);
		}
		static get watchers() {
			return {
				iconWidth: ['validateWidth'],
			};
		}
		static get style() {
			return ontarioIconCss;
		}
	},
	[
		1,
		'ontario-icon-alert-error',
		{
			iconWidth: [2, 'icon-width'],
			iconWidthState: [32],
		},
		undefined,
		{
			iconWidth: ['validateWidth'],
		},
	],
);
function defineCustomElement() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-icon-alert-error'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-icon-alert-error':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioIconAlertError);
				}
				break;
		}
	});
}

export { OntarioIconAlertError as O, defineCustomElement as d };
