// @ongov/ontario-design-system-component-library/dist/components/p-oOEM9Utl.js
import { p as proxyCustomElement, H, h } from './index.js';
import { C as ConsoleMessageClass } from './p-CJXRCL8b.js';

const ontarioIconCss =
	'.ontario-icon svg,.ontario-icon{display:inline-block;stroke-width:0;stroke:currentColor;fill:currentColor;vertical-align:middle;overflow:hidden}.ontario-icon--blue svg{stroke:#0066CC;fill:#0066CC}.ontario-icon--grey svg{stroke:#666666;fill:#666666}.ontario-icon--black svg{stroke:#1A1A1A;fill:#1A1A1A}.ontario-icon--white svg{stroke:#FFFFFF;fill:#FFFFFF}.ontario-icon--white svg{stroke:#FFFFFF;fill:#FFFFFF}';

const OntarioIconAlertWarning = /*@__PURE__*/ proxyCustomElement(
	class OntarioIconAlertWarning extends H {
		constructor() {
			super();
			this.__registerHost();
			this.__attachShadow();
		}
		/**
		 * The icon width will autogenerate the height since the icons are in square format, thus preserving
		 * the aspect ratio.
		 */
		iconWidth = 24;
		/**
		 * Mutable variable, for internal use only.
		 * Set the icon's width depending on validation result.
		 */
		iconWidthState;
		/**
		 * Watch for changes in the `iconWidth` variable for validation purpose.
		 * If the user input is not a number or is a negative number then `iconWidth` will be set to its default (24).
		 */
		validateWidth() {
			if (isNaN(this.iconWidth) || (!isNaN(this.iconWidth) && this.iconWidth <= 0)) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' icon-width ')
					.addRegularText('on')
					.addMonospaceText(' <ontario-icon-alert-warning> ')
					.addRegularText(
						`${isNaN(this.iconWidth) ? 'was set to a non-numeric value' : 'was set to a negative number'}; only a positive number is allowed. The default size of`,
					)
					.addMonospaceText(' 24px ')
					.addRegularText('was assumed.')
					.printMessage();
				this.iconWidthState = 24;
			} else {
				this.iconWidthState = this.iconWidth;
			}
		}
		get host() {
			return this;
		}
		/**
		 * Validate that the `colour` attribute is not set by users
		 * Prints a warning message if the `colour` attribute is set
		 */
		validateColour() {
			if (this.host.hasAttribute('colour')) {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(' colour ')
					.addRegularText('on')
					.addMonospaceText(' <ontario-icon-alert-warning> ')
					.addRegularText('cannot be set. The provided colour is ignored.')
					.printMessage();
			}
		}
		/**
		 * Stencil component lifecycle method that is called once after the component is first connected to the DOM.
		 */
		componentWillLoad() {
			this.validateColour();
			this.validateWidth();
		}
		/**
		 * Returns the HTML code to be rendered into a custom element.
		 */
		render() {
			return h(
				'div',
				{
					key: '7fc27406ecbba9ea1c361e68dd2bdf195cd0f482',
					class: `ontario-icon ontario-icon--width-${this.iconWidthState}`,
					style: { width: `${this.iconWidthState}px` },
				},
				h(
					'svg',
					{
						key: '134ba2924fd2969dbc70ba0f99ec644f548de60f',
						class: 'svg-icon',
						role: 'img',
						xmlns: 'http://www.w3.org/2000/svg',
						fill: 'none',
						viewBox: '0 0 24 24',
						id: 'alert-warning',
					},
					h('path', {
						key: '0aa5ae9fccf4db8d83e00eb75479ac1380829dce',
						d: 'M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z',
						fill: '#ffd440',
					}),
					h('path', { key: '307169b7a8f7ad94c6c7c838cddccd2d3df2dd90', d: 'M11 10h2v4h-2zm0 6h2v2h-2z', fill: '#000' }),
				),
			);
		}
		static get watchers() {
			return {
				iconWidth: ['validateWidth'],
			};
		}
		static get style() {
			return ontarioIconCss;
		}
	},
	[
		1,
		'ontario-icon-alert-warning',
		{
			iconWidth: [2, 'icon-width'],
			iconWidthState: [32],
		},
		undefined,
		{
			iconWidth: ['validateWidth'],
		},
	],
);
function defineCustomElement() {
	if (typeof customElements === 'undefined') {
		return;
	}
	const components = ['ontario-icon-alert-warning'];
	components.forEach((tagName) => {
		switch (tagName) {
			case 'ontario-icon-alert-warning':
				if (!customElements.get(tagName)) {
					customElements.define(tagName, OntarioIconAlertWarning);
				}
				break;
		}
	});
}

export { OntarioIconAlertWarning as O, defineCustomElement as d };
