// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-page-alert/ontario-page-alert.js
import { h } from '@stencil/core';
export class OntarioPageAlert {
	/**
	 * The type of page alert to render. If no value is provided, the `informational` type alert would be rendered.
	 *
	 * There are four possible values for page alert: `informational`, `warning`, `success` or `error`.
	 *
	 * @example
	 * <ontario-page-alert type="error">
	 * </ontario-page-alert>
	 */
	type = 'informational';
	/**
	 * The heading for the page alert.
	 *
	 * @example
	 * <ontario-page-alert heading="Licence plates">
	 * </ontario-page-alert>
	 */
	heading;
	/**
	 * The main content for the page alert. This can be rendered as either string or HTML content.
	 *
	 * @example
	 * <ontario-page-alert content="Please look out for an email confirmation with your receipt and order number.">
	 * </ontario-page-alert>
	 *
	 * or
	 *
	 * <ontario-page-alert>
	 *  <p>This is a sample page alert component using slots. <a href="#">Learn more</a>.</p>
	 * </ontario-page-alert>
	 */
	content;
	/**
	 * @returns the classes of the page alert container based on the alert `type`.
	 */
	getClass() {
		return `ontario-alert ontario-alert--${this.type}`;
	}
	renderIcon() {
		const iconProps = { 'icon-width': 36 };
		switch (this.type) {
			case 'informational':
				return h('ontario-icon-alert-information', { ...iconProps });
			case 'success':
				return h('ontario-icon-alert-success', { ...iconProps });
			case 'warning':
				return h('ontario-icon-alert-warning', { ...iconProps });
			case 'error':
				return h('ontario-icon-alert-error', { ...iconProps });
		}
	}
	renderContent() {
		const body = this.content;
		if (typeof body === 'string') {
			return h('p', null, body);
		}
		return h('slot', null);
	}
	render() {
		return h(
			'div',
			{ key: 'bc89ba1ae733df97405d1431c21e94b6bde31659', class: this.getClass() },
			h(
				'div',
				{ key: '823d43473763c86b7d62e50c6a9f38377eb85931', class: 'ontario-alert__header' },
				h(
					'div',
					{ key: '186731136cf022db116adcb017c1d53be3589312', class: 'ontario-alert__header-icon' },
					this.renderIcon(),
				),
				h(
					'h2',
					{ key: '01f2f4f91f6567cff09130359ac2cd6fb949574e', class: 'ontario-alert__header-title ontario-h4' },
					this.heading,
				),
			),
			h('div', { key: 'c72fdccf42f2669259856c58be84085152dde302', class: 'ontario-alert__body' }, this.renderContent()),
		);
	}
	static get is() {
		return 'ontario-page-alert';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-page-alert.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-page-alert.css'],
		};
	}
	static get properties() {
		return {
			type: {
				type: 'string',
				attribute: 'type',
				mutable: false,
				complexType: {
					original: 'PageAlertType',
					resolved: '"error" | "informational" | "success" | "warning"',
					references: {
						PageAlertType: {
							location: 'import',
							path: './ontario-page-alert.interface',
							id: 'src/components/ontario-page-alert/ontario-page-alert.interface.ts::PageAlertType',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-page-alert type="error">\n</ontario-page-alert>',
						},
					],
					text: 'The type of page alert to render. If no value is provided, the `informational` type alert would be rendered.\n\nThere are four possible values for page alert: `informational`, `warning`, `success` or `error`.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'informational'",
			},
			heading: {
				type: 'string',
				attribute: 'heading',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-page-alert heading="Licence plates">\n</ontario-page-alert>',
						},
					],
					text: 'The heading for the page alert.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			content: {
				type: 'string',
				attribute: 'content',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-page-alert content="Please look out for an email confirmation with your receipt and order number.">\n</ontario-page-alert>\n\nor\n\n<ontario-page-alert>\n <p>This is a sample page alert component using slots. <a href="#">Learn more</a>.</p>\n</ontario-page-alert>',
						},
					],
					text: 'The main content for the page alert. This can be rendered as either string or HTML content.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
}
