// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-checkbox/ontario-checkboxes.js
import { h } from '@stencil/core';
import { Input } from '../../utils/common/input/input';
import { InputCaption } from '../../utils/common/input-caption/input-caption';
import {
	validateObjectExists,
	validatePropExists,
	validateLanguage,
} from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { constructHintTextObject } from '../../utils/components/hints/hints';
import { EventType } from '../../utils/events/event-handler.interface';
import { handleInputEvent } from '../../utils/events/event-handler';
import { default as translations } from '../../translations/global.i18n.json';
import { ErrorMessage } from '../../utils/components/error-message/error-message';
export class OntarioCheckboxes {
	element;
	internals;
	hintTextRef;
	/**
	 * The text to display for the checkbox legend.
	 *
	 * @example
	 * <ontario-checkboxes
	 *   caption='{
	 *     "captionText": "Checkbox legend",
	 *     "captionType": "heading",
	 *   }
	 *   ...>
	 * </ontario-checkboxes>
	 */
	caption;
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
	 */
	language;
	/**
	 * The name for the checkboxes. The name value is used to reference form data after a form is submitted.
	 */
	name;
	/**
	 * Used to include the ontario-hint-text component for the checkbox group.
	 * This is optional.
	 */
	hintText;
	/**
	 * Used to include the ontario-hint-expander component for the checkbox group.
	 * This is passed in as an object with key-value pairs.
	 *
	 * This is optional.
	 *
	 * @example
	 * <ontario-checkboxes
	 *   caption='{
	 *     "captionText": "Checkbox legend",
	 *     "captionType": "heading",
	 *   }
	 *   name='ontario-checkboxes'
	 *   options='[
	 *		{
	 *			"value": "checkbox-option-1",
	 *			"label": "Checkbox option 1 label",
	 *			"elementId": "checkbox-1"
	 *		}
	 *   }]'
	 *   hint-expander='{
	 *    "hint": "Hint expander for the checkbox group",
	 *    "content": "Example hint expander content for the checkbox group"
	 *   }'
	 *   required="true"
	 * >
	 * </ontario-checkboxes>
	 */
	hintExpander;
	/**
	 * The options for the checkbox group.
	 *
	 * Each property will be passed in through an object in the options array.
	 * This can either be passed in as an object directly (if using react), or as a string in HTML.
	 * If there are multiple checkboxes in a fieldset, each checkbox will be displayed as an option.
	 *
	 * In the example below, the options are being passed in as a string and there are two checkboxes to be displayed in the fieldset.
	 *
	 * @example
	 * <ontario-checkboxes
	 *   caption='{
	 *		"captionText": "Checkbox legend",
	 *		"captionType": "heading",
	 *   }
	 *   name="ontario-checkboxes",
	 *   hint-text="Hint text for the checkbox group."
	 *   options='[
	 *		{
	 *			"value": "checkbox-option-1",
	 *			"label": "Checkbox option 1 label"
	 *			"elementId": "checkbox-1"
	 *     },
	 *     {
	 *        "value": "checkbox-option-2",
	 *        "label": "Checkbox option 2 label",
	 * 		  "elementId": "checkbox-2",
	 *	      "hintExpander": {
	 *				"hint": "Hint expander for checkbox option 2",
	 *              "content": "Example hint expander content for checkbox option 2"
	 *        }
	 *      }
	 *   ]'
	 *   required="true"
	 * >
	 * </ontario-checkboxes>
	 */
	options;
	/**
	 * This is used to determine whether the checkbox is required or not.
	 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
	 * If no prop is set, it will default to false (optional).
	 */
	required = false;
	/**
	 * Set this to display an error message
	 */
	errorMessage;
	/**
	 * Used to add a custom function to the checkbox onChange event.
	 */
	customOnChange;
	/**
	 * Used to add a custom function to the checkbox onBlur event.
	 */
	customOnBlur;
	/**
	 * Used to add a custom function to the checkbox onFocus event.
	 */
	customOnFocus;
	/**
	 * Used for the `aria-describedby` value of the checkbox fieldset. This will match with the id of the hint text.
	 */
	hintTextId;
	/**
	 * Instantiate an InputCaption object for internal logic use
	 */
	captionState;
	/**
	 * The hint text options are re-assigned to the internalHintText array.
	 */
	internalHintText;
	/**
	 * The hint expander options are re-assigned to the internalHintExpander array.
	 */
	internalHintExpander;
	/**
	 * The options are re-assigned to the internalOptions array.
	 */
	internalOptions;
	/**
	 * Emitted when a keyboard input or mouse event occurs when a checkbox option has been changed.
	 */
	checkboxOnChange;
	/**
	 * Emitted when a keyboard input event occurs when a checkbox option has lost focus.
	 */
	checkboxOnBlur;
	/**
	 * Emitted when a keyboard input event occurs when a checkbox option has gained focus.
	 */
	checkboxOnFocus;
	/**
	 * Emitted when an error message is reported to the component.
	 */
	inputErrorOccurred;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 */
	handleSetAppLanguage(event) {
		if (!this.language) {
			this.language = validateLanguage(event);
		}
	}
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event);
	}
	/**
	 * Watch for changes to the `hintText` prop.
	 *
	 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
	 */
	parseHintText() {
		if (this.hintText) {
			const hintTextObject = constructHintTextObject(this.hintText);
			this.internalHintText = hintTextObject;
		}
	}
	/**
	 * Watch for changes to the `hintExpander` prop.
	 *
	 * If a `hintExpander` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalHintExpander` state.
	 */
	parseHintExpander() {
		const hintExpander = this.hintExpander;
		if (hintExpander) {
			if (typeof hintExpander === 'string') this.internalHintExpander = JSON.parse(hintExpander);
			else this.internalHintExpander = hintExpander;
		}
	}
	/**
	 * Watch for changes to the `options` prop.
	 *
	 * If an `options` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalOptions` state. The result will be run through a validation function.
	 */
	parseOptions() {
		if (typeof this.options !== 'undefined') {
			if (!Array.isArray(this.options)) {
				this.internalOptions = JSON.parse(this.options);
			} else {
				this.internalOptions = this.options;
			}
		}
		this.validateOptions(this.internalOptions);
	}
	/**
	 * Validate the `options` and make sure the `options` prop has a value.
	 * Log a warning if user doesn't input a value for the `options`.
	 *
	 * @param newValue object to be validated
	 */
	validateOptions(newValue) {
		if (validateObjectExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' options ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-checkboxes> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/*
	 * Watch for changes in the `name` prop for validation purposes.
	 *
	 * Validate the `name` and make sure the `name` prop has a value.
	 * Log a warning if user doesn't input a value for the `name`.
	 */
	validateName(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' name ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-checkboxes> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/**
	 * Watch for changes to the `caption` prop.
	 *
	 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
	 * @param newValue: Caption | string
	 */
	updateCaptionState(newValue) {
		this.captionState = new InputCaption(
			this.element.tagName,
			newValue,
			translations,
			this.language,
			true,
			this.required,
		);
	}
	/**
	 * Watch for changes to the `language` prop to render either the English or French translations
	 */
	updateLanguage() {
		this.updateCaptionState(this.caption);
	}
	broadcastInputErrorOccurredEvent() {
		// Emit event to notify anyone who wants to listen for errors occurring
		this.inputErrorOccurred.emit({ errorMessage: this.errorMessage ?? '' });
	}
	/**
	 * Function to handle checkbox events and the information pertaining to the checkbox to emit.
	 */
	handleEvent(event, eventType) {
		const input = event.target;
		if (input) {
			input.checked = input.checked ?? '';
		}
		// Update internalOptions checked state
		const changedOption = this.internalOptions.find((x) => x.value === input?.value);
		if (changedOption) changedOption.checked = !changedOption?.checked;
		// Set the value within the form
		this.internals?.setFormValue?.(
			this.internalOptions
				.filter((x) => !!x.checked)
				.reduce((formData, currentValue) => {
					formData.append(this.name, currentValue.value);
					return formData;
				}, new FormData()),
		);
		handleInputEvent(
			event,
			eventType,
			input,
			this.checkboxOnChange,
			this.checkboxOnFocus,
			this.checkboxOnBlur,
			undefined,
			'checkbox',
			this.customOnChange,
			this.customOnFocus,
			this.customOnBlur,
			undefined,
			this.element,
		);
	}
	/**
	 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the fieldset `aria-describedBy` attribute.
	 */
	async componentDidLoad() {
		this.hintTextId = await this.hintTextRef?.getHintTextId();
	}
	componentWillLoad() {
		this.updateCaptionState(this.caption);
		this.parseOptions();
		this.parseHintExpander();
		this.parseHintText();
		this.validateName(this.name);
		this.language = validateLanguage(this.language);
	}
	render() {
		const error = !!this.errorMessage;
		return h(
			'div',
			{
				key: '5436155c32e692dee15fbf93fc4a80cec4ef6325',
				class: `ontario-form-group ${error ? 'ontario-input--error' : ''}`,
			},
			h(
				'fieldset',
				{
					'key': '13830e1c232fd935c8085a0a443b1a0c6f7655a9',
					'class': 'ontario-fieldset',
					'aria-describedby': this.hintTextId,
				},
				this.captionState.getCaption(undefined, !!this.internalHintExpander),
				this.internalHintText &&
					h('ontario-hint-text', {
						key: 'd176e8693779d38e3508b89c24320b9e4f9c0b6c',
						hint: this.internalHintText.hint,
						hintContentType: this.internalHintText.hintContentType,
						ref: (el) => (this.hintTextRef = el),
					}),
				h(ErrorMessage, { key: 'c775a4a1ec00311fcb831473e9b6a51add3cb569', message: this.errorMessage, error: error }),
				h(
					'div',
					{ key: '491cd03960d0e69d1ad5fc8c2827409af2721609', class: 'ontario-checkboxes' },
					this.internalOptions?.map((checkbox) =>
						h(
							'div',
							{ class: 'ontario-checkboxes__item' },
							h(Input, {
								className: 'ontario-checkboxes__input',
								id: checkbox.elementId,
								name: this.name,
								type: 'checkbox',
								value: checkbox.value,
								required: !!this.required,
								checked: !!checkbox.checked,
								onChange: (e) => this.handleEvent(e, EventType.Change),
								onBlur: (e) => this.handleEvent(e, EventType.Blur),
								onFocus: (e) => this.handleEvent(e, EventType.Focus),
							}),
							h(
								'label',
								{ class: 'ontario-checkboxes__label', htmlFor: checkbox.elementId },
								checkbox.label,
								checkbox.hintExpander && this.captionState.getHintExpanderAccessibilityText(checkbox.label, true),
							),
							checkbox.hintExpander &&
								h(
									'div',
									{ class: 'ontario-checkboxes__hint-expander' },
									h('ontario-hint-expander', {
										'hint': checkbox.hintExpander.hint,
										'content': checkbox.hintExpander.content,
										'hintContentType': checkbox.hintExpander.hintContentType,
										'input-exists': true,
									}),
								),
						),
					),
					this.internalHintExpander &&
						h(
							'div',
							{ key: '982eb7debd2038a33166c0e3b4db9887c632a2f2', class: 'ontario-checkboxes__hint-expander' },
							h('ontario-hint-expander', {
								'key': '624fa6eb7ec9dd1840651f4a536792655b8aa0fc',
								'hint': this.internalHintExpander.hint,
								'content': this.internalHintExpander.content,
								'hintContentType': this.internalHintExpander.hintContentType,
								'input-exists': true,
							}),
						),
				),
			),
		);
	}
	static get is() {
		return 'ontario-checkboxes';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get formAssociated() {
		return true;
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-checkboxes.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-checkboxes.css'],
		};
	}
	static get properties() {
		return {
			caption: {
				type: 'string',
				attribute: 'caption',
				mutable: false,
				complexType: {
					original: 'Caption | string',
					resolved: 'Caption | string',
					references: {
						Caption: {
							location: 'import',
							path: '../../utils/common/input-caption/caption.interface',
							id: 'src/utils/common/input-caption/caption.interface.ts::Caption',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-checkboxes\n  caption=\'{\n    "captionText": "Checkbox legend",\n    "captionType": "heading",\n  }\n  ...>\n</ontario-checkboxes>',
						},
					],
					text: 'The text to display for the checkbox legend.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr" | undefined',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			name: {
				type: 'string',
				attribute: 'name',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The name for the checkboxes. The name value is used to reference form data after a form is submitted.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			hintText: {
				type: 'string',
				attribute: 'hint-text',
				mutable: true,
				complexType: {
					original: 'string | Hint',
					resolved: 'Hint | string | undefined',
					references: {
						Hint: {
							location: 'import',
							path: '../../utils/common/common.interface',
							id: 'src/utils/common/common.interface.ts::Hint',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to include the ontario-hint-text component for the checkbox group.\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			hintExpander: {
				type: 'string',
				attribute: 'hint-expander',
				mutable: false,
				complexType: {
					original: 'HintExpander | string',
					resolved: 'HintExpander | string | undefined',
					references: {
						HintExpander: {
							location: 'import',
							path: '../ontario-hint-expander/hint-expander.interface',
							id: 'src/components/ontario-hint-expander/hint-expander.interface.ts::HintExpander',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-checkboxes\n  caption=\'{\n    "captionText": "Checkbox legend",\n    "captionType": "heading",\n  }\n  name=\'ontario-checkboxes\'\n  options=\'[\n\t{\n\t\t"value": "checkbox-option-1",\n\t\t"label": "Checkbox option 1 label",\n\t\t"elementId": "checkbox-1"\n\t}\n  }]\'\n  hint-expander=\'{\n   "hint": "Hint expander for the checkbox group",\n   "content": "Example hint expander content for the checkbox group"\n  }\'\n  required="true"\n>\n</ontario-checkboxes>',
						},
					],
					text: 'Used to include the ontario-hint-expander component for the checkbox group.\nThis is passed in as an object with key-value pairs.\n\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			options: {
				type: 'string',
				attribute: 'options',
				mutable: false,
				complexType: {
					original: 'CheckboxOption[] | string',
					resolved: 'CheckboxOption[] | string',
					references: {
						CheckboxOption: {
							location: 'import',
							path: './checkbox-option.interface',
							id: 'src/components/ontario-checkbox/checkbox-option.interface.ts::CheckboxOption',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-checkboxes\n  caption=\'{\n\t"captionText": "Checkbox legend",\n\t"captionType": "heading",\n  }\n  name="ontario-checkboxes",\n  hint-text="Hint text for the checkbox group."\n  options=\'[\n\t{\n\t\t"value": "checkbox-option-1",\n\t\t"label": "Checkbox option 1 label"\n\t\t"elementId": "checkbox-1"\n    },\n    {\n       "value": "checkbox-option-2",\n       "label": "Checkbox option 2 label",\n\t\t  "elementId": "checkbox-2",\n      "hintExpander": {\n\t\t\t"hint": "Hint expander for checkbox option 2",\n             "content": "Example hint expander content for checkbox option 2"\n       }\n     }\n  ]\'\n  required="true"\n>\n</ontario-checkboxes>',
						},
					],
					text: 'The options for the checkbox group.\n\nEach property will be passed in through an object in the options array.\nThis can either be passed in as an object directly (if using react), or as a string in HTML.\nIf there are multiple checkboxes in a fieldset, each checkbox will be displayed as an option.\n\nIn the example below, the options are being passed in as a string and there are two checkboxes to be displayed in the fieldset.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			required: {
				type: 'boolean',
				attribute: 'required',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'This is used to determine whether the checkbox is required or not.\nThis prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.\nIf no prop is set, it will default to false (optional).',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			errorMessage: {
				type: 'string',
				attribute: 'error-message',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Set this to display an error message',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			customOnChange: {
				type: 'unknown',
				attribute: 'custom-on-change',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the checkbox onChange event.',
				},
				getter: false,
				setter: false,
			},
			customOnBlur: {
				type: 'unknown',
				attribute: 'custom-on-blur',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the checkbox onBlur event.',
				},
				getter: false,
				setter: false,
			},
			customOnFocus: {
				type: 'unknown',
				attribute: 'custom-on-focus',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the checkbox onFocus event.',
				},
				getter: false,
				setter: false,
			},
		};
	}
	static get states() {
		return {
			hintTextId: {},
			captionState: {},
			internalHintText: {},
			internalHintExpander: {},
			internalOptions: {},
		};
	}
	static get events() {
		return [
			{
				method: 'checkboxOnChange',
				name: 'checkboxOnChange',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input or mouse event occurs when a checkbox option has been changed.',
				},
				complexType: {
					original: 'RadioAndCheckboxChangeEvent',
					resolved: 'InputInteractionEvent & { checked: boolean; }',
					references: {
						RadioAndCheckboxChangeEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::RadioAndCheckboxChangeEvent',
						},
					},
				},
			},
			{
				method: 'checkboxOnBlur',
				name: 'checkboxOnBlur',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when a checkbox option has lost focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'checkboxOnFocus',
				name: 'checkboxOnFocus',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when a checkbox option has gained focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'inputErrorOccurred',
				name: 'inputErrorOccurred',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when an error message is reported to the component.',
				},
				complexType: {
					original: '{ errorMessage: string }',
					resolved: '{ errorMessage: string; }',
					references: {},
				},
			},
		];
	}
	static get elementRef() {
		return 'element';
	}
	static get watchers() {
		return [
			{
				propName: 'hintText',
				methodName: 'parseHintText',
			},
			{
				propName: 'hintExpander',
				methodName: 'parseHintExpander',
			},
			{
				propName: 'options',
				methodName: 'parseOptions',
			},
			{
				propName: 'name',
				methodName: 'validateName',
			},
			{
				propName: 'caption',
				methodName: 'updateCaptionState',
			},
			{
				propName: 'language',
				methodName: 'updateLanguage',
			},
			{
				propName: 'errorMessage',
				methodName: 'broadcastInputErrorOccurredEvent',
			},
		];
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
		];
	}
	static get attachInternalsMemberName() {
		return 'internals';
	}
}
