// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-input/ontario-input.js
import { h } from '@stencil/core';
import { v4 as uuid } from 'uuid';
import { Input } from '../../utils/common/input/input';
import { InputCaption } from '../../utils/common/input-caption/input-caption';
import { validateLanguage, validatePropExists } from '../../utils/validation/validation-functions';
import { translations as globalTranslations } from '../../translations';
import { constructHintTextObject } from '../../utils/components/hints/hints';
import { EventType } from '../../utils/events/event-handler.interface';
import { handleInputEvent } from '../../utils/events/event-handler';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { ErrorMessage } from '../../utils/components/error-message/error-message';
export class OntarioInput {
	element;
	internals;
	hintTextRef;
	/**
	 * The text to display as the input label
	 *
	 * @example
	 * <ontario-input
	 *   caption='{
	 *     "captionText": "Address",
	 *     "captionType": "heading",
	 *   }
	 *   required="true"
	 *   ...>
	 * </ontario-input>
	 */
	caption;
	/**
	 * The unique identifier of the input. This is optional - if no ID is passed, one will be generated.
	 */
	elementId;
	/**
	 * The width of the input field. If no value is assigned, it will present as the `default` input width.
	 */
	inputWidth = 'default';
	/**
	 * The name assigned to the input. The name value is used to reference form data after a form is submitted.
	 */
	name;
	/**
	 * Used to include the ontario-hint-text component for the input.
	 * This is optional.
	 */
	hintText;
	/**
	 * This is used to determine whether the input is required or not.
	 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
	 * If no prop is set, it will default to false (optional).
	 *
	 * _Please add a validation messaging using `requiredValidationMessage` if setting this property._
	 *
	 * @example
	 * <ontario-input
	 *		id="address-line-1"
	 *		caption="Address line 1"
	 *		required
	 *		required-validation-message="Please enter an address, including street number and street name"
	 *		name="address-line-1"
	 *		hint-text="Street and number or P.O. box."
	 *	></ontario-input>
	 */
	required = false;
	/**
	 * The input type value.
	 *
	 * If no `type` is provided, it will default to 'text'.
	 */
	type = 'text';
	/**
	 * The input content value.
	 *
	 * This is optional.
	 */
	value;
	/**
	 * Set this to display an
	 */
	errorMessage;
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
	 */
	language = 'en';
	/**
	 * Used to include the ontario-hint-expander component for the input component.
	 * This is passed in as an object with key-value pairs.
	 *
	 * This is optional.
	 *
	 * @example
	 * <ontario-input
	 *   caption='{
	 *     "caption": "Address",
	 *     "captionType": "heading",
	 *   }
	 *   hint-expander='{
	 *    "hint": "Hint expander",
	 *    "content": "This is the content"
	 *   }'
	 *   required="true"
	 * >
	 * </ontario-input>
	 */
	hintExpander;
	/**
	 * Used for the `aria-describedby` value of the input. This will match with the id of the hint text.
	 */
	hintTextId;
	/**
	 * Enable live validation on the input.  Custom live validation can be performed using an `inputValidator`
	 * validation function.  It will also validate the `required` state if no errors are returned from
	 * the `inputValidator`.  Please set a `requiredValidationMessage` to report concisely to the end user what
	 * they are required to set.
	 */
	enableLiveValidation = false;
	/**
	 * Validate the validity of the input value `onBlur`.  This `async` function should return a result
	 * to trigger an error message.  Returning `undefined` or `null` will clear it.
	 */
	inputValidator;
	/**
	 * Used to add a custom function to the input onInput event.
	 */
	customOnInput;
	/**
	 * Used to add a custom function to the input onChange event.
	 */
	customOnChange;
	/**
	 * Used to add a custom function to the input onBlur event.
	 */
	customOnBlur;
	/**
	 * Used to add a custom function to the input onFocus event.
	 */
	customOnFocus;
	/**
	 * Custom error message to display if a required field is not filled out.  _Please add a
	 * custom message when setting an input as required_.
	 */
	requiredValidationMessage;
	/**
	 * The hint text options are re-assigned to the internalHintText array.
	 */
	internalHintText;
	/**
	 * The hint expander options are re-assigned to the internalHintExpander array.
	 */
	internalHintExpander;
	/**
	 * Instantiate an InputCaption object for internal logic use
	 */
	captionState;
	/**
	 * Track if the input has been interacted with, used to validate if
	 * a `required` field is in error.
	 */
	hasBeenInteractedWith = false;
	/**
	 * Global translations for accessing built-in translations
	 */
	translations = globalTranslations;
	/**
	 * Emitted when a input  occurs when an input has been changed.
	 */
	inputOnInput;
	/**
	 * Emitted when a keyboard input or mouse event occurs when an input has been changed.
	 */
	inputOnChange;
	/**
	 * Emitted when a keyboard input event occurs when an input has lost focus.
	 */
	inputOnBlur;
	/**
	 * Emitted when a keyboard input event occurs when an input has gained focus.
	 */
	inputOnFocus;
	/**
	 * Emitted when an error message is reported to the component.
	 */
	inputErrorOccurred;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 * @param event The language that has been selected.
	 */
	handleSetAppLanguage(event) {
		this.language = validateLanguage(event);
	}
	/**
	 * Handles an update to the language should the user request a language update from the language toggle.
	 * @param event The language that has been selected.
	 */
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event.detail.newLanguage);
	}
	/**
	 * Handle the change in the `value` property and validate if the input has been interacted with by
	 * the user to aid in determining if the required state should produce an error.
	 */
	handleValueChange() {
		this.hasBeenInteractedWith = this.hasBeenInteractedWith || !!this.value;
	}
	/*
	 * Watch for changes in the `name` prop for validation purposes.
	 *
	 * Validate the `name` and make sure the `name` prop has a value.
	 * Log a warning if user doesn't input a value for the `name`.
	 */
	validateName(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' name ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-input> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/**
	 * Watch for changes to the `hintText` prop.
	 *
	 * If a `hintText` prop is passed, the `constructHintTextObject` function will convert it to the correct format, and set the result to the `internalHintText` state.
	 */
	parseHintText() {
		if (this.hintText) {
			const hintTextObject = constructHintTextObject(this.hintText);
			this.internalHintText = hintTextObject;
		}
	}
	/**
	 * Watch for changes to the `hintExpander` prop.
	 *
	 * If a `hintExpander` prop is passed, it will be parsed (if it is a string), and the result will be set to the `internalHintExpander` state.
	 */
	parseHintExpander() {
		const hintExpander = this.hintExpander;
		if (hintExpander) {
			if (typeof hintExpander === 'string') this.internalHintExpander = JSON.parse(hintExpander);
			else this.internalHintExpander = hintExpander;
		}
	}
	/**
	 * Watch for changes to the `caption` prop.
	 *
	 * The caption will be run through the InputCaption constructor to convert it to the correct format, and set the result to the `captionState` state.
	 * @param newValue: Caption | string
	 */
	updateCaptionState(newValue) {
		this.captionState = new InputCaption(
			this.element.tagName,
			newValue,
			this.translations,
			this.language,
			false,
			this.required,
		);
	}
	/**
	 * Watch for changes in the `language` prop to render either the English or French translations
	 */
	updateLanguage() {
		this.updateCaptionState(this.caption);
	}
	/**
	 * Handle the component being blurred and perform validation logic on the input.  Custom validation
	 * takes persistance, followed by validating the required state.
	 *
	 * Finally, an event is emitted to notify anything listening for the `inputErrorOccurred` that
	 * an error occurred.
	 */
	async handleComponentBlur() {
		if (this.enableLiveValidation) {
			// Call inputValidator function to perform custom validation
			const validationResult = this.inputValidator && this.inputValidator(this.value);
			await validationResult?.then((x) => (this.errorMessage = x?.errorMessage));
			// Validate the `required` field
			// Only report a required error if no other error message is reported via validation
			if (this.required && this.hasBeenInteractedWith && !validationResult)
				if (!this.value)
					this.errorMessage =
						this.requiredValidationMessage || this.translations.input.requiredFieldError[this.getComponentLanguage()];
				else this.errorMessage = undefined;
		}
	}
	broadcastInputErrorOccurredEvent() {
		// Emit event to notify anyone who wants to listen for errors occurring
		this.inputErrorOccurred.emit({ inputId: this.getId(), errorMessage: this.errorMessage ?? '' });
	}
	/**
	 * Function to handle input events and the information pertaining to the input to emit.
	 */
	handleEvent(event, eventType) {
		const input = event.target;
		// Update the component value to match the value of the input element.
		this.value = input?.value;
		this.internals?.setFormValue?.(this.value ?? '');
		handleInputEvent(
			event,
			eventType,
			input,
			this.inputOnChange,
			this.inputOnFocus,
			this.inputOnBlur,
			this.inputOnInput,
			'input',
			this.customOnChange,
			this.customOnFocus,
			this.customOnBlur,
			this.customOnInput,
			this.element,
		);
	}
	getId() {
		// A UUID is assigned in `componentWillLoad` if there is no given `elementId`.
		return this.elementId ?? '';
	}
	getValue() {
		return this.value ?? '';
	}
	getClass() {
		if (this.hintExpander) {
			return this.inputWidth === 'default'
				? `ontario-input ontario-input-hint-expander--true`
				: `ontario-input ontario-input--${this.inputWidth} ontario-input-hint-expander--true`;
		} else {
			return this.inputWidth === 'default' ? `ontario-input` : `ontario-input ontario-input--${this.inputWidth}`;
		}
	}
	getComponentLanguage() {
		return this.language ?? 'en';
	}
	/**
	 * If a `hintText` prop is passed, the id generated from it will be set to the internal `hintTextId` state to match with the input `aria-describedBy` attribute.
	 */
	async componentDidLoad() {
		this.hintTextId = await this.hintTextRef?.getHintTextId();
	}
	componentWillLoad() {
		this.updateCaptionState(this.caption);
		this.elementId = this.elementId ?? uuid();
		this.parseHintText();
		this.parseHintExpander();
		this.validateName(this.name);
		this.language = validateLanguage(this.language);
	}
	render() {
		const error = !!this.errorMessage;
		return h(
			'div',
			{
				key: 'a7222e6476454f5d3d7063306b93db212abb3b01',
				class: `ontario-form-group ${error ? 'ontario-input--error' : ''}`,
			},
			this.captionState.getCaption(this.getId(), !!this.internalHintExpander),
			this.internalHintText &&
				h('ontario-hint-text', {
					key: '89d1b50018069434bf7ff24d784b6dc065ea7205',
					hint: this.internalHintText.hint,
					hintContentType: this.internalHintText.hintContentType,
					ref: (el) => (this.hintTextRef = el),
				}),
			h(ErrorMessage, { key: 'dd096e7cfe445817532bc7c0cf681048f417c390', message: this.errorMessage, error: error }),
			h(Input, {
				'key': '6f8e1f78a9c9d3f21335ff5b3dca8f4d0d53e3a7',
				'aria-describedBy': this.hintTextId,
				'className': this.getClass(),
				'id': this.getId(),
				'name': this.name,
				'onInput': (e) => this.handleEvent(e, EventType.Input),
				'onChange': (e) => this.handleEvent(e, EventType.Change),
				'onBlur': (e) => this.handleEvent(e, EventType.Blur),
				'onFocus': (e) => this.handleEvent(e, EventType.Focus),
				'type': this.type,
				'value': this.getValue(),
				'required': !!this.required,
			}),
			this.internalHintExpander &&
				h('ontario-hint-expander', {
					key: 'f25a94ba009fa257660a0dbe6c77d00d28668351',
					hint: this.internalHintExpander.hint,
					content: this.internalHintExpander.content,
					hintContentType: this.internalHintExpander.hintContentType,
				}),
		);
	}
	static get is() {
		return 'ontario-input';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get formAssociated() {
		return true;
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-input.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-input.css'],
		};
	}
	static get properties() {
		return {
			caption: {
				type: 'string',
				attribute: 'caption',
				mutable: false,
				complexType: {
					original: 'Caption | string',
					resolved: 'Caption | string',
					references: {
						Caption: {
							location: 'import',
							path: '../../utils/common/input-caption/caption.interface',
							id: 'src/utils/common/input-caption/caption.interface.ts::Caption',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-input\n  caption=\'{\n    "captionText": "Address",\n    "captionType": "heading",\n  }\n  required="true"\n  ...>\n</ontario-input>',
						},
					],
					text: 'The text to display as the input label',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			elementId: {
				type: 'string',
				attribute: 'element-id',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The unique identifier of the input. This is optional - if no ID is passed, one will be generated.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			inputWidth: {
				type: 'string',
				attribute: 'input-width',
				mutable: true,
				complexType: {
					original:
						"| '2-char-width'\n\t\t| '3-char-width'\n\t\t| '4-char-width'\n\t\t| '5-char-width'\n\t\t| '7-char-width'\n\t\t| '10-char-width'\n\t\t| '20-char-width'\n\t\t| 'default'",
					resolved:
						'"10-char-width" | "2-char-width" | "20-char-width" | "3-char-width" | "4-char-width" | "5-char-width" | "7-char-width" | "default"',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The width of the input field. If no value is assigned, it will present as the `default` input width.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'default'",
			},
			name: {
				type: 'string',
				attribute: 'name',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The name assigned to the input. The name value is used to reference form data after a form is submitted.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			hintText: {
				type: 'string',
				attribute: 'hint-text',
				mutable: false,
				complexType: {
					original: 'string | Hint',
					resolved: 'Hint | string | undefined',
					references: {
						Hint: {
							location: 'import',
							path: '../../utils/common/common.interface',
							id: 'src/utils/common/common.interface.ts::Hint',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to include the ontario-hint-text component for the input.\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			required: {
				type: 'boolean',
				attribute: 'required',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-input\n\tid="address-line-1"\n\tcaption="Address line 1"\n\trequired\n\trequired-validation-message="Please enter an address, including street number and street name"\n\tname="address-line-1"\n\thint-text="Street and number or P.O. box."\n></ontario-input>',
						},
					],
					text: 'This is used to determine whether the input is required or not.\nThis prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.\nIf no prop is set, it will default to false (optional).\n\n_Please add a validation messaging using `requiredValidationMessage` if setting this property._',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			type: {
				type: 'string',
				attribute: 'type',
				mutable: true,
				complexType: {
					original: "'text' | 'tel' | 'email' | 'password'",
					resolved: '"email" | "password" | "tel" | "text"',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "The input type value.\n\nIf no `type` is provided, it will default to 'text'.",
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'text'",
			},
			value: {
				type: 'string',
				attribute: 'value',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The input content value.\n\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			errorMessage: {
				type: 'string',
				attribute: 'error-message',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Set this to display an',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr" | undefined',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'en'",
			},
			hintExpander: {
				type: 'string',
				attribute: 'hint-expander',
				mutable: false,
				complexType: {
					original: 'HintExpander | string',
					resolved: 'HintExpander | string | undefined',
					references: {
						HintExpander: {
							location: 'import',
							path: '../ontario-hint-expander/hint-expander.interface',
							id: 'src/components/ontario-hint-expander/hint-expander.interface.ts::HintExpander',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-input\n  caption=\'{\n    "caption": "Address",\n    "captionType": "heading",\n  }\n  hint-expander=\'{\n   "hint": "Hint expander",\n   "content": "This is the content"\n  }\'\n  required="true"\n>\n</ontario-input>',
						},
					],
					text: 'Used to include the ontario-hint-expander component for the input component.\nThis is passed in as an object with key-value pairs.\n\nThis is optional.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			enableLiveValidation: {
				type: 'boolean',
				attribute: 'enable-live-validation',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'Enable live validation on the input.  Custom live validation can be performed using an `inputValidator`\nvalidation function.  It will also validate the `required` state if no errors are returned from\nthe `inputValidator`.  Please set a `requiredValidationMessage` to report concisely to the end user what\nthey are required to set.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'false',
			},
			inputValidator: {
				type: 'unknown',
				attribute: 'input-validator',
				mutable: false,
				complexType: {
					original: '(value?: string) => Promise<{ errorMessage?: string } | null | undefined>',
					resolved:
						'((value?: string | undefined) => Promise<{ errorMessage?: string | undefined; } | null | undefined>) | undefined',
					references: {
						Promise: {
							location: 'global',
							id: 'global::Promise',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Validate the validity of the input value `onBlur`.  This `async` function should return a result\nto trigger an error message.  Returning `undefined` or `null` will clear it.',
				},
				getter: false,
				setter: false,
			},
			customOnInput: {
				type: 'unknown',
				attribute: 'custom-on-input',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the input onInput event.',
				},
				getter: false,
				setter: false,
			},
			customOnChange: {
				type: 'unknown',
				attribute: 'custom-on-change',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the input onChange event.',
				},
				getter: false,
				setter: false,
			},
			customOnBlur: {
				type: 'unknown',
				attribute: 'custom-on-blur',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the input onBlur event.',
				},
				getter: false,
				setter: false,
			},
			customOnFocus: {
				type: 'unknown',
				attribute: 'custom-on-focus',
				mutable: false,
				complexType: {
					original: '(event: globalThis.Event) => void',
					resolved: '((event: Event) => void) | undefined',
					references: {
						globalThis: {
							location: 'global',
							id: 'global::globalThis',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to add a custom function to the input onFocus event.',
				},
				getter: false,
				setter: false,
			},
			requiredValidationMessage: {
				type: 'string',
				attribute: 'required-validation-message',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'Custom error message to display if a required field is not filled out.  _Please add a\ncustom message when setting an input as required_.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			hintTextId: {},
			internalHintText: {},
			internalHintExpander: {},
			captionState: {},
			hasBeenInteractedWith: {},
			translations: {},
		};
	}
	static get events() {
		return [
			{
				method: 'inputOnInput',
				name: 'inputOnInput',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a input  occurs when an input has been changed.',
				},
				complexType: {
					original: 'InputInputEvent',
					resolved: 'InputInteractionEvent & { inputType?: string | undefined; }',
					references: {
						InputInputEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputInputEvent',
						},
					},
				},
			},
			{
				method: 'inputOnChange',
				name: 'inputOnChange',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input or mouse event occurs when an input has been changed.',
				},
				complexType: {
					original: 'InputInteractionEvent',
					resolved: '{ id?: string | undefined; value?: string | undefined; }',
					references: {
						InputInteractionEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputInteractionEvent',
						},
					},
				},
			},
			{
				method: 'inputOnBlur',
				name: 'inputOnBlur',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when an input has lost focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'inputOnFocus',
				name: 'inputOnFocus',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input event occurs when an input has gained focus.',
				},
				complexType: {
					original: 'InputFocusBlurEvent',
					resolved: 'InputInteractionEvent & { focused: boolean; }',
					references: {
						InputFocusBlurEvent: {
							location: 'import',
							path: '../../utils/events/event-handler.interface',
							id: 'src/utils/events/event-handler.interface.ts::InputFocusBlurEvent',
						},
					},
				},
			},
			{
				method: 'inputErrorOccurred',
				name: 'inputErrorOccurred',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when an error message is reported to the component.',
				},
				complexType: {
					original: '{ inputId: string; errorMessage: string }',
					resolved: '{ inputId: string; errorMessage: string; }',
					references: {},
				},
			},
		];
	}
	static get elementRef() {
		return 'element';
	}
	static get watchers() {
		return [
			{
				propName: 'value',
				methodName: 'handleValueChange',
			},
			{
				propName: 'name',
				methodName: 'validateName',
			},
			{
				propName: 'hintText',
				methodName: 'parseHintText',
			},
			{
				propName: 'hintExpander',
				methodName: 'parseHintExpander',
			},
			{
				propName: 'caption',
				methodName: 'updateCaptionState',
			},
			{
				propName: 'language',
				methodName: 'updateLanguage',
			},
			{
				propName: 'errorMessage',
				methodName: 'broadcastInputErrorOccurredEvent',
			},
		];
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'blur',
				method: 'handleComponentBlur',
				target: undefined,
				capture: true,
				passive: false,
			},
		];
	}
	static get attachInternalsMemberName() {
		return 'internals';
	}
}
