// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-footer/ontario-footer.js
import { h } from '@stencil/core';
import { ExpandedFooterWrapper, FooterColumn, SimpleFooter } from './components';
import { isInvalidTwoColumnOptions, isInvalidThreeColumnOptions } from './utils';
import { validateLanguage } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { ConsoleType } from '../../utils/console-message/console-message.enum';
import { getImageAssetSrcPath } from '../../utils/helper/assets';
import translations from '../../translations/global.i18n.json';
export class OntarioFooter {
	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.
	 */
	language;
	/**
	 * The type of footer to be rendered. If no prop is provided, it will default to the 'default' type.
	 */
	type = 'default';
	/**
	 * A prop that stores the required links for all footers.
	 * Available options are 'accessibilityLink', 'privacyLink', 'contactLink','termsOfUseLink' and 'printerLink'
	 */
	footerLinks;
	/**
	 * Social media links to render in the footer.
	 * Available options are 'facebook', 'twitter', 'instagram' and 'youtube'
	 */
	socialLinks;
	/**
	 * Stores the titles and content for the expanded two column footer.
	 */
	twoColumnOptions;
	/**
	 * Stores the titles and content for the expanded three column footer.
	 */
	threeColumnOptions;
	/**
	 * Top margin for the footer. By default, this prop is set to `true`, which adds a margin top value of `5rem`.
	 * If set to `false`, the top margin value will be set to zero.
	 */
	topMargin = true;
	/**
	 * The base path to an assets folder containing the Design System assets
	 */
	assetBasePath;
	translations = translations;
	footerLinksState;
	socialLinksState;
	twoColumnState;
	threeColumnState;
	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is
	 * connected to the DOM. It is used for the initial language when the input component loads.
	 */
	handleSetAppLanguage(event) {
		if (!this.language) {
			this.language = validateLanguage(event);
		}
	}
	handleHeaderLanguageToggled(event) {
		this.language = validateLanguage(event);
	}
	processFooterLinks() {
		this.parseOptions(this.footerLinks);
	}
	processSocialLinks() {
		this.parseOptions(this.socialLinks);
	}
	processTwoColumnOptions() {
		this.twoColumnOptions && this.parseOptions(this.twoColumnOptions);
		this.verifyTwoColumnOptions();
	}
	processThreeColumnOptions() {
		this.threeColumnOptions && this.parseOptions(this.threeColumnOptions);
		this.verifyThreeColumnOptions();
	}
	isTwoColumnLayout = () => this.type === 'twoColumn';
	isThreeColumnLayout = () => this.type === 'threeColumn';
	isExpandedLayout = () => this.isTwoColumnLayout() || this.isThreeColumnLayout();
	verifyTwoColumnOptions() {
		if (this.isTwoColumnLayout() && isInvalidTwoColumnOptions(this.twoColumnState)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' twoColumnOptions ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-footer> ')
				.addRegularText('were not fully set. Please review your values and ensure all options are set.')
				.printMessage();
		}
	}
	verifyThreeColumnOptions() {
		if (this.isThreeColumnLayout() && isInvalidThreeColumnOptions(this.threeColumnState)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' threeColumnOptions ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-footer> ')
				.addRegularText('were not fully set. Please review your values and ensure all options are set.')
				.printMessage();
		}
	}
	parseOptions(optionType) {
		const options = optionType;
		const isString = typeof options === 'string';
		if (!options) {
			return;
		}
		try {
			if (options === this.footerLinks) {
				this.footerLinksState = isString ? JSON.parse(options) : options;
			} else if (options === this.socialLinks) {
				this.socialLinksState = isString ? JSON.parse(options) : options;
			} else if (options === this.twoColumnOptions) {
				this.twoColumnState = isString ? JSON.parse(options) : options;
			} else {
				this.threeColumnState = isString ? JSON.parse(options) : options;
			}
		} catch (error) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addRegularText(' failed to parse props for ')
				.addMonospaceText('<ontario-footer>')
				.addRegularText(' in ')
				.addMonospaceText('parseOptions()')
				.addRegularText(' method \n ')
				.addMonospaceText(error.stack)
				.printMessage(ConsoleType.Error);
		}
	}
	/**
	 * Generate CSS url to the background image
	 * @returns path to the background image
	 */
	getBackgroundImagePath() {
		const supergraphicLogoFile = this.isExpandedLayout()
			? 'footer-expanded-supergraphic-logo.svg'
			: 'footer-default-supergraphic-logo.svg';
		return `url(${getImageAssetSrcPath(supergraphicLogoFile, this.assetBasePath)})`;
	}
	getFooterClasses() {
		let classes = 'ontario-footer ontario-footer--default';
		if (!this.topMargin) {
			classes = `${classes}  ontario-margin-top-0-!`;
		}
		return classes;
	}
	getFooterLinks() {
		const { language, translations, footerLinksState } = this;
		const { accessibilityLink, privacyLink, contactLink, termsOfUseLink, printerLink } = footerLinksState ?? {};
		const links = {
			accessibilityLink: {
				href: accessibilityLink?.href ?? translations.accessibilityLink.link[language],
				text: accessibilityLink?.text ?? translations.accessibilityLink.text[language],
			},
			privacyLink: {
				href: privacyLink?.href ?? translations.privacyLink.link[language],
				text: privacyLink?.text ?? translations.privacyLink.text[language],
			},
			printerLink: {
				href: printerLink?.href ?? translations.printerLink.link[language],
				text: printerLink?.text ?? translations.printerLink.text[language],
			},
		};
		if (contactLink) {
			links.contactLink = {
				href: contactLink.href,
				text: contactLink.text ?? translations.contactUs[language],
			};
		}
		if (termsOfUseLink) {
			links.termsOfUseLink = {
				href: termsOfUseLink.href,
				text: termsOfUseLink.text ?? translations.termsOfUse[language],
			};
		}
		return links;
	}
	componentWillLoad() {
		this.processFooterLinks();
		this.processSocialLinks();
		this.processTwoColumnOptions();
		this.processThreeColumnOptions();
		this.language = validateLanguage(this.language);
	}
	render() {
		const { socialLinksState, twoColumnState, threeColumnState, topMargin } = this;
		const footerLinks = this.getFooterLinks();
		if (this.isTwoColumnLayout()) {
			return h(
				ExpandedFooterWrapper,
				{ footerLinks: footerLinks, topMargin: topMargin, backgroundImagePath: this.getBackgroundImagePath() },
				h(FooterColumn, { data: twoColumnState.column1 }),
				h(FooterColumn, { data: twoColumnState.column2, socialLinks: socialLinksState }),
			);
		}
		if (this.isThreeColumnLayout()) {
			return h(
				ExpandedFooterWrapper,
				{ footerLinks: footerLinks, topMargin: topMargin, backgroundImagePath: this.getBackgroundImagePath() },
				h(FooterColumn, { data: threeColumnState.column1, isThreeColLayout: true, isFullWidthInMediumLayout: true }),
				h(FooterColumn, { data: threeColumnState.column2, isThreeColLayout: true }),
				h(FooterColumn, { data: threeColumnState.column3, socialLinks: socialLinksState, isThreeColLayout: true }),
			);
		}
		return h(
			'footer',
			{ class: this.getFooterClasses(), style: { '--imagePath': this.getBackgroundImagePath() } },
			h(SimpleFooter, { ...footerLinks }),
		);
	}
	static get is() {
		return 'ontario-footer';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-footer.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-footer.css'],
		};
	}
	static get assetsDirs() {
		return ['assets'];
	}
	static get properties() {
		return {
			language: {
				type: 'string',
				attribute: 'language',
				mutable: true,
				complexType: {
					original: 'Language',
					resolved: '"en" | "fr"',
					references: {
						Language: {
							location: 'import',
							path: '../../utils/common/language-types',
							id: 'src/utils/common/language-types.ts::Language',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The language of the component.\nThis is used for translations, and is by default set through event listeners checking for a language property from the header. If no language is passed, it will default to English.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			type: {
				type: 'string',
				attribute: 'type',
				mutable: false,
				complexType: {
					original: 'OntarioFooterType',
					resolved: '"default" | "threeColumn" | "twoColumn"',
					references: {
						OntarioFooterType: {
							location: 'import',
							path: './ontario-footer-interface',
							id: 'src/components/ontario-footer/ontario-footer-interface.tsx::OntarioFooterType',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "The type of footer to be rendered. If no prop is provided, it will default to the 'default' type.",
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'default'",
			},
			footerLinks: {
				type: 'string',
				attribute: 'footer-links',
				mutable: false,
				complexType: {
					original: 'FooterLinks | string',
					resolved: 'FooterLinks | string',
					references: {
						FooterLinks: {
							location: 'import',
							path: './ontario-footer-interface',
							id: 'src/components/ontario-footer/ontario-footer-interface.tsx::FooterLinks',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "A prop that stores the required links for all footers.\nAvailable options are 'accessibilityLink', 'privacyLink', 'contactLink','termsOfUseLink' and 'printerLink'",
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			socialLinks: {
				type: 'string',
				attribute: 'social-links',
				mutable: false,
				complexType: {
					original: 'FooterSocialLinksProps | string',
					resolved:
						'string | { facebook?: string | undefined; twitter?: string | undefined; instagram?: string | undefined; youtube?: string | undefined; }',
					references: {
						FooterSocialLinksProps: {
							location: 'import',
							path: './components',
							id: 'src/components/ontario-footer/components/index.ts::FooterSocialLinksProps',
						},
					},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: "Social media links to render in the footer.\nAvailable options are 'facebook', 'twitter', 'instagram' and 'youtube'",
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			twoColumnOptions: {
				type: 'string',
				attribute: 'two-column-options',
				mutable: false,
				complexType: {
					original: 'TwoColumnOptions | string',
					resolved: 'TwoColumnOptions | string | undefined',
					references: {
						TwoColumnOptions: {
							location: 'import',
							path: './ontario-footer-interface',
							id: 'src/components/ontario-footer/ontario-footer-interface.tsx::TwoColumnOptions',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Stores the titles and content for the expanded two column footer.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			threeColumnOptions: {
				type: 'string',
				attribute: 'three-column-options',
				mutable: false,
				complexType: {
					original: 'ThreeColumnOptions | string',
					resolved: 'ThreeColumnOptions | string | undefined',
					references: {
						ThreeColumnOptions: {
							location: 'import',
							path: './ontario-footer-interface',
							id: 'src/components/ontario-footer/ontario-footer-interface.tsx::ThreeColumnOptions',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Stores the titles and content for the expanded three column footer.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			topMargin: {
				type: 'boolean',
				attribute: 'top-margin',
				mutable: false,
				complexType: {
					original: 'boolean',
					resolved: 'boolean',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'Top margin for the footer. By default, this prop is set to `true`, which adds a margin top value of `5rem`.\nIf set to `false`, the top margin value will be set to zero.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: 'true',
			},
			assetBasePath: {
				type: 'string',
				attribute: 'asset-base-path',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'The base path to an assets folder containing the Design System assets',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			translations: {},
			footerLinksState: {},
			socialLinksState: {},
			twoColumnState: {},
			threeColumnState: {},
		};
	}
	static get watchers() {
		return [
			{
				propName: 'footerLinks',
				methodName: 'processFooterLinks',
			},
			{
				propName: 'socialLinks',
				methodName: 'processSocialLinks',
			},
			{
				propName: 'twoColumnOptions',
				methodName: 'processTwoColumnOptions',
			},
			{
				propName: 'threeColumnOptions',
				methodName: 'processThreeColumnOptions',
			},
		];
	}
	static get listeners() {
		return [
			{
				name: 'setAppLanguage',
				method: 'handleSetAppLanguage',
				target: 'window',
				capture: false,
				passive: false,
			},
			{
				name: 'headerLanguageToggled',
				method: 'handleHeaderLanguageToggled',
				target: 'window',
				capture: false,
				passive: false,
			},
		];
	}
}
