// @ongov/ontario-design-system-component-library/dist/collection/components/ontario-hint-expander/ontario-hint-expander.js
import { h } from '@stencil/core';
import { v4 as uuid } from 'uuid';
import { validatePropExists } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
export class OntarioHintExpander {
	host;
	/**
	 * The content type of the hint.
	 * If no prop is passed, it will default to a string.
	 * If the hint requires multiple lines or HTML, the `hintContentType` prop should be set to `html`.
	 */
	hintContentType = 'string';
	/**
	 * Text to display as the hint expander question/statement
	 */
	hint;
	/**
	 * Content to display as the hint, once the expander is toggled open.
	 * Please note that any content that is passed into this prop will only be displayed as a string.
	 * If you would like to add HTML content, supply child content to the component.
	 *
	 * @example
	 * <ontario-hint-expander hint="This is the hint"
	 *   <img src="https://www.jquery-az.com/html/images/banana.jpg" title="Title of image" alt="alt text here"/>
	 *   <p> Here is the content beside the image </p>
	 * </ontario-hint-expander>
	 */
	content;
	/**
	 * Used to used to establish a relationship between hint text content and elements using aria-describedby. This is optional - if no ID is passed, one will be generated.
	 */
	elementId;
	hintState;
	/**
	 * Emitted when a keyboard input or mouse event occurs.
	 */
	toggleExpanderEvent;
	/**
	 * Watch for changes to the `hintContentType` prop for validation purposes.
	 * If none is provided, or the wrong type is provided, it will default to `string`.
	 */
	checkHintContentType() {
		if (this.hintContentType !== 'string' && this.hintContentType !== 'html') {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' hintContentType ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-hint-text> ')
				.addRegularText('was not one of the permitted types. A default type of `string` will be applied.')
				.printMessage();
			return (this.hintContentType = 'string');
		}
		return this.hintContentType;
	}
	/*
	 * Watch for changes in the `hint` prop for validation purposes.
	 * If no `hint` prop is provided, the `hint` will be set to the host element textContent (if it exists).
	 */
	updateHintContent() {
		this.hintState = this.hint ?? this.host.textContent ?? '';
		this.validateHint(this.hintState);
	}
	/*
	 * Watch for changes to the `content` prop for validation purposes.
	 *
	 * Validate the `content` and make sure the `content` has a value.
	 * Log a warning if user doesn't input a value for the `content` or element content.
	 */
	validateContent(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' content ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-hint-expander> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	onClick = (ev) => {
		const hintExpander = ev.target;
		const hintExpanderParent = hintExpander?.parentElement;
		hintExpanderParent?.classList.toggle('ontario-expander--active');
		let content = hintExpanderParent?.querySelector("[data-toggle='ontario-expander-content']");
		content?.classList.toggle('ontario-expander__content--opened');
		content?.classList.contains('ontario-expander__content--opened')
			? content.setAttribute('aria-hidden', 'false')
			: content?.setAttribute('aria-hidden', 'true');
		hintExpanderParent?.classList.contains('ontario-expander--active')
			? hintExpanderParent?.setAttribute('aria-expanded', 'true')
			: hintExpanderParent?.setAttribute('aria-expanded', 'false');
		this.toggleExpanderEvent.emit(ev);
	};
	/*
	 * Watch for changes in the `hint` prop for validation purposes.
	 *
	 * Validate the `hint` and make sure the `hint` has a value.
	 * Log a warning if user doesn't input a value for the `hint`.
	 */
	validateHint(newValue) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' hint ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-hint-expander> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	/**
	 * Set `hint` using internal component logic
	 */
	componentWillLoad() {
		this.elementId = this.elementId ?? uuid();
		this.updateHintContent();
		this.checkHintContentType();
		this.validateContent(this.content);
	}
	/**
	 * This helper is used to help load translations for any slots + text content passed in by the user.
	 */
	componentDidLoad() {
		const observer = new MutationObserver((mutations) => {
			mutations.forEach((mutation) => {
				if (mutation.type === 'attributes') {
					this.updateHintContent();
				}
			});
		});
		const options = { attributes: true };
		observer.observe(this.host, options);
	}
	getId() {
		return this.elementId ?? '';
	}
	render() {
		return h(
			'div',
			{ key: 'de3d223cc99fe939b37472aeb127046ad1e7cfb4', class: 'ontario-hint-expander__container' },
			h(
				'button',
				{
					'key': '3102d97aa969066c32c20a8967eff971e0134f76',
					'class': 'ontario-hint-expander__button',
					'onClick': this.onClick,
					'id': `hint-expander-button-${this.getId()}`,
					'aria-controls': `hint-expander-content-${this.getId()}`,
					'aria-expanded': 'false',
					'data-toggle': 'ontario-collapse',
				},
				h(
					'span',
					{
						key: 'b8856b64b6f4319ccac743a6f2c9a742648a2b18',
						class: 'ontario-hint-expander__button-icon--close ontario-icon',
					},
					h('ontario-icon-chevron-up', { key: '6aa07816393bc8768e0304b58c66226e7502ca8a', colour: 'inherit' }),
				),
				h(
					'span',
					{ key: '5bfe9d549dfcce83d26ec351856d3cc1c1f4d9c2', class: 'ontario-hint-expander__button-icon--open' },
					h('ontario-icon-chevron-down', { key: '24650b2b0254fe5429cb43195d93f2b3a653b168', colour: 'inherit' }),
				),
				this.hint,
			),
			h(
				'div',
				{
					'key': '7925aa9f5f883ecb0739cdb068b1a1cfe2b40714',
					'class': 'ontario-hint-expander__content',
					'id': `hint-expander-content-${this.getId()}`,
					'aria-labelledby': `hint-expander-button-${this.getId()}`,
					'aria-hidden': 'true',
					'data-toggle': 'ontario-expander-content',
				},
				this.hintContentType === 'string' ? this.content : h('span', { innerHTML: this.content }),
			),
		);
	}
	static get is() {
		return 'ontario-hint-expander';
	}
	static get encapsulation() {
		return 'shadow';
	}
	static get originalStyleUrls() {
		return {
			$: ['ontario-hint-expander.scss'],
		};
	}
	static get styleUrls() {
		return {
			$: ['ontario-hint-expander.css'],
		};
	}
	static get properties() {
		return {
			hintContentType: {
				type: 'string',
				attribute: 'hint-content-type',
				mutable: true,
				complexType: {
					original: 'HintContentType',
					resolved: '"html" | "string" | undefined',
					references: {
						HintContentType: {
							location: 'import',
							path: '../../utils/common/common.interface',
							id: 'src/utils/common/common.interface.ts::HintContentType',
						},
					},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'The content type of the hint.\nIf no prop is passed, it will default to a string.\nIf the hint requires multiple lines or HTML, the `hintContentType` prop should be set to `html`.',
				},
				getter: false,
				setter: false,
				reflect: false,
				defaultValue: "'string'",
			},
			hint: {
				type: 'string',
				attribute: 'hint',
				mutable: false,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [],
					text: 'Text to display as the hint expander question/statement',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			content: {
				type: 'string',
				attribute: 'content',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string',
					references: {},
				},
				required: false,
				optional: false,
				docs: {
					tags: [
						{
							name: 'example',
							text: '<ontario-hint-expander hint="This is the hint"\n  <img src="https://www.jquery-az.com/html/images/banana.jpg" title="Title of image" alt="alt text here"/>\n  <p> Here is the content beside the image </p>\n</ontario-hint-expander>',
						},
					],
					text: 'Content to display as the hint, once the expander is toggled open.\nPlease note that any content that is passed into this prop will only be displayed as a string.\nIf you would like to add HTML content, supply child content to the component.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
			elementId: {
				type: 'string',
				attribute: 'element-id',
				mutable: true,
				complexType: {
					original: 'string',
					resolved: 'string | undefined',
					references: {},
				},
				required: false,
				optional: true,
				docs: {
					tags: [],
					text: 'Used to used to establish a relationship between hint text content and elements using aria-describedby. This is optional - if no ID is passed, one will be generated.',
				},
				getter: false,
				setter: false,
				reflect: false,
			},
		};
	}
	static get states() {
		return {
			hintState: {},
		};
	}
	static get events() {
		return [
			{
				method: 'toggleExpanderEvent',
				name: 'toggleExpanderEvent',
				bubbles: true,
				cancelable: true,
				composed: true,
				docs: {
					tags: [],
					text: 'Emitted when a keyboard input or mouse event occurs.',
				},
				complexType: {
					original: 'MouseEvent | KeyboardEvent',
					resolved: 'KeyboardEvent | MouseEvent',
					references: {
						MouseEvent: {
							location: 'global',
							id: 'global::MouseEvent',
						},
						KeyboardEvent: {
							location: 'global',
							id: 'global::KeyboardEvent',
						},
					},
				},
			},
		];
	}
	static get elementRef() {
		return 'host';
	}
	static get watchers() {
		return [
			{
				propName: 'hintContentType',
				methodName: 'checkHintContentType',
			},
			{
				propName: 'hint',
				methodName: 'updateHintContent',
			},
			{
				propName: 'content',
				methodName: 'validateContent',
			},
		];
	}
}
